{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Gene-regulatory inference\n",
    "\n",
    "This notebook contains the pipeline for checking the applicability of regulatory network inference methods based on scRNA-seq. The methods are extracted from [here](https://academic.oup.com/bib/article/22/3/bbaa190/5904505), [here](https://saezlab.org/project/modeling-signaling-networks-from-single-cell-data/) and [here](https://www.nature.com/articles/s41592-019-0690-6)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The overview of GRN-inferring methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "| Method                               | Availability                                                                    | Code         | Pkg.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | Ref.   | Year    | Usability | Comment | Output | Installed |\n",
    "| - | - | - | - | - | - | - | - | - | - |\n",
    "| **Boolean model**                |\n",
    "| Boolean Pseudotime                   | https://github.com/fionahamey/Pseudotime-network-inference                      | Python       | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx6.jpeg?Expires=1643307662&Signature=Ah3Dhko7icUOOeDaCtXFt-gHkxhat1gSTDM3EBkdEQr9zyMuG5~bGX95~tfJgmvMuBNLtPq2epCuinIbR7jVEkG0oEN~NfkCDCnITbOQBqtx0ef8SBGbir0-8c8M51rmJTT9AlLatr7ftCPjPcRTCwcQkr2jAgp4s3qh37u1vrDap3fTmH9nccfKR4cxp474CSOiyd~8Oz1vNomwfkYp3uTRAQDDLKZMOascvxUZaBQwZjSJ-IuKG3LSNnwT2nAlecpa2JiKH2pXI0c6IzR0-g2AXNAZBoei4PTZK7kLhlUivpA~ZxKU1slCv1D2Jn0SbvBF-gXB3FdNFnw5n5bLFg__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA)  | [link](https://www.pnas.org/content/114/23/5822) | 2017 | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx7.jpeg?Expires=1643307662&Signature=mXlAW4~41tUP3wHnNNPtCP7FIqRnJTdp~MFd95pk3adn9NEhy4oJxb470f1nPLXNXFY8gmBgoQtfd3T13WONpL2P5jH1RnflBFYRKeEu8bYzPIlwf52lOiN-SikPt0JtUY9YfJEmBY6mdGyKKA2KuIxmVvOZYoBnGwhSKQzBVdfckb7iaAR7NicVdmIEuhB2wtnLU92ExJ82P6qx4Th5lQw41dphaufLjBMM89Naj20kkkwGJgyi~qGbjn76Tlpna3aUBNhdFzIOjP5AwJ53c9beqpEqLTtgQUvixxPq0Vj8uGbkLAqXaiAhk3c3o4--UlPiKnb2T~O~I~TpcKSSbg__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA)  | - | - | - |\n",
    "| BTR                                  | https://cran.r-project.org/package=BTR                                          | R            | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx8.jpeg?Expires=1643307662&Signature=ANEMHcKy6H9xbKfKNEC8c0UsN4PqK-nPVsVOkNozyhRcNAYOvZV53ffjuVQ9FJKHdhqoPu8RoDKihC6a8NculdeiLjk327A55liwFH8PV~lQUAAPC3epJ78~6eQvU~dOiUfDG~m6kFtW3XN3fhxGc1n22BOjOWdpX7AeyyYr7x8NLLzcwcwskuK0OvR8tRQgiKqoQCDROFy0Gv5Nu1NVJc4Xtxr6N1bCpbQ2Iz6Cik3ErB1CXBDK18IVDtbqYy-JP6Kuf6Cs1-xaWauu28Nbk2ajOp1L6qxNfbKUJNM0sLMahQ8BTJkzdUkQnP729aJN5Yw3lBzxxYwRAe2li1naIQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA)  | [link](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1235-y) | 2016      | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx9.jpeg?Expires=1643307662&Signature=mXMPQE9V9lmARSt2M69z2QSBSBkML384XOBeVuTnFjRYKOrz23rnddgeYpYdK9WwMMs5s~XhOf1LrgbsCNgcjzAWLe5E2BwWfWBHbEKbxg203MuwFNli5pXU8CEL~XsE954~HlYe2G9FjGVv2sBmtF6UzmCQwIVsqlLxRxn-M1T2fq3W4T8BM8AkVhVBpzVD-djCzq6cMvnR4nFsTjVtyobIhazF4NhNZQ54najY390kbggJ28lGfL3up9ryzL-2rT2dEjl6Me20IGvQyTz9cq3th07eC30F~AHYr8KEoQ8ASiGQPB6D0nhAPUTDgu1Hr2A5qDm8tr0uhDOVvUN~IQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA)  | - | - | - |\n",
    "| <u>SCNS</u>                                 | https://github.com/swoodhouse/SCNS-GUI                                          | F#           | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx10.jpeg?Expires=1643307662&Signature=Yw9IApy3JEHq8bCiLX0TF4P0xLCuqxsHGNybQsZLeo9Fcni1EFyMBw2JrbqNrArZyd597tCJlooGks-ijhlYGIpESzaf5v~Ne0ORsGnolICQrPL5y6C2I5gDwn2G~kPyK6sVI8GPWOUD7t8XJLy0w1tMoNbl3VKA0Yhwx2hfoK0fViyjjUlYAfLBK2ZXZLRttY51eogmlqPDFg4NV9J0Ri4O3V9bYEVSWS7ZFGzPx~dte5dEDQNTuHA5hOe7XCplT4RezKBALQt3-7A6W1jWpmu~B9cYKKcINjqxc4aAervpGEld5F4p1aY8tqjm1VulmV73W7ZC9fxMLm-FWFSs3g__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://bmcsystbiol.biomedcentral.com/articles/10.1186/s12918-018-0581-y) | 2018       | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx11.jpeg?Expires=1643307662&Signature=jefGbZ-FWVi6pHlAP3Lr21Z5vKKU1v1m281ZPVeZ7VLL9DfymfM1eQYK1jSnk2SsjVaSLX0GXKD9gCoOxnSV5D5YrypKnJ1SxBJrvuFTcoXe6FWerj1itLlMub6ZQe~sxRcdf6JGzpgwEItxQntwDHyYBYmdsrstlRcZuGUtEjWXNucmOPWl~9dPpc1a7voBMzP-riMNFd4JWw0D61wierR~ijtGYWMDUVDcdcQ-uEZTOTeC-zqR~2ZOO0drxpC~HB8pnUms1QBeRpQbG~~Scug2qJyRAsPZnGQ2dwjgD8mMjJaJuAjjMLRAKYCtRB9w4F6zOb4uGq1c3Y5zQTrWJA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | ![graphic](https://findicons.com/files/icons/2360/spirit20/20/system_delete_alt_02.png) |\n",
    "| **Differential equation**              |\n",
    "| Inference Snapshot                   | https://www.helmholtz-muenchen.de/fileadmin/ICB/software/inferenceSnapshot.zip | C++/Matlab   | \\*                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | [link](https://academic.oup.com/bioinformatics/article/31/12/i89/216346) | 2015        | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx12.jpeg?Expires=1643307662&Signature=ch8QnzJuE7F99E2sg3jrJJvHH2Bs7H2tEXTn0DpdP0uKkEe1rzDLz~Avn6d5RrXaDxwBTU2BmPEO9uGRR43cX20f-93YMURu2RpZ4wVpCGMKf-8wnUgjdsRgAo7jfLy6Elm4FdGH9C73kXf5Rw0zJ-utw9OHuDJVN-AP5pYXggFr~GC4rXTWT4spUBcHRyQ7IRzQyOktTTLYQAhOpkHNNbBgyGM8ngauVZbvSuWZhirpH7o93EFq4Jk2hb7QWR5Os3o4jd2R7BIAPPrZqbDKFO0RYQ0pqkJngUD5GHoSsqZsfdHTxoaSCEaBGHfB7A6IfHQ0l5lM-jWQeLrQFzVMlQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| <u>SCODE</u>                                | https://github.com/hmatsu1226/SCODE                                             | R/Julia/Ruby | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx13.jpeg?Expires=1643307662&Signature=aOBN9lHKAqHmblgX2HpnlrH6XX62e9T0Qlp4-OGORBfy5bt1MksSoD-cdX-Wix8hOhInSIeNyc7WZguGFss-5wsxJAlBysSDV~jJIwQMqLnEBeQc7ZbkLT6QuRdCyYAALmeV63k3NQqINOaNDuzscoukq2yKOQ3BdMmX4oKC3Cb9LGQ5XdMMQGT9kXtc25CCmAevFV-SxGz9mvJlWhENGi37NejaPEdOmZum~rXj-ASos0N0vSLo4HBWxUzQlY2XoxXb-uru3GpnySlwyMr5lde-00uWZ2uKQB45VbBrTwBj3xAbxudJ-XJHj22avC-0iFLT8~zdM0ceYZuLLt9m3Q__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://academic.oup.com/bioinformatics/article/33/15/2314/3100331) | 2017        | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx14.jpeg?Expires=1643307662&Signature=xLEWpYD4qFeLiIXnRI8KlbS-6aultYBLGUr2tQV5FzAJjgeiYWm4sEpGgBB5Vd4DIsc0Kqbm~4yQ9OGsDSuiysXpS4KuBThUg9mg-v2nGSF85lmPDX8DGnVJAzKq9euTB~A2uLSmTRQXyVeH2eXw~MqldZypZ48F0yEI9p7U2BKI4tpa2U8OP7eg-3ohkE4AdbcecALlN6B68E-w0tbxffBKsVXqvJARUG3PKhKEiLsAG-dipfWOTlLwBQWUdk3OgtrXe5BWK49Y4mkB0JuiM31EHMg-kfWrE2yE7QZAOzsaD5p3f-f02uagR4IDVMPZRVwFVOzwFMt7dr1jl7QMlQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| SCOUP                                | https://github.com/hmatsu1226/SCOUP                                             | C++          | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx15.jpeg?Expires=1643307662&Signature=mCKLp5BL6A8q-o3DVpIBBVnOBH3ci7U1paUiroVIBulKsqBk03FUH8UUnEDqqWlpxnLhrOLqwIf1y5L6GzySkCqShnxXVUMhCfceOfwF4pIYzYxQMvixq7DEQOCA5OEdjSv-VtEhXUeWAzFa81NnWgeLwDkKXvgQ5yu8Q0rW9QKg3wDBuQMT9tVW-y5Uz1rH7aUBOBnu6DV9qcrJ2uDFoJlknOzsMuHxXiJAzsjQXzR~L-hVep3xkIipC60aBwU7BFe0bDZZJkPqhIBEzJqpItlIFcoj5qJuJ5MYfDWrNlOsSSYh-quixSa5X969MF9lAkSQi8EAuc2ANED2EmUtBA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-1109-3) | 2016       | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx16.jpeg?Expires=1643307662&Signature=EQ9XctoIbrwesVb9Hlx62CMzaiHbRQbcEnEStBp1FlrJmhSgvF62DTJNiH79WEnKKWsax52iru4S--JzSsnU0~MqXkcIbYqG1HbHoU0pG5YGyYOt~ajqucO6vcfYcPMooxg501BFnnqQgux2b~I3GEpLdcAyFOa~p~k0zInelj21mKn28iSwz-lqoaduqPWjmFmRlhN1rFqLWRrz3q8R-yZOlj4lqfxKU1ayiS~lfXBFtaGM0pJJxqdWVB4Oc2HWZHiOPXJ~z0JYGpH5Nq8ixVFOV4CUkTRUzOMZd~SF-HML9I139raPtkrBvwJ6CMZbPTs5QNbYlgfP92MV3Ce~Sw__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| <u>GRISLI</u>                          | https://github.com/PCAubin/GRISLI                                               | Matlab     | - | [link](https://academic.oup.com/bioinformatics/article/36/18/4774/5858974) | 2020    | - | - | - | - |\n",
    "| **Gene correlation**                    |\n",
    "| Empirical Bayes                      | https://github.com/ananth-pallaseni/EmpiricalBayes.jl                           | Julia        | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx17.jpeg?Expires=1643307662&Signature=Rwng9AWKU1sbKovNL1rkAoJPvXMjLa2gT1t0y9Qo7lXYbTq4dBSuQOqeWraonsXGaBv1H9fL0TLj1YZ3dkgtCOU72GzW~SoNFhmWOV56cKFg4SoEZeTdaYX9HfU0GK5eCOPpoGK9oxrTUgdNEPQu78i2TUiAlBzn817DjvYdjQUB2J0~9xBamBiag7ed3iCLIAG6oyuedP~5hQHsj9EZ71NNTNTI1e~CYM0DIL1x9qceuStlCZ3Ab3xfCts~u2Ywj52utUMrjeGyrBehIYiFO7vHdVIZSeLgZBLQmeRg~l2~e0CTxxEUhM9xJh0oWQ7yjMaP6K6ch1~XOfPIcNfroQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://www.biorxiv.org/content/10.1101/264853v1.abstract) | 2018       | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx18.jpeg?Expires=1643307662&Signature=wvpua2t31sj9nZYR6QyJD0wr5VP0AgZFKcUbDpeqBx~Y5Al9LGa1olHFKlipu21oNJDz5O~d~5KDZC0DdOiJANNlKoxfLZHyJBfqQT6EJKhKDri3zLsQU-176mZC3xMwkh8baaEIyufDGhHW-Z6p6ExK672eRDcFXcVdinM~ulI7wMQxWoH~dK4cVLdUqcjHkGfJ5yXKawd57RxpHTT1ulcOWmR8dDsTKaRfUeMehKBEsBvLa1VDBFS1jciS~SiAbZr7Dfl8HyXCN1-huI1QpcmtbqNDqxH36Te~q5GsmW0IMgVi9AecrCg3xzHnuJIGz1tsyj~-btoISyPPZeM0Eg__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| <span style=\"color:green\"><u>Information Measures (PIDC)</u></span>               | https://github.com/Tchanders/InformationMeasures.jl                             | Julia        | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx19.jpeg?Expires=1643307662&Signature=cTUVEkISLo0XylwGeKLQSrPHA7qSE9RljGsK1evvYSFDQM2LRpHRvIgFFWmHSLCdR19jbxaNDmx1IOz~konALQHYN5An-FzCatp9nk8tdMXGON32rmD4hQhTBGhT~TG8I6XvG5iMT~GgZv5zk2AK3ODPqtH6Ctp84336-L07bYss3uj32JdMCPZYOyh8jjF~mikt3kJH5TBVEv8KBjjKSvj8YJERxInPVyCr8UBda7RGtXAWUytWCJPQVmCVdYcUCR~dn4C7bAFzEuF5r7BwCzLAjlKmlsHt7qGo9944uR1j24ZG2FEPYC47O1hWtYc0x3v~~Q7D9swcZtm-ghy1hA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://www.cell.com/cell-systems/fulltext/S2405-4712(17)30386-1?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2405471217303861%3Fshowall%3Dtrue) | 2017       | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx20.jpeg?Expires=1643307662&Signature=f26qJZXKiyXCSkUOa-FynC1nCO5qs4Df2PxRO0X~Ns1SVC5OLu28S8y7FScCFDeHT858wPqAXkubTGc6KbfvvEKe1uf~7noDR6xGumgLsRbYkqPE6TJ11y0fkwIkOXVvoxueu3GDJt8xi7vPe1COdH0KGWJM-iA3djaJ7Kpzk6Q05dXyucN5jcQ4in1dfwWGG1cFDT-9~YBJkSgqeBINgi492sjByxYi-J1vDmmL8MUbyQyTlVaC9eJ82EheMs3loh8~EM08jccQnA4-uJROJsMYarAQQlX07y81bjoBidDA1j45bRHROXNqKgPP7O9TUd~F51vtDJfHzDZUEdmNcQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | ![graphic](https://findicons.com/files/icons/2360/spirit20/20/system_tick_alt_03.png) |\n",
    "| <span style=\"color:red\"><u>SCRIBE</u></span>                        | https://github.com/aristoteleo/Scribe-py                                | Python     | - | [link](https://www.cell.com/cell-systems/fulltext/S2405-4712(20)30036-3?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2405471220300363%3Fshowall%3Dtrue) | 2020a        | - | - | - | - |\n",
    "| NLNET                                | https://cran.r-project.org/package=nlnet                                        | R            | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx21.jpeg?Expires=1643307662&Signature=DaH~pSS6ZexUd197UIflIY22~b3mBQKDwN-VX~ITVtyPcc6PZC9XUIVU2Dgc~HcFUYv8ca11BAu456olkJE7UF53XGDfcYWbDDtqfhuDdysRri80ZNhmKfmEVwTEUPgs2iiw108olfiB1Kd7mBM5Ly5aktAdMy5zz9Nw~SrHEvKT3i0cvt~ALlglEtFR98RpJttS1kHPmvr0wqz5XdA9HX1txBnGfz2pRRLDLnjbAR42wT2CtP32FrMs~HsFy339jzZ2C4EeMaTEIk-xyL1nZ6UXSMnChgoxHfmsf2vvyoXvVskGQtQXhnYD2s79z5MsjAKapRVQBRDu5Kcmml1jZQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0158247) | 2016     | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx22.jpeg?Expires=1643307662&Signature=becMRq~xvCh14X4aXiAJAk-wVQgwUadCa6BH86GIpvXyrGNFkPLuPimfOBA7vue-~3wm-794-yjpTWmICfFBPq7gGZQM0XbEHTU5VTXf67X813wqDZdE-XxBgN6xpwJPdxhcIbMJbbRt9gDlWBZqqUJ0K75Fj6m7u05OeJHlCBA2eNs7AW5KOobhm48NMQLEbZmpgkQexBwOKczqU0Yo3nGug2bOUoh9VkQPq5iFYTJ1zwgrfAhLK3fuKvPx8qwsMxB6HmGRfDrrmjoj8hDHHnyHMjhyrFajudISs904Wk5xEVqVtuFtYjxR2O7dEhQgWgb-2z9UT3ae2uNZgYMuKg__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| SINCERA                              | https://research.cchmc.org/pbge/sincera.html                                    | R            | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx23.jpeg?Expires=1643307662&Signature=SPIeoiG6-gFhHv0S7isxHMcaP2WzqQDQQtNA1-uHBbX6chsMPsi21IQMle6ylKWCa9vdHWX2qyqbPrj0y8L-r07o1Kr-q7~3ufB9cjbl9t-dMpxuVE3oOvHMB-ZVVrErKaW2erGxwvxanmPUMI29jXD-Wo-4m~IDTA-As5N5u4sF9uRgvvbUJGs0JfeOCfHFqY6aN6SaDzeMfnMeJu-9bxOx5L0~hXPL-bTS3iHsB14lrpuFi1Lmi0t3l9-BYWjQSo9s-nBYbx8QwZef43yMUQ5arebcYfzt56lRpWLFLwj~ebjv-NxsbOTtbPdx6JOB5DRux4qff94m1EDGFgY6Gw__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004575) | 2015     | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx24.jpeg?Expires=1643307662&Signature=H~1sds9Mmiql6lpDLkQc~Z-UTcXQZnihCTZcj~p6qbGVwszWsthslh3mVEbCmRWzkcIphzHnnJYy1glyXJmOT-YGWf51Dt7OccK4axFjet8Sme2pnOLKPY-1aL-HEjzY0wk4L3ZdJoqr0VvRBxvB6KVPchslJylh~W5fLNaAVlwn5GSyMQa5klMazcOjLG1cYfPB5vDLkU4FPRiOLdFGG0XXylSrqKiADCccCiP~ty-hvXBDq2XKvgFrHY-i21d6kw4x7L09A0w0BigghpYFDybbOKuAqC-cbMKM7KowQoKXYhPNVP5w5wZCLDrjKZ5BG8k6B3uMbHmf-~iEFxeqGg__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| SCENIC                               | https://github.com/aertslab/SCENIC                                              | R/Python     | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx25.jpeg?Expires=1643307662&Signature=wEZE~TvGPgytwT96v~~HQRCMnKMk68~S4EkaZ80QRu8VJSy9oCXgs5LQ-u4C10IC9gg4cCjlGDDx2aYj0wY6J-pfdlLKXw9eKViqW3SGi-XnnTe84aFp7ed2D9mqp3sSneLNLjDi~KX15O0r-8LUGbEsfw2BmwAErP-DsYJQSYFEs8t89BNmGy49JawBKpC2H-sv3864yXJQlBTh3emfD2RY17SmF649tNXrAFrbMdqwLBFx-JoMW7PV3JzQpyi8YmlAmmeL5LkWk1QqTUJKsNP2OkSNprp9deSYieZUfDeEJaPGR48Y~MB2Fgg9WKchoxVt06BqtktBS0Yhtl-QjQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://www.nature.com/articles/nmeth.4463) | 2017    | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx26.jpeg?Expires=1643307662&Signature=nfVQE49hEuoSKYSqZInp059IaVBkaTXLpL3eOhZZczQsKZuSuKSFo-56F-sE52bT-TGvmfcD6RuP7KNnDE7hEW7KhD7z2Kux2ke80c23hTEVEWTY0WOvhzJunK~viHkdiFMlJwf0Z9duhIcZ4FCaWhChV78dy~eWx74E~~yyzig9IJF1jMdIMpB22pTnup7yvQkfgkdb4Vn8FGteUlQNk4FnvmVIEPPAMmsO8~BTS0AsCAScIxus1AG27FlzG-1Ql-8HT5dFawx1YXjLbagm~3qx7XeUtF2MkNMeiJHp4RhroxLyPQt7t4kH0yZMMqYSLmAc6WSQUfl4zt0cX5pGxQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | Very popular, most citations, require a database of TFs and their targets | - | ![graphic](https://findicons.com/files/icons/2360/spirit20/20/system_tick_alt_03.png) |\n",
    "| <span style=\"color:green\"><u>GENIE3</u></span>                           | https://github.com/vahuynh/GENIE3                                               | R/Python/C/Matlab     | - | [link](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0012776) | 2010    | - | - | - | ![graphic](https://findicons.com/files/icons/2360/spirit20/20/system_tick_alt_03.png) |\n",
    "| <span style=\"color:green\"><u>GRNBoost2</u></span>                           | https://arboreto.readthedocs.io/                                               | Python     | - | [link](https://academic.oup.com/bioinformatics/article/35/12/2159/5184284?login=true) | 2019    | - | - | - | ![graphic](https://findicons.com/files/icons/2360/spirit20/20/system_tick_alt_03.png) |\n",
    "| <u>PPCOR</u>                          | https://cran.r-project.org/web/packages/ppcor                                               | R     | - | [link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4681537/) | 2015    | - | - | - | - |\n",
    "| **Correlation ensemble over pseudotime** |\n",
    "| <u>LEAP</u>                                 | https://cran.r-project.org/package=LEAP                                         | R            | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx27.jpeg?Expires=1643307662&Signature=lMNvHTqJlWPHRbzaQSswuL38AwczZw1nFB7YdTSJRb1dLKkCbYO-wlJEjiAqopzjy988Tc1zsNEWTGZ~klOkVfczT~RyfIzqGBdgU~Qm3iA-42wIL7vsGpuflBi7DBYw1ABskfalbph3zAtkCxT-8sfqNqUqZ0UY6HJRqhIG7T1sI653XGD1-erKj389Yt8CJ60xf3twbaQ~cccK7x4Q3gqs~zat8dYdN4ALwKGkHBzGCHxlXII~G6Cb90Wgb0IrN4CpBgWpvR9g~ON2~~rAno~buqAX9acpmQbPPK1IcWGBqdUM5t4Tzdnnc6G93BIaeN7IB2PWoUAhG0BoSc67lA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://academic.oup.com/bioinformatics/article/33/5/764/2557687?login=true) | 2016     | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx28.jpeg?Expires=1643307662&Signature=vnqtL1HtdzCJV2jZooS98IILedSTxd62QsCuxD~HFLUpHj2kKO7MAhuLdOec0eW5VLyB9zl-f7NLIk3ct7U57EejV21EXCxXJ51aMF97Li3aEtLcTEBPPjgOXWUnmpyPRiy~IezmJvYz9YU1HynskbvEhkME4gSgQfhJtqRPeQabZmqMy4EIhN6wwA0NQxsH2mN3awAlxHVq9PHkOQOJWWXXq3wrWX-K-WJbKwE6tEUfttpPQrAxuSEzY4gbil~uLPMrljCgM1dSr0phZmLEWZxTKg9wqUlvNb2RMLI3ok81AloVPd8Jw1legrfGbzfKUwjxPf5pv0AnWTIE2wuDRg__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| <u>SINCERITIES</u>                          | http://www.cabsel.ethz.ch/tools/sincerities.html                                | R/Matlab     | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx29.jpeg?Expires=1643307662&Signature=FsALgnaUh8YhVdm-5FU2KvbmRnjYycgXgZBS-YGNujrrr6OuraBzsCCkaFSxpyg1qNGDU4qIvoUNuq-cvDTtT6ETNfZ1ooAvv~nf9XNW3b3UA68ulckYGQXt4Z23MDpG2va1VyuUM2Ipio39d7d~C5pLDVpp07l09ph0Dm37DdDIUlGk71dT0J2HSmJsk2mgDi-HSE~74Unsd~Ko6pktcRc-gZX1Hse0Demxih8RMIOlOPOW93n0~PPmH3~UlMJ39P4dPDJjT6nUWx4IZTykLeu0F429jwrot-lhr-D40snYfLJ8dkdRl9GGdCV7GxOzrie8-~NTV2CZN8zZ1mk6HQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://academic.oup.com/bioinformatics/article/34/2/258/4158033) | 2017        | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx30.jpeg?Expires=1643307662&Signature=a3E~XcAjqMk9DshBBnyTTtzajNuWFUjl8wfBNKutZQQhWpENM6K~tvWy0GMbLWahvG-NqVlpX1QOe~zhHWSgmdQQBetxlEAXzwioYF9S5NftAsbq2cFGSqJzQ0Linnn0mookhp6dbvlQ7xe2AjJfeBDFjDlF83u9he-Yc8nJ5KdQ7k5Aw2cn805KnztPwl-kdHStGTTnqcHZocOy4gZGdV7V5auggB7U9whf-pEsL67NuuyYWRnzvUzjoV9GJdxxk1VLK2ogqqtwFAAfHKc7EgkeIH5ofyeIuQx4Qh6YyYa5jTTgQS51iief104Ee~26SB1eSJ5M3311V~GDDQTrMA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| SCIMITAR                             | https://github.com/dimenwarper/scimitar                                         | Python       | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx31.jpeg?Expires=1643307662&Signature=zTIoGQ4YyL1YerwBXaG2x348cYEfvUbRZcp3YZmTJ7rHDV7jygPZT8go8kNXcFLzwBQJnnCD7zWib2ZvEfm-CnaoGEyFjPecAkv7hxATyR9wVtFv4zdEL6bio7ItAQdgwA7ph~9ClbT8yBUR9vPAVTjtMWRlQEHN3bWPuxcnri0Ti2uBSCkN3GvSIITj7BzBHFfylNuBkRaf-9nhlMJDCWVbY~MbNd0lgXLLyA6Q5N7V1KSKUZvD-qH~aQLmVt3YNf2R06DUAosgy6tUzqS6Zz7KxhNMkaBVS07aE0-pUUVaf7w5Ma2ke42xNnuBF7kMJ4m3jxOC5RdQj765mU2B~g__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://www.worldscientific.com/doi/abs/10.1142/9789813207813_0053) | 2017       | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx32.jpeg?Expires=1643307662&Signature=VACNr~Ojj2TFmq6TWDCATnO3VOcpc-8uf-fhsd8ROK4br7cKEXOznZk4lH9HrxRBRJQt1y1Gqn3n9KjShu33f8-emIkGcqI5zlCz9Ij6vs86fhOY-y6aeXIT0bAgj0-U~L2G312is2LiubptHlKUEw4ZR7hSxZYff~ObVYITDAZDh-UQQyJ9SNNW0COcRYodykFH4b350X9XF95Y1dVtLxA4Du0r26IKICeSsaqLsxltu9UX7zjK5tBZk6KJpcniTdG1DKU4nsiGGt8gh0zNwkw8RvVlqlwy-6hRrRuphZYiMH-dPydyN4aGlCdFVRQCcqBajDaU2cP-rQdlZYk0oQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| <u>SCINGE (SINGE)</u>                               | https://github.com/gitter-lab/SCINGE                                            | Matlab       | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx33.jpeg?Expires=1643307662&Signature=3fiC0QRaWwUBJOaoeAUlXcymVbV6JHeHs6j6a2VbxZXhrehFFRXiVQZxkoLEkomukywIXKk39vXF9FUbJC2TuoyPncAqCf4aCD6617jcniJ9IMk7WvKzlzlmNYo3l0g8y9OsV5ZfjPxbW5aScJQUnnVIQIhvT6sV0p-3HSWWO-qaogvS5nyhyzjiTB7WKm7eVqht3c7YI9zFSyAkebM6yTUmP22OLT8E3lcnrTjgl~vF9-PMvenEG-pRvBhTI5xjNEbQ2w7CavsM2EABAqCHByXZyvidflNdol-gsn6kDijp22dRd9032J46GedcdoVgG0ouOiP9zekMLsSG1f3pnA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://www.biorxiv.org/content/10.1101/534834v2.abstract) | 2019        | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx34.jpeg?Expires=1643307662&Signature=VVtdvKfYYDX1c2wagrkA8SI5RM2W6OqI46kkuOZ4Pl-BNH9Z1OcGIHfy7jOiE2govhnZYYa6YCvrcEjvxRNqDVEqiBit~mxMrCTA2VHDWyvwEVSfF9MOBuiwVRjXAjuTWN3LJvyoUu4aMrKJkoxXQzOoq2rall3nXW6bxLuW2dTsEbKCbzXQrk5x1KSTfza37qZIVa-0hKHJ36mOtnzntTfrIHPxU3ExzlWDiOLRJVM33IIdszCyn7yRXxK346UVUXTRLBYl-sATwGSLzvgFx3OVtj5Vbp7vx1OwDhfIRU0W-BXo6cfVv9FGB45~a6MUWpyeKPY23cV~XAl9TdeClQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | - | - | - |\n",
    "| <u>GRNVBEM</u>.                          | https://github.com/mscastillo/GRNVBEM                                               | Matlab     | - | [link](https://academic.oup.com/bioinformatics/article/34/6/964/4222631) | 2018    | - | - | - | - |\n",
    "| **Saez-Rodriguez tools** |\n",
    "| PROGENy                                 | https://saezlab.github.io/progeny                                         | R            | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx27.jpeg?Expires=1643307662&Signature=lMNvHTqJlWPHRbzaQSswuL38AwczZw1nFB7YdTSJRb1dLKkCbYO-wlJEjiAqopzjy988Tc1zsNEWTGZ~klOkVfczT~RyfIzqGBdgU~Qm3iA-42wIL7vsGpuflBi7DBYw1ABskfalbph3zAtkCxT-8sfqNqUqZ0UY6HJRqhIG7T1sI653XGD1-erKj389Yt8CJ60xf3twbaQ~cccK7x4Q3gqs~zat8dYdN4ALwKGkHBzGCHxlXII~G6Cb90Wgb0IrN4CpBgWpvR9g~ON2~~rAno~buqAX9acpmQbPPK1IcWGBqdUM5t4Tzdnnc6G93BIaeN7IB2PWoUAhG0BoSc67lA__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://www.nature.com/articles/s41467-017-02391-6) | 2018    | - | Signatures of pathways | - | ![graphic](https://findicons.com/files/icons/2360/spirit20/20/system_tick_alt_03.png) |\n",
    "| DoRothEA                          | https://saezlab.github.io/dorothea                                | R     | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx29.jpeg?Expires=1643307662&Signature=FsALgnaUh8YhVdm-5FU2KvbmRnjYycgXgZBS-YGNujrrr6OuraBzsCCkaFSxpyg1qNGDU4qIvoUNuq-cvDTtT6ETNfZ1ooAvv~nf9XNW3b3UA68ulckYGQXt4Z23MDpG2va1VyuUM2Ipio39d7d~C5pLDVpp07l09ph0Dm37DdDIUlGk71dT0J2HSmJsk2mgDi-HSE~74Unsd~Ko6pktcRc-gZX1Hse0Demxih8RMIOlOPOW93n0~PPmH3~UlMJ39P4dPDJjT6nUWx4IZTykLeu0F429jwrot-lhr-D40snYfLJ8dkdRl9GGdCV7GxOzrie8-~NTV2CZN8zZ1mk6HQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6673718/) | 2018        | - | Inferred TFs activity | - | ![graphic](https://findicons.com/files/icons/2360/spirit20/20/system_tick_alt_03.png) |\n",
    "| CARNIVAL                          | https://saezlab.github.io/CARNIVAL                                | R     | ![graphic](https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bib/22/3/10.1093_bib_bbaa190/2/m_bbaa190fx29.jpeg?Expires=1643307662&Signature=FsALgnaUh8YhVdm-5FU2KvbmRnjYycgXgZBS-YGNujrrr6OuraBzsCCkaFSxpyg1qNGDU4qIvoUNuq-cvDTtT6ETNfZ1ooAvv~nf9XNW3b3UA68ulckYGQXt4Z23MDpG2va1VyuUM2Ipio39d7d~C5pLDVpp07l09ph0Dm37DdDIUlGk71dT0J2HSmJsk2mgDi-HSE~74Unsd~Ko6pktcRc-gZX1Hse0Demxih8RMIOlOPOW93n0~PPmH3~UlMJ39P4dPDJjT6nUWx4IZTykLeu0F429jwrot-lhr-D40snYfLJ8dkdRl9GGdCV7GxOzrie8-~NTV2CZN8zZ1mk6HQ__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA) | [link](https://www.nature.com/articles/s41540-019-0118-z) | 2018        | - | Inferred paths | - | ![graphic](https://findicons.com/files/icons/2360/spirit20/20/system_tick_alt_03.png) |\n",
    "\n",
    "*<div style=\"text-align: right\"><u>Underlined</u> methods are benchmarked</div>*\n",
    "*<div style=\"text-align: right\"><span style=\"color:green\">Green</span> are best </div>*\n",
    "*<div style=\"text-align: right\"><span style=\"color:red\">Red</span> are worst </div>*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data\n",
    "\n",
    "We will start with the scRNA-seq dataset from COVID-19 patients, which is provided [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926). The data contains samples from 10x genomics pipeline that were obtains from the lung immune microenvironment in the bronchoalveolar lavage fluid (BALF) from <span style=\"color:red\">6 severe</span> and <span style=\"color:yellow\">3 moderate</span> COVID-19 patients and <span style=\"color:green\">3 healthy</span> controls."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparations..\n",
    "### Setting up Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# General purpose\n",
    "import sys\n",
    "import os\n",
    "import inspect\n",
    "import warnings\n",
    "import time\n",
    "import re\n",
    "from glob import glob  # for managing the files\n",
    "\n",
    "# Data managing\n",
    "import json  # working with json format\n",
    "import ndex2  # GRN database\n",
    "import math\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "# Tools/utils\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from collections import Counter\n",
    "from pyscenic.utils import load_motifs\n",
    "from functools import reduce  # for aggregate functions\n",
    "from termcolor import colored  # colored text output\n",
    "from tqdm.notebook import tqdm  # logging\n",
    "\n",
    "# Graphs\n",
    "import networkx as nx\n",
    "import pygraphviz as pgv\n",
    "import igraph as ig\n",
    "import leidenalg as la\n",
    "from community import community_louvain\n",
    "\n",
    "# Visualization\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import colorcet as cc\n",
    "from matplotlib.colors import ListedColormap\n",
    "from wordcloud import WordCloud\n",
    "\n",
    "# Setting working directory as home\n",
    "home_dir = os.path.expanduser('~')\n",
    "os.chdir(os.path.expanduser('~'))\n",
    "\n",
    "# Custom funcs\n",
    "from src.func import *\n",
    "\n",
    "# Arboreto\n",
    "from arboreto.utils import load_tf_names\n",
    "from arboreto.algo import genie3, grnboost2\n",
    "\n",
    "# Jupyter utils\n",
    "from IPython.display import display, HTML, SVG, set_matplotlib_formats\n",
    "from ipywidgets import *\n",
    "\n",
    "# Adding home directory to the PYTHONPATH\n",
    "sys.path.insert(0, home_dir)\n",
    "\n",
    "# Bridge to R\n",
    "import rpy2.rinterface\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "# To reload packages\n",
    "%load_ext autoreload\n",
    "# %autoreload 2\n",
    "%config InlineBackend.print_figure_kwargs = {'facecolor' : \"w\"}\n",
    "# To obtain line-by-line profile\n",
    "%load_ext line_profiler\n",
    "\n",
    "# PARAMETERS\n",
    "_NODE_SIZE = 1200\n",
    "plt.rcParams['xtick.labelsize'] = 15\n",
    "plt.rcParams['ytick.labelsize'] = 15\n",
    "plt.rcParams['axes.labelsize'] = 17\n",
    "plt.rcParams['axes.titlesize'] = 20\n",
    "plt.rcParams['figure.dpi'] = 200"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Setting up R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressPackageStartupMessages({\n",
    "  #if (!require(\"pacman\")) install.packages(\"pacman\")\n",
    "  #list.of.packages <- c(\"BiocManager\",\"dplyr\",\"SingleR\",\"Matrix\",\"Seurat\",\"future\",\"pheatmap\",\"ggplot2\",\"optparse\",\"hdf5r\")\n",
    "  #BiocManager::install(\"SingleR\")\n",
    "  #BiocManager::install('limma')\n",
    "  #BiocManager::install('SingleCellExperiment')\n",
    "  #pacman::p_load(list.of.packages, character.only = TRUE)\n",
    "  library(SingleR)\n",
    "  library(Matrix)\n",
    "  library(Seurat)\n",
    "  library(dplyr)\n",
    "  library(future)\n",
    "  library(pheatmap)\n",
    "  library(ggplot2)\n",
    "  library(optparse)\n",
    "})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importing data + Setting parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "As an example, the expression matrix of C141 patient contains 1449 cells and 20271 genes:\n",
      "And as an example, the dataset from C141 patient contains 2000 highly-variable genes.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL669831.2</th>\n",
       "      <th>AL669831.5</th>\n",
       "      <th>FAM87B</th>\n",
       "      <th>LINC00115</th>\n",
       "      <th>FAM41C</th>\n",
       "      <th>SAMD11</th>\n",
       "      <th>NOC2L</th>\n",
       "      <th>KLHL17</th>\n",
       "      <th>PLEKHN1</th>\n",
       "      <th>...</th>\n",
       "      <th>BX004987.1</th>\n",
       "      <th>AC145212.1</th>\n",
       "      <th>MAFIP</th>\n",
       "      <th>AC011043.1</th>\n",
       "      <th>AL592183.1</th>\n",
       "      <th>AC007325.1</th>\n",
       "      <th>AC007325.4</th>\n",
       "      <th>AC007325.2</th>\n",
       "      <th>AL354822.1</th>\n",
       "      <th>AC240274.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGATGTCGG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAATCCGAT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATGGTCAT-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTTTAGCTG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCAATCACG-1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 20271 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    AL627309.1  AL669831.2  AL669831.5  FAM87B  LINC00115  \\\n",
       "AAACCTGAGATGTCGG-1         0.0         0.0         0.0     0.0        0.0   \n",
       "AAACCTGCAATCCGAT-1         0.0         0.0         0.0     0.0        0.0   \n",
       "AAACCTGCATGGTCAT-1         0.0         0.0         0.0     0.0        0.0   \n",
       "AAACCTGGTTTAGCTG-1         0.0         0.0         0.0     0.0        0.0   \n",
       "AAACCTGTCAATCACG-1         0.0         0.0         0.0     0.0        0.0   \n",
       "\n",
       "                    FAM41C  SAMD11  NOC2L  KLHL17  PLEKHN1  ...  BX004987.1  \\\n",
       "AAACCTGAGATGTCGG-1     0.0     0.0    0.0     0.0      0.0  ...         0.0   \n",
       "AAACCTGCAATCCGAT-1     0.0     0.0    0.0     0.0      0.0  ...         0.0   \n",
       "AAACCTGCATGGTCAT-1     0.0     0.0    0.0     0.0      0.0  ...         0.0   \n",
       "AAACCTGGTTTAGCTG-1     0.0     0.0    0.0     0.0      0.0  ...         0.0   \n",
       "AAACCTGTCAATCACG-1     0.0     0.0    0.0     0.0      0.0  ...         0.0   \n",
       "\n",
       "                    AC145212.1  MAFIP  AC011043.1  AL592183.1  AC007325.1  \\\n",
       "AAACCTGAGATGTCGG-1         0.0    0.0         0.0         0.0         0.0   \n",
       "AAACCTGCAATCCGAT-1         0.0    0.0         0.0         0.0         0.0   \n",
       "AAACCTGCATGGTCAT-1         0.0    0.0         0.0         0.0         0.0   \n",
       "AAACCTGGTTTAGCTG-1         0.0    0.0         0.0         0.0         0.0   \n",
       "AAACCTGTCAATCACG-1         0.0    0.0         0.0         0.0         0.0   \n",
       "\n",
       "                    AC007325.4  AC007325.2  AL354822.1  AC240274.1  \n",
       "AAACCTGAGATGTCGG-1         0.0         0.0         0.0         0.0  \n",
       "AAACCTGCAATCCGAT-1         0.0         0.0         0.0         0.0  \n",
       "AAACCTGCATGGTCAT-1         0.0         0.0         0.0         0.0  \n",
       "AAACCTGGTTTAGCTG-1         0.0         0.0         0.0         0.0  \n",
       "AAACCTGTCAATCACG-1         0.0         0.0         0.0         0.0  \n",
       "\n",
       "[5 rows x 20271 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mFull metadata:\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>file</th>\n",
       "      <th>num_cells</th>\n",
       "      <th>Macrophage</th>\n",
       "      <th>T_cells</th>\n",
       "      <th>DC</th>\n",
       "      <th>Pre-B_cell_CD34-</th>\n",
       "      <th>Monocyte</th>\n",
       "      <th>NK_cell</th>\n",
       "      <th>B_cell</th>\n",
       "      <th>...</th>\n",
       "      <th>CMP</th>\n",
       "      <th>GMP</th>\n",
       "      <th>Gametocytes</th>\n",
       "      <th>Erythroblast</th>\n",
       "      <th>Neurons</th>\n",
       "      <th>Fibroblasts</th>\n",
       "      <th>Smooth_muscle_cells</th>\n",
       "      <th>Hepatocytes</th>\n",
       "      <th>Keratinocytes</th>\n",
       "      <th>Pro-Myelocyte</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C51</th>\n",
       "      <td>C</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>9431.0</td>\n",
       "      <td>8348.0</td>\n",
       "      <td>608.0</td>\n",
       "      <td>215.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C52</th>\n",
       "      <td>C</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>8696.0</td>\n",
       "      <td>8611.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C141</th>\n",
       "      <td>M</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>1449.0</td>\n",
       "      <td>197.0</td>\n",
       "      <td>932.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C142</th>\n",
       "      <td>M</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>1790.0</td>\n",
       "      <td>482.0</td>\n",
       "      <td>996.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C148</th>\n",
       "      <td>S</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>1165.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>122.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>641.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C149</th>\n",
       "      <td>S</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>1936.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>691.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C100</th>\n",
       "      <td>C</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>907.0</td>\n",
       "      <td>338.0</td>\n",
       "      <td>411.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C144</th>\n",
       "      <td>M</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>452.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C143</th>\n",
       "      <td>S</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>14933.0</td>\n",
       "      <td>2048.0</td>\n",
       "      <td>1394.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>7489.0</td>\n",
       "      <td>562.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C145</th>\n",
       "      <td>S</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>15550.0</td>\n",
       "      <td>6960.0</td>\n",
       "      <td>719.0</td>\n",
       "      <td>859.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>5616.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C146</th>\n",
       "      <td>S</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>2545.0</td>\n",
       "      <td>247.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>127.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C152</th>\n",
       "      <td>S</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>2557.0</td>\n",
       "      <td>466.0</td>\n",
       "      <td>397.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>795.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>322.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     group                                               file  num_cells  \\\n",
       "id                                                                         \n",
       "C51      C  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     9431.0   \n",
       "C52      C  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     8696.0   \n",
       "C141     M  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     1449.0   \n",
       "C142     M  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     1790.0   \n",
       "C148     S  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     1165.0   \n",
       "C149     S  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     1936.0   \n",
       "C100     C  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...      907.0   \n",
       "C144     M  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...      452.0   \n",
       "C143     S  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...    14933.0   \n",
       "C145     S  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...    15550.0   \n",
       "C146     S  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     2545.0   \n",
       "C152     S  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     2557.0   \n",
       "\n",
       "      Macrophage  T_cells     DC  Pre-B_cell_CD34-  Monocyte  NK_cell  B_cell  \\\n",
       "id                                                                              \n",
       "C51       8348.0    608.0  215.0              98.0      70.0     68.0     9.0   \n",
       "C52       8611.0     13.0   23.0               3.0      14.0      5.0     2.0   \n",
       "C141       197.0    932.0   48.0               5.0      86.0     96.0    33.0   \n",
       "C142       482.0    996.0   39.0              13.0      67.0    113.0    20.0   \n",
       "C148        98.0    122.0   24.0               NaN     641.0     36.0     8.0   \n",
       "C149       176.0    681.0   80.0               1.0     691.0     59.0    38.0   \n",
       "C100       338.0    411.0   45.0               5.0      51.0     20.0    12.0   \n",
       "C144        37.0    181.0   41.0               8.0      73.0     34.0    14.0   \n",
       "C143      2048.0   1394.0  154.0              33.0    7489.0    562.0    72.0   \n",
       "C145      6960.0    719.0  859.0              46.0    5616.0    421.0    58.0   \n",
       "C146       247.0     61.0   36.0               NaN     127.0     14.0     3.0   \n",
       "C152       466.0    397.0   41.0             174.0     795.0     74.0   322.0   \n",
       "\n",
       "      ...  CMP   GMP  Gametocytes  Erythroblast  Neurons  Fibroblasts  \\\n",
       "id    ...                                                               \n",
       "C51   ...  NaN   NaN          NaN           NaN      NaN          NaN   \n",
       "C52   ...  NaN   NaN          NaN           NaN      NaN          NaN   \n",
       "C141  ...  3.0   1.0          NaN           NaN      NaN          NaN   \n",
       "C142  ...  3.0   1.0          NaN           NaN      NaN          NaN   \n",
       "C148  ...  3.0   1.0          1.0           NaN      NaN          NaN   \n",
       "C149  ...  NaN   NaN          NaN           NaN      NaN          NaN   \n",
       "C100  ...  5.0   NaN          NaN           NaN      NaN          NaN   \n",
       "C144  ...  1.0   NaN          NaN           2.0      NaN          NaN   \n",
       "C143  ...  1.0   2.0          2.0           NaN      1.0          NaN   \n",
       "C145  ...  NaN   2.0          NaN           NaN      NaN          NaN   \n",
       "C146  ...  NaN   NaN          1.0           NaN      NaN          2.0   \n",
       "C152  ...  NaN  13.0          NaN           NaN      NaN          NaN   \n",
       "\n",
       "      Smooth_muscle_cells  Hepatocytes  Keratinocytes  Pro-Myelocyte  \n",
       "id                                                                    \n",
       "C51                   NaN          NaN            NaN            NaN  \n",
       "C52                   NaN          NaN            NaN            NaN  \n",
       "C141                  NaN          NaN            NaN            NaN  \n",
       "C142                  NaN          NaN            NaN            NaN  \n",
       "C148                  NaN          NaN            NaN            NaN  \n",
       "C149                  NaN          NaN            NaN            NaN  \n",
       "C100                  NaN          NaN            NaN            NaN  \n",
       "C144                  NaN          NaN            NaN            NaN  \n",
       "C143                  NaN          NaN            NaN            NaN  \n",
       "C145                  NaN          NaN            NaN            NaN  \n",
       "C146                  1.0          1.0            1.0            NaN  \n",
       "C152                  NaN          NaN            NaN            1.0  \n",
       "\n",
       "[12 rows x 25 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mCurrent metadata:\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>file</th>\n",
       "      <th>num_cells</th>\n",
       "      <th>T_cells</th>\n",
       "      <th>Macrophage</th>\n",
       "      <th>NK_cell</th>\n",
       "      <th>Monocyte</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C51</th>\n",
       "      <td>C</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>9431.0</td>\n",
       "      <td>608.0</td>\n",
       "      <td>8348.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>70.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C52</th>\n",
       "      <td>C</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>8696.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>8611.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C141</th>\n",
       "      <td>M</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>1449.0</td>\n",
       "      <td>932.0</td>\n",
       "      <td>197.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>86.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C142</th>\n",
       "      <td>M</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>1790.0</td>\n",
       "      <td>996.0</td>\n",
       "      <td>482.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C148</th>\n",
       "      <td>S</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>1165.0</td>\n",
       "      <td>122.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C149</th>\n",
       "      <td>S</td>\n",
       "      <td>/gpfs/projects/bsc08/bsc08890/data/GSE145926_R...</td>\n",
       "      <td>1936.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>691.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     group                                               file  num_cells  \\\n",
       "id                                                                         \n",
       "C51      C  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     9431.0   \n",
       "C52      C  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     8696.0   \n",
       "C141     M  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     1449.0   \n",
       "C142     M  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     1790.0   \n",
       "C148     S  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     1165.0   \n",
       "C149     S  /gpfs/projects/bsc08/bsc08890/data/GSE145926_R...     1936.0   \n",
       "\n",
       "      T_cells  Macrophage  NK_cell  Monocyte  \n",
       "id                                            \n",
       "C51     608.0      8348.0     68.0      70.0  \n",
       "C52      13.0      8611.0      5.0      14.0  \n",
       "C141    932.0       197.0     96.0      86.0  \n",
       "C142    996.0       482.0    113.0      67.0  \n",
       "C148    122.0        98.0     36.0     641.0  \n",
       "C149    681.0       176.0     59.0     691.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mCommon cell types among different patients (full metadata):\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>id</th>\n",
       "      <th>C51</th>\n",
       "      <th>C52</th>\n",
       "      <th>C141</th>\n",
       "      <th>C142</th>\n",
       "      <th>C148</th>\n",
       "      <th>C149</th>\n",
       "      <th>C100</th>\n",
       "      <th>C144</th>\n",
       "      <th>C143</th>\n",
       "      <th>C145</th>\n",
       "      <th>C146</th>\n",
       "      <th>C152</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Macrophage</th>\n",
       "      <td>8348.0</td>\n",
       "      <td>8611.0</td>\n",
       "      <td>197.0</td>\n",
       "      <td>482.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>338.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2048.0</td>\n",
       "      <td>6960.0</td>\n",
       "      <td>247.0</td>\n",
       "      <td>466.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T_cells</th>\n",
       "      <td>608.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>932.0</td>\n",
       "      <td>996.0</td>\n",
       "      <td>122.0</td>\n",
       "      <td>681.0</td>\n",
       "      <td>411.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>1394.0</td>\n",
       "      <td>719.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>397.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DC</th>\n",
       "      <td>215.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>154.0</td>\n",
       "      <td>859.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>41.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Monocyte</th>\n",
       "      <td>70.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>641.0</td>\n",
       "      <td>691.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>7489.0</td>\n",
       "      <td>5616.0</td>\n",
       "      <td>127.0</td>\n",
       "      <td>795.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NK_cell</th>\n",
       "      <td>68.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>562.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>74.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B_cell</th>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>322.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Epithelial_cells</th>\n",
       "      <td>7.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>145.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>417.0</td>\n",
       "      <td>201.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "id                   C51     C52   C141   C142   C148   C149   C100   C144  \\\n",
       "Macrophage        8348.0  8611.0  197.0  482.0   98.0  176.0  338.0   37.0   \n",
       "T_cells            608.0    13.0  932.0  996.0  122.0  681.0  411.0  181.0   \n",
       "DC                 215.0    23.0   48.0   39.0   24.0   80.0   45.0   41.0   \n",
       "Monocyte            70.0    14.0   86.0   67.0  641.0  691.0   51.0   73.0   \n",
       "NK_cell             68.0     5.0   96.0  113.0   36.0   59.0   20.0   34.0   \n",
       "B_cell               9.0     2.0   33.0   20.0    8.0   38.0   12.0   14.0   \n",
       "Epithelial_cells     7.0    25.0   37.0   38.0   52.0   41.0   18.0   54.0   \n",
       "\n",
       "id                  C143    C145   C146   C152  \n",
       "Macrophage        2048.0  6960.0  247.0  466.0  \n",
       "T_cells           1394.0   719.0   61.0  397.0  \n",
       "DC                 154.0   859.0   36.0   41.0  \n",
       "Monocyte          7489.0  5616.0  127.0  795.0  \n",
       "NK_cell            562.0   421.0   14.0   74.0  \n",
       "B_cell              72.0    58.0    3.0  322.0  \n",
       "Epithelial_cells   145.0   207.0  417.0  201.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36mALL\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_97c53_row0_col0, #T_97c53_row0_col1, #T_97c53_row0_col2, #T_97c53_row0_col3, #T_97c53_row0_col4, #T_97c53_row0_col5, #T_97c53_row0_col6, #T_97c53_row0_col7, #T_97c53_row0_col10, #T_97c53_row0_col11, #T_97c53_row1_col0, #T_97c53_row1_col1, #T_97c53_row1_col2, #T_97c53_row1_col3, #T_97c53_row1_col4, #T_97c53_row1_col5, #T_97c53_row1_col6, #T_97c53_row1_col7, #T_97c53_row1_col8, #T_97c53_row1_col9, #T_97c53_row1_col10, #T_97c53_row1_col11, #T_97c53_row2_col0, #T_97c53_row2_col1, #T_97c53_row2_col2, #T_97c53_row2_col3, #T_97c53_row2_col4, #T_97c53_row2_col5, #T_97c53_row2_col6, #T_97c53_row2_col7, #T_97c53_row2_col8, #T_97c53_row2_col9, #T_97c53_row2_col10, #T_97c53_row2_col11, #T_97c53_row3_col0, #T_97c53_row3_col1, #T_97c53_row3_col2, #T_97c53_row3_col3, #T_97c53_row3_col4, #T_97c53_row3_col5, #T_97c53_row3_col6, #T_97c53_row3_col7, #T_97c53_row3_col8, #T_97c53_row3_col9, #T_97c53_row3_col10, #T_97c53_row3_col11, #T_97c53_row4_col0, #T_97c53_row4_col1, #T_97c53_row4_col3, #T_97c53_row4_col5, #T_97c53_row4_col6, #T_97c53_row4_col7, #T_97c53_row4_col8, #T_97c53_row4_col9, #T_97c53_row4_col11, #T_97c53_row5_col0, #T_97c53_row5_col1, #T_97c53_row5_col2, #T_97c53_row5_col3, #T_97c53_row5_col4, #T_97c53_row5_col5, #T_97c53_row5_col6, #T_97c53_row5_col7, #T_97c53_row5_col8, #T_97c53_row5_col9, #T_97c53_row5_col10, #T_97c53_row5_col11, #T_97c53_row6_col0, #T_97c53_row6_col1, #T_97c53_row6_col2, #T_97c53_row6_col3, #T_97c53_row6_col4, #T_97c53_row6_col5, #T_97c53_row6_col6, #T_97c53_row6_col7, #T_97c53_row6_col8, #T_97c53_row6_col9, #T_97c53_row6_col10, #T_97c53_row6_col11, #T_97c53_row7_col0, #T_97c53_row7_col1, #T_97c53_row7_col2, #T_97c53_row7_col3, #T_97c53_row7_col4, #T_97c53_row7_col5, #T_97c53_row7_col6, #T_97c53_row7_col7, #T_97c53_row7_col8, #T_97c53_row7_col9, #T_97c53_row7_col10, #T_97c53_row7_col11, #T_97c53_row8_col0, #T_97c53_row8_col1, #T_97c53_row8_col2, #T_97c53_row8_col3, #T_97c53_row8_col4, #T_97c53_row8_col5, #T_97c53_row8_col6, #T_97c53_row8_col7, #T_97c53_row8_col8, #T_97c53_row8_col9, #T_97c53_row8_col10, #T_97c53_row8_col11, #T_97c53_row9_col0, #T_97c53_row9_col3, #T_97c53_row9_col7, #T_97c53_row9_col11, #T_97c53_row10_col0, #T_97c53_row10_col3, #T_97c53_row10_col4, #T_97c53_row10_col6, #T_97c53_row10_col7, #T_97c53_row10_col8, #T_97c53_row10_col9, #T_97c53_row10_col11, #T_97c53_row11_col0, #T_97c53_row11_col3, #T_97c53_row11_col5, #T_97c53_row11_col7, #T_97c53_row11_col8, #T_97c53_row11_col9, #T_97c53_row11_col10, #T_97c53_row12_col3, #T_97c53_row12_col4, #T_97c53_row12_col5, #T_97c53_row12_col7, #T_97c53_row12_col8, #T_97c53_row12_col9, #T_97c53_row12_col10, #T_97c53_row12_col11, #T_97c53_row13_col3, #T_97c53_row13_col4, #T_97c53_row13_col6, #T_97c53_row13_col7, #T_97c53_row13_col8, #T_97c53_row14_col3, #T_97c53_row14_col4, #T_97c53_row14_col8, #T_97c53_row14_col9, #T_97c53_row14_col11, #T_97c53_row15_col4, #T_97c53_row15_col8, #T_97c53_row15_col10, #T_97c53_row16_col7, #T_97c53_row17_col8, #T_97c53_row18_col10, #T_97c53_row19_col10, #T_97c53_row20_col10, #T_97c53_row21_col10, #T_97c53_row22_col11 {\n",
       "  background-color: green;\n",
       "}\n",
       "#T_97c53_row0_col8, #T_97c53_row0_col9, #T_97c53_row4_col2, #T_97c53_row12_col2, #T_97c53_row13_col2, #T_97c53_row14_col2 {\n",
       "  background-color: red;\n",
       "}\n",
       "#T_97c53_row4_col4, #T_97c53_row4_col10, #T_97c53_row9_col1, #T_97c53_row9_col2, #T_97c53_row9_col4, #T_97c53_row9_col5, #T_97c53_row9_col6, #T_97c53_row9_col8, #T_97c53_row9_col9, #T_97c53_row9_col10, #T_97c53_row10_col1, #T_97c53_row10_col2, #T_97c53_row10_col5, #T_97c53_row10_col10, #T_97c53_row11_col1, #T_97c53_row11_col2, #T_97c53_row11_col4, #T_97c53_row11_col6, #T_97c53_row11_col11, #T_97c53_row12_col0, #T_97c53_row12_col1, #T_97c53_row12_col6, #T_97c53_row13_col0, #T_97c53_row13_col1, #T_97c53_row13_col5, #T_97c53_row13_col9, #T_97c53_row13_col10, #T_97c53_row13_col11, #T_97c53_row14_col0, #T_97c53_row14_col1, #T_97c53_row14_col5, #T_97c53_row14_col6, #T_97c53_row14_col7, #T_97c53_row14_col10, #T_97c53_row15_col0, #T_97c53_row15_col1, #T_97c53_row15_col2, #T_97c53_row15_col3, #T_97c53_row15_col5, #T_97c53_row15_col6, #T_97c53_row15_col7, #T_97c53_row15_col9, #T_97c53_row15_col11, #T_97c53_row16_col0, #T_97c53_row16_col1, #T_97c53_row16_col2, #T_97c53_row16_col3, #T_97c53_row16_col4, #T_97c53_row16_col5, #T_97c53_row16_col6, #T_97c53_row16_col8, #T_97c53_row16_col9, #T_97c53_row16_col10, #T_97c53_row16_col11, #T_97c53_row17_col0, #T_97c53_row17_col1, #T_97c53_row17_col2, #T_97c53_row17_col3, #T_97c53_row17_col4, #T_97c53_row17_col5, #T_97c53_row17_col6, #T_97c53_row17_col7, #T_97c53_row17_col9, #T_97c53_row17_col10, #T_97c53_row17_col11, #T_97c53_row18_col0, #T_97c53_row18_col1, #T_97c53_row18_col2, #T_97c53_row18_col3, #T_97c53_row18_col4, #T_97c53_row18_col5, #T_97c53_row18_col6, #T_97c53_row18_col7, #T_97c53_row18_col8, #T_97c53_row18_col9, #T_97c53_row18_col11, #T_97c53_row19_col0, #T_97c53_row19_col1, #T_97c53_row19_col2, #T_97c53_row19_col3, #T_97c53_row19_col4, #T_97c53_row19_col5, #T_97c53_row19_col6, #T_97c53_row19_col7, #T_97c53_row19_col8, #T_97c53_row19_col9, #T_97c53_row19_col11, #T_97c53_row20_col0, #T_97c53_row20_col1, #T_97c53_row20_col2, #T_97c53_row20_col3, #T_97c53_row20_col4, #T_97c53_row20_col5, #T_97c53_row20_col6, #T_97c53_row20_col7, #T_97c53_row20_col8, #T_97c53_row20_col9, #T_97c53_row20_col11, #T_97c53_row21_col0, #T_97c53_row21_col1, #T_97c53_row21_col2, #T_97c53_row21_col3, #T_97c53_row21_col4, #T_97c53_row21_col5, #T_97c53_row21_col6, #T_97c53_row21_col7, #T_97c53_row21_col8, #T_97c53_row21_col9, #T_97c53_row21_col11, #T_97c53_row22_col0, #T_97c53_row22_col1, #T_97c53_row22_col2, #T_97c53_row22_col3, #T_97c53_row22_col4, #T_97c53_row22_col5, #T_97c53_row22_col6, #T_97c53_row22_col7, #T_97c53_row22_col8, #T_97c53_row22_col9, #T_97c53_row22_col10 {\n",
       "  background: white;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_97c53_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >C51</th>\n",
       "      <th class=\"col_heading level0 col1\" >C52</th>\n",
       "      <th class=\"col_heading level0 col2\" >C141</th>\n",
       "      <th class=\"col_heading level0 col3\" >C142</th>\n",
       "      <th class=\"col_heading level0 col4\" >C148</th>\n",
       "      <th class=\"col_heading level0 col5\" >C149</th>\n",
       "      <th class=\"col_heading level0 col6\" >C100</th>\n",
       "      <th class=\"col_heading level0 col7\" >C144</th>\n",
       "      <th class=\"col_heading level0 col8\" >C143</th>\n",
       "      <th class=\"col_heading level0 col9\" >C145</th>\n",
       "      <th class=\"col_heading level0 col10\" >C146</th>\n",
       "      <th class=\"col_heading level0 col11\" >C152</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row0\" class=\"row_heading level0 row0\" >raw_data</th>\n",
       "      <td id=\"T_97c53_row0_col0\" class=\"data row0 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row0_col1\" class=\"data row0 col1\" >+</td>\n",
       "      <td id=\"T_97c53_row0_col2\" class=\"data row0 col2\" >+</td>\n",
       "      <td id=\"T_97c53_row0_col3\" class=\"data row0 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row0_col4\" class=\"data row0 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row0_col5\" class=\"data row0 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row0_col6\" class=\"data row0 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row0_col7\" class=\"data row0 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row0_col8\" class=\"data row0 col8\" >-</td>\n",
       "      <td id=\"T_97c53_row0_col9\" class=\"data row0 col9\" >-</td>\n",
       "      <td id=\"T_97c53_row0_col10\" class=\"data row0 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row0_col11\" class=\"data row0 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row1\" class=\"row_heading level0 row1\" >raw_data_Macrophage</th>\n",
       "      <td id=\"T_97c53_row1_col0\" class=\"data row1 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col1\" class=\"data row1 col1\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col2\" class=\"data row1 col2\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col3\" class=\"data row1 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col4\" class=\"data row1 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col5\" class=\"data row1 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col6\" class=\"data row1 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col7\" class=\"data row1 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col8\" class=\"data row1 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col9\" class=\"data row1 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col10\" class=\"data row1 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row1_col11\" class=\"data row1 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row2\" class=\"row_heading level0 row2\" >raw_data_T_cells</th>\n",
       "      <td id=\"T_97c53_row2_col0\" class=\"data row2 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col1\" class=\"data row2 col1\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col2\" class=\"data row2 col2\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col3\" class=\"data row2 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col4\" class=\"data row2 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col5\" class=\"data row2 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col6\" class=\"data row2 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col7\" class=\"data row2 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col8\" class=\"data row2 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col9\" class=\"data row2 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col10\" class=\"data row2 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row2_col11\" class=\"data row2 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row3\" class=\"row_heading level0 row3\" >raw_data_DC</th>\n",
       "      <td id=\"T_97c53_row3_col0\" class=\"data row3 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col1\" class=\"data row3 col1\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col2\" class=\"data row3 col2\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col3\" class=\"data row3 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col4\" class=\"data row3 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col5\" class=\"data row3 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col6\" class=\"data row3 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col7\" class=\"data row3 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col8\" class=\"data row3 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col9\" class=\"data row3 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col10\" class=\"data row3 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row3_col11\" class=\"data row3 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row4\" class=\"row_heading level0 row4\" >raw_data_Pre-B_cell_CD34-</th>\n",
       "      <td id=\"T_97c53_row4_col0\" class=\"data row4 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row4_col1\" class=\"data row4 col1\" >+</td>\n",
       "      <td id=\"T_97c53_row4_col2\" class=\"data row4 col2\" >-</td>\n",
       "      <td id=\"T_97c53_row4_col3\" class=\"data row4 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row4_col4\" class=\"data row4 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row4_col5\" class=\"data row4 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row4_col6\" class=\"data row4 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row4_col7\" class=\"data row4 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row4_col8\" class=\"data row4 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row4_col9\" class=\"data row4 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row4_col10\" class=\"data row4 col10\" >nan</td>\n",
       "      <td id=\"T_97c53_row4_col11\" class=\"data row4 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row5\" class=\"row_heading level0 row5\" >raw_data_Monocyte</th>\n",
       "      <td id=\"T_97c53_row5_col0\" class=\"data row5 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col1\" class=\"data row5 col1\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col2\" class=\"data row5 col2\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col3\" class=\"data row5 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col4\" class=\"data row5 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col5\" class=\"data row5 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col6\" class=\"data row5 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col7\" class=\"data row5 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col8\" class=\"data row5 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col9\" class=\"data row5 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col10\" class=\"data row5 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row5_col11\" class=\"data row5 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row6\" class=\"row_heading level0 row6\" >raw_data_NK_cell</th>\n",
       "      <td id=\"T_97c53_row6_col0\" class=\"data row6 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col1\" class=\"data row6 col1\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col2\" class=\"data row6 col2\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col3\" class=\"data row6 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col4\" class=\"data row6 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col5\" class=\"data row6 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col6\" class=\"data row6 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col7\" class=\"data row6 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col8\" class=\"data row6 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col9\" class=\"data row6 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col10\" class=\"data row6 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row6_col11\" class=\"data row6 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row7\" class=\"row_heading level0 row7\" >raw_data_B_cell</th>\n",
       "      <td id=\"T_97c53_row7_col0\" class=\"data row7 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col1\" class=\"data row7 col1\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col2\" class=\"data row7 col2\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col3\" class=\"data row7 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col4\" class=\"data row7 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col5\" class=\"data row7 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col6\" class=\"data row7 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col7\" class=\"data row7 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col8\" class=\"data row7 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col9\" class=\"data row7 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col10\" class=\"data row7 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row7_col11\" class=\"data row7 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row8\" class=\"row_heading level0 row8\" >raw_data_Epithelial_cells</th>\n",
       "      <td id=\"T_97c53_row8_col0\" class=\"data row8 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col1\" class=\"data row8 col1\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col2\" class=\"data row8 col2\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col3\" class=\"data row8 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col4\" class=\"data row8 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col5\" class=\"data row8 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col6\" class=\"data row8 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col7\" class=\"data row8 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col8\" class=\"data row8 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col9\" class=\"data row8 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col10\" class=\"data row8 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row8_col11\" class=\"data row8 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row9\" class=\"row_heading level0 row9\" >raw_data_BM</th>\n",
       "      <td id=\"T_97c53_row9_col0\" class=\"data row9 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row9_col1\" class=\"data row9 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row9_col2\" class=\"data row9 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row9_col3\" class=\"data row9 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row9_col4\" class=\"data row9 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row9_col5\" class=\"data row9 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row9_col6\" class=\"data row9 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row9_col7\" class=\"data row9 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row9_col8\" class=\"data row9 col8\" >nan</td>\n",
       "      <td id=\"T_97c53_row9_col9\" class=\"data row9 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row9_col10\" class=\"data row9 col10\" >nan</td>\n",
       "      <td id=\"T_97c53_row9_col11\" class=\"data row9 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row10\" class=\"row_heading level0 row10\" >raw_data_Pro-B_cell_CD34+</th>\n",
       "      <td id=\"T_97c53_row10_col0\" class=\"data row10 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row10_col1\" class=\"data row10 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row10_col2\" class=\"data row10 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row10_col3\" class=\"data row10 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row10_col4\" class=\"data row10 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row10_col5\" class=\"data row10 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row10_col6\" class=\"data row10 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row10_col7\" class=\"data row10 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row10_col8\" class=\"data row10 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row10_col9\" class=\"data row10 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row10_col10\" class=\"data row10 col10\" >nan</td>\n",
       "      <td id=\"T_97c53_row10_col11\" class=\"data row10 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row11\" class=\"row_heading level0 row11\" >raw_data_HSC_-G-CSF</th>\n",
       "      <td id=\"T_97c53_row11_col0\" class=\"data row11 col0\" >+</td>\n",
       "      <td id=\"T_97c53_row11_col1\" class=\"data row11 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row11_col2\" class=\"data row11 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row11_col3\" class=\"data row11 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row11_col4\" class=\"data row11 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row11_col5\" class=\"data row11 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row11_col6\" class=\"data row11 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row11_col7\" class=\"data row11 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row11_col8\" class=\"data row11 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row11_col9\" class=\"data row11 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row11_col10\" class=\"data row11 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row11_col11\" class=\"data row11 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row12\" class=\"row_heading level0 row12\" >raw_data_Neutrophils</th>\n",
       "      <td id=\"T_97c53_row12_col0\" class=\"data row12 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row12_col1\" class=\"data row12 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row12_col2\" class=\"data row12 col2\" >-</td>\n",
       "      <td id=\"T_97c53_row12_col3\" class=\"data row12 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row12_col4\" class=\"data row12 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row12_col5\" class=\"data row12 col5\" >+</td>\n",
       "      <td id=\"T_97c53_row12_col6\" class=\"data row12 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row12_col7\" class=\"data row12 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row12_col8\" class=\"data row12 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row12_col9\" class=\"data row12 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row12_col10\" class=\"data row12 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row12_col11\" class=\"data row12 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row13\" class=\"row_heading level0 row13\" >raw_data_CMP</th>\n",
       "      <td id=\"T_97c53_row13_col0\" class=\"data row13 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row13_col1\" class=\"data row13 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row13_col2\" class=\"data row13 col2\" >-</td>\n",
       "      <td id=\"T_97c53_row13_col3\" class=\"data row13 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row13_col4\" class=\"data row13 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row13_col5\" class=\"data row13 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row13_col6\" class=\"data row13 col6\" >+</td>\n",
       "      <td id=\"T_97c53_row13_col7\" class=\"data row13 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row13_col8\" class=\"data row13 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row13_col9\" class=\"data row13 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row13_col10\" class=\"data row13 col10\" >nan</td>\n",
       "      <td id=\"T_97c53_row13_col11\" class=\"data row13 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row14\" class=\"row_heading level0 row14\" >raw_data_GMP</th>\n",
       "      <td id=\"T_97c53_row14_col0\" class=\"data row14 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row14_col1\" class=\"data row14 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row14_col2\" class=\"data row14 col2\" >-</td>\n",
       "      <td id=\"T_97c53_row14_col3\" class=\"data row14 col3\" >+</td>\n",
       "      <td id=\"T_97c53_row14_col4\" class=\"data row14 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row14_col5\" class=\"data row14 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row14_col6\" class=\"data row14 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row14_col7\" class=\"data row14 col7\" >nan</td>\n",
       "      <td id=\"T_97c53_row14_col8\" class=\"data row14 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row14_col9\" class=\"data row14 col9\" >+</td>\n",
       "      <td id=\"T_97c53_row14_col10\" class=\"data row14 col10\" >nan</td>\n",
       "      <td id=\"T_97c53_row14_col11\" class=\"data row14 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row15\" class=\"row_heading level0 row15\" >raw_data_Gametocytes</th>\n",
       "      <td id=\"T_97c53_row15_col0\" class=\"data row15 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row15_col1\" class=\"data row15 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row15_col2\" class=\"data row15 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row15_col3\" class=\"data row15 col3\" >nan</td>\n",
       "      <td id=\"T_97c53_row15_col4\" class=\"data row15 col4\" >+</td>\n",
       "      <td id=\"T_97c53_row15_col5\" class=\"data row15 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row15_col6\" class=\"data row15 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row15_col7\" class=\"data row15 col7\" >nan</td>\n",
       "      <td id=\"T_97c53_row15_col8\" class=\"data row15 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row15_col9\" class=\"data row15 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row15_col10\" class=\"data row15 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row15_col11\" class=\"data row15 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row16\" class=\"row_heading level0 row16\" >raw_data_Erythroblast</th>\n",
       "      <td id=\"T_97c53_row16_col0\" class=\"data row16 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col1\" class=\"data row16 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col2\" class=\"data row16 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col3\" class=\"data row16 col3\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col4\" class=\"data row16 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col5\" class=\"data row16 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col6\" class=\"data row16 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col7\" class=\"data row16 col7\" >+</td>\n",
       "      <td id=\"T_97c53_row16_col8\" class=\"data row16 col8\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col9\" class=\"data row16 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col10\" class=\"data row16 col10\" >nan</td>\n",
       "      <td id=\"T_97c53_row16_col11\" class=\"data row16 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row17\" class=\"row_heading level0 row17\" >raw_data_Neurons</th>\n",
       "      <td id=\"T_97c53_row17_col0\" class=\"data row17 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col1\" class=\"data row17 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col2\" class=\"data row17 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col3\" class=\"data row17 col3\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col4\" class=\"data row17 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col5\" class=\"data row17 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col6\" class=\"data row17 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col7\" class=\"data row17 col7\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col8\" class=\"data row17 col8\" >+</td>\n",
       "      <td id=\"T_97c53_row17_col9\" class=\"data row17 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col10\" class=\"data row17 col10\" >nan</td>\n",
       "      <td id=\"T_97c53_row17_col11\" class=\"data row17 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row18\" class=\"row_heading level0 row18\" >raw_data_Fibroblasts</th>\n",
       "      <td id=\"T_97c53_row18_col0\" class=\"data row18 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col1\" class=\"data row18 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col2\" class=\"data row18 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col3\" class=\"data row18 col3\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col4\" class=\"data row18 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col5\" class=\"data row18 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col6\" class=\"data row18 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col7\" class=\"data row18 col7\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col8\" class=\"data row18 col8\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col9\" class=\"data row18 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row18_col10\" class=\"data row18 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row18_col11\" class=\"data row18 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row19\" class=\"row_heading level0 row19\" >raw_data_Smooth_muscle_cells</th>\n",
       "      <td id=\"T_97c53_row19_col0\" class=\"data row19 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col1\" class=\"data row19 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col2\" class=\"data row19 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col3\" class=\"data row19 col3\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col4\" class=\"data row19 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col5\" class=\"data row19 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col6\" class=\"data row19 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col7\" class=\"data row19 col7\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col8\" class=\"data row19 col8\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col9\" class=\"data row19 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row19_col10\" class=\"data row19 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row19_col11\" class=\"data row19 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row20\" class=\"row_heading level0 row20\" >raw_data_Hepatocytes</th>\n",
       "      <td id=\"T_97c53_row20_col0\" class=\"data row20 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col1\" class=\"data row20 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col2\" class=\"data row20 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col3\" class=\"data row20 col3\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col4\" class=\"data row20 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col5\" class=\"data row20 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col6\" class=\"data row20 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col7\" class=\"data row20 col7\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col8\" class=\"data row20 col8\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col9\" class=\"data row20 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row20_col10\" class=\"data row20 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row20_col11\" class=\"data row20 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row21\" class=\"row_heading level0 row21\" >raw_data_Keratinocytes</th>\n",
       "      <td id=\"T_97c53_row21_col0\" class=\"data row21 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col1\" class=\"data row21 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col2\" class=\"data row21 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col3\" class=\"data row21 col3\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col4\" class=\"data row21 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col5\" class=\"data row21 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col6\" class=\"data row21 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col7\" class=\"data row21 col7\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col8\" class=\"data row21 col8\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col9\" class=\"data row21 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row21_col10\" class=\"data row21 col10\" >+</td>\n",
       "      <td id=\"T_97c53_row21_col11\" class=\"data row21 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_97c53_level0_row22\" class=\"row_heading level0 row22\" >raw_data_Pro-Myelocyte</th>\n",
       "      <td id=\"T_97c53_row22_col0\" class=\"data row22 col0\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col1\" class=\"data row22 col1\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col2\" class=\"data row22 col2\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col3\" class=\"data row22 col3\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col4\" class=\"data row22 col4\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col5\" class=\"data row22 col5\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col6\" class=\"data row22 col6\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col7\" class=\"data row22 col7\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col8\" class=\"data row22 col8\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col9\" class=\"data row22 col9\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col10\" class=\"data row22 col10\" >nan</td>\n",
       "      <td id=\"T_97c53_row22_col11\" class=\"data row22 col11\" >+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7fe3cb723050>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36mTF\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_d3fd8_row0_col0, #T_d3fd8_row0_col1, #T_d3fd8_row0_col2, #T_d3fd8_row0_col3, #T_d3fd8_row0_col4, #T_d3fd8_row0_col5, #T_d3fd8_row0_col6, #T_d3fd8_row0_col7, #T_d3fd8_row0_col8, #T_d3fd8_row0_col9, #T_d3fd8_row0_col10, #T_d3fd8_row0_col11, #T_d3fd8_row1_col0, #T_d3fd8_row1_col1, #T_d3fd8_row1_col2, #T_d3fd8_row1_col3, #T_d3fd8_row1_col4, #T_d3fd8_row1_col5, #T_d3fd8_row1_col6, #T_d3fd8_row1_col7, #T_d3fd8_row1_col8, #T_d3fd8_row1_col9, #T_d3fd8_row1_col10, #T_d3fd8_row1_col11, #T_d3fd8_row2_col0, #T_d3fd8_row2_col1, #T_d3fd8_row2_col2, #T_d3fd8_row2_col3, #T_d3fd8_row2_col4, #T_d3fd8_row2_col5, #T_d3fd8_row2_col6, #T_d3fd8_row2_col7, #T_d3fd8_row2_col8, #T_d3fd8_row2_col9, #T_d3fd8_row2_col10, #T_d3fd8_row2_col11, #T_d3fd8_row3_col0, #T_d3fd8_row3_col1, #T_d3fd8_row3_col2, #T_d3fd8_row3_col3, #T_d3fd8_row3_col4, #T_d3fd8_row3_col5, #T_d3fd8_row3_col6, #T_d3fd8_row3_col7, #T_d3fd8_row3_col8, #T_d3fd8_row3_col9, #T_d3fd8_row3_col10, #T_d3fd8_row3_col11, #T_d3fd8_row4_col0, #T_d3fd8_row4_col1, #T_d3fd8_row4_col3, #T_d3fd8_row4_col5, #T_d3fd8_row4_col6, #T_d3fd8_row4_col7, #T_d3fd8_row4_col8, #T_d3fd8_row4_col9, #T_d3fd8_row4_col11, #T_d3fd8_row5_col0, #T_d3fd8_row5_col1, #T_d3fd8_row5_col2, #T_d3fd8_row5_col3, #T_d3fd8_row5_col4, #T_d3fd8_row5_col5, #T_d3fd8_row5_col6, #T_d3fd8_row5_col7, #T_d3fd8_row5_col8, #T_d3fd8_row5_col9, #T_d3fd8_row5_col10, #T_d3fd8_row5_col11, #T_d3fd8_row6_col0, #T_d3fd8_row6_col1, #T_d3fd8_row6_col2, #T_d3fd8_row6_col3, #T_d3fd8_row6_col4, #T_d3fd8_row6_col5, #T_d3fd8_row6_col6, #T_d3fd8_row6_col7, #T_d3fd8_row6_col8, #T_d3fd8_row6_col9, #T_d3fd8_row6_col10, #T_d3fd8_row6_col11, #T_d3fd8_row7_col0, #T_d3fd8_row7_col1, #T_d3fd8_row7_col2, #T_d3fd8_row7_col3, #T_d3fd8_row7_col4, #T_d3fd8_row7_col5, #T_d3fd8_row7_col6, #T_d3fd8_row7_col7, #T_d3fd8_row7_col8, #T_d3fd8_row7_col9, #T_d3fd8_row7_col10, #T_d3fd8_row7_col11, #T_d3fd8_row8_col0, #T_d3fd8_row8_col1, #T_d3fd8_row8_col2, #T_d3fd8_row8_col3, #T_d3fd8_row8_col4, #T_d3fd8_row8_col5, #T_d3fd8_row8_col6, #T_d3fd8_row8_col7, #T_d3fd8_row8_col8, #T_d3fd8_row8_col9, #T_d3fd8_row8_col10, #T_d3fd8_row8_col11, #T_d3fd8_row9_col0, #T_d3fd8_row9_col3, #T_d3fd8_row9_col7, #T_d3fd8_row9_col11, #T_d3fd8_row10_col0, #T_d3fd8_row10_col3, #T_d3fd8_row10_col4, #T_d3fd8_row10_col6, #T_d3fd8_row10_col7, #T_d3fd8_row10_col8, #T_d3fd8_row10_col9, #T_d3fd8_row10_col11, #T_d3fd8_row11_col0, #T_d3fd8_row11_col3, #T_d3fd8_row11_col5, #T_d3fd8_row11_col7, #T_d3fd8_row11_col8, #T_d3fd8_row11_col9, #T_d3fd8_row11_col10, #T_d3fd8_row12_col3, #T_d3fd8_row12_col4, #T_d3fd8_row12_col5, #T_d3fd8_row12_col7, #T_d3fd8_row12_col8, #T_d3fd8_row12_col9, #T_d3fd8_row12_col10, #T_d3fd8_row12_col11, #T_d3fd8_row13_col3, #T_d3fd8_row13_col4, #T_d3fd8_row13_col6, #T_d3fd8_row13_col7, #T_d3fd8_row13_col8, #T_d3fd8_row14_col3, #T_d3fd8_row14_col4, #T_d3fd8_row14_col8, #T_d3fd8_row14_col9, #T_d3fd8_row14_col11, #T_d3fd8_row15_col4, #T_d3fd8_row15_col8, #T_d3fd8_row15_col10, #T_d3fd8_row16_col7, #T_d3fd8_row17_col8, #T_d3fd8_row18_col10, #T_d3fd8_row19_col10, #T_d3fd8_row20_col10, #T_d3fd8_row21_col10, #T_d3fd8_row22_col11 {\n",
       "  background-color: green;\n",
       "}\n",
       "#T_d3fd8_row4_col2, #T_d3fd8_row12_col2, #T_d3fd8_row13_col2, #T_d3fd8_row14_col2 {\n",
       "  background-color: red;\n",
       "}\n",
       "#T_d3fd8_row4_col4, #T_d3fd8_row4_col10, #T_d3fd8_row9_col1, #T_d3fd8_row9_col2, #T_d3fd8_row9_col4, #T_d3fd8_row9_col5, #T_d3fd8_row9_col6, #T_d3fd8_row9_col8, #T_d3fd8_row9_col9, #T_d3fd8_row9_col10, #T_d3fd8_row10_col1, #T_d3fd8_row10_col2, #T_d3fd8_row10_col5, #T_d3fd8_row10_col10, #T_d3fd8_row11_col1, #T_d3fd8_row11_col2, #T_d3fd8_row11_col4, #T_d3fd8_row11_col6, #T_d3fd8_row11_col11, #T_d3fd8_row12_col0, #T_d3fd8_row12_col1, #T_d3fd8_row12_col6, #T_d3fd8_row13_col0, #T_d3fd8_row13_col1, #T_d3fd8_row13_col5, #T_d3fd8_row13_col9, #T_d3fd8_row13_col10, #T_d3fd8_row13_col11, #T_d3fd8_row14_col0, #T_d3fd8_row14_col1, #T_d3fd8_row14_col5, #T_d3fd8_row14_col6, #T_d3fd8_row14_col7, #T_d3fd8_row14_col10, #T_d3fd8_row15_col0, #T_d3fd8_row15_col1, #T_d3fd8_row15_col2, #T_d3fd8_row15_col3, #T_d3fd8_row15_col5, #T_d3fd8_row15_col6, #T_d3fd8_row15_col7, #T_d3fd8_row15_col9, #T_d3fd8_row15_col11, #T_d3fd8_row16_col0, #T_d3fd8_row16_col1, #T_d3fd8_row16_col2, #T_d3fd8_row16_col3, #T_d3fd8_row16_col4, #T_d3fd8_row16_col5, #T_d3fd8_row16_col6, #T_d3fd8_row16_col8, #T_d3fd8_row16_col9, #T_d3fd8_row16_col10, #T_d3fd8_row16_col11, #T_d3fd8_row17_col0, #T_d3fd8_row17_col1, #T_d3fd8_row17_col2, #T_d3fd8_row17_col3, #T_d3fd8_row17_col4, #T_d3fd8_row17_col5, #T_d3fd8_row17_col6, #T_d3fd8_row17_col7, #T_d3fd8_row17_col9, #T_d3fd8_row17_col10, #T_d3fd8_row17_col11, #T_d3fd8_row18_col0, #T_d3fd8_row18_col1, #T_d3fd8_row18_col2, #T_d3fd8_row18_col3, #T_d3fd8_row18_col4, #T_d3fd8_row18_col5, #T_d3fd8_row18_col6, #T_d3fd8_row18_col7, #T_d3fd8_row18_col8, #T_d3fd8_row18_col9, #T_d3fd8_row18_col11, #T_d3fd8_row19_col0, #T_d3fd8_row19_col1, #T_d3fd8_row19_col2, #T_d3fd8_row19_col3, #T_d3fd8_row19_col4, #T_d3fd8_row19_col5, #T_d3fd8_row19_col6, #T_d3fd8_row19_col7, #T_d3fd8_row19_col8, #T_d3fd8_row19_col9, #T_d3fd8_row19_col11, #T_d3fd8_row20_col0, #T_d3fd8_row20_col1, #T_d3fd8_row20_col2, #T_d3fd8_row20_col3, #T_d3fd8_row20_col4, #T_d3fd8_row20_col5, #T_d3fd8_row20_col6, #T_d3fd8_row20_col7, #T_d3fd8_row20_col8, #T_d3fd8_row20_col9, #T_d3fd8_row20_col11, #T_d3fd8_row21_col0, #T_d3fd8_row21_col1, #T_d3fd8_row21_col2, #T_d3fd8_row21_col3, #T_d3fd8_row21_col4, #T_d3fd8_row21_col5, #T_d3fd8_row21_col6, #T_d3fd8_row21_col7, #T_d3fd8_row21_col8, #T_d3fd8_row21_col9, #T_d3fd8_row21_col11, #T_d3fd8_row22_col0, #T_d3fd8_row22_col1, #T_d3fd8_row22_col2, #T_d3fd8_row22_col3, #T_d3fd8_row22_col4, #T_d3fd8_row22_col5, #T_d3fd8_row22_col6, #T_d3fd8_row22_col7, #T_d3fd8_row22_col8, #T_d3fd8_row22_col9, #T_d3fd8_row22_col10 {\n",
       "  background: white;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_d3fd8_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >C51</th>\n",
       "      <th class=\"col_heading level0 col1\" >C52</th>\n",
       "      <th class=\"col_heading level0 col2\" >C141</th>\n",
       "      <th class=\"col_heading level0 col3\" >C142</th>\n",
       "      <th class=\"col_heading level0 col4\" >C148</th>\n",
       "      <th class=\"col_heading level0 col5\" >C149</th>\n",
       "      <th class=\"col_heading level0 col6\" >C100</th>\n",
       "      <th class=\"col_heading level0 col7\" >C144</th>\n",
       "      <th class=\"col_heading level0 col8\" >C143</th>\n",
       "      <th class=\"col_heading level0 col9\" >C145</th>\n",
       "      <th class=\"col_heading level0 col10\" >C146</th>\n",
       "      <th class=\"col_heading level0 col11\" >C152</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row0\" class=\"row_heading level0 row0\" >raw_data</th>\n",
       "      <td id=\"T_d3fd8_row0_col0\" class=\"data row0 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col1\" class=\"data row0 col1\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col2\" class=\"data row0 col2\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col3\" class=\"data row0 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col4\" class=\"data row0 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col5\" class=\"data row0 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col6\" class=\"data row0 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col7\" class=\"data row0 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col8\" class=\"data row0 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col9\" class=\"data row0 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col10\" class=\"data row0 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row0_col11\" class=\"data row0 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row1\" class=\"row_heading level0 row1\" >raw_data_Macrophage</th>\n",
       "      <td id=\"T_d3fd8_row1_col0\" class=\"data row1 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col1\" class=\"data row1 col1\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col2\" class=\"data row1 col2\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col3\" class=\"data row1 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col4\" class=\"data row1 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col5\" class=\"data row1 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col6\" class=\"data row1 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col7\" class=\"data row1 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col8\" class=\"data row1 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col9\" class=\"data row1 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col10\" class=\"data row1 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row1_col11\" class=\"data row1 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row2\" class=\"row_heading level0 row2\" >raw_data_T_cells</th>\n",
       "      <td id=\"T_d3fd8_row2_col0\" class=\"data row2 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col1\" class=\"data row2 col1\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col2\" class=\"data row2 col2\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col3\" class=\"data row2 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col4\" class=\"data row2 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col5\" class=\"data row2 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col6\" class=\"data row2 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col7\" class=\"data row2 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col8\" class=\"data row2 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col9\" class=\"data row2 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col10\" class=\"data row2 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row2_col11\" class=\"data row2 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row3\" class=\"row_heading level0 row3\" >raw_data_DC</th>\n",
       "      <td id=\"T_d3fd8_row3_col0\" class=\"data row3 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col1\" class=\"data row3 col1\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col2\" class=\"data row3 col2\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col3\" class=\"data row3 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col4\" class=\"data row3 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col5\" class=\"data row3 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col6\" class=\"data row3 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col7\" class=\"data row3 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col8\" class=\"data row3 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col9\" class=\"data row3 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col10\" class=\"data row3 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row3_col11\" class=\"data row3 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row4\" class=\"row_heading level0 row4\" >raw_data_Pre-B_cell_CD34-</th>\n",
       "      <td id=\"T_d3fd8_row4_col0\" class=\"data row4 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row4_col1\" class=\"data row4 col1\" >+</td>\n",
       "      <td id=\"T_d3fd8_row4_col2\" class=\"data row4 col2\" >-</td>\n",
       "      <td id=\"T_d3fd8_row4_col3\" class=\"data row4 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row4_col4\" class=\"data row4 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row4_col5\" class=\"data row4 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row4_col6\" class=\"data row4 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row4_col7\" class=\"data row4 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row4_col8\" class=\"data row4 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row4_col9\" class=\"data row4 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row4_col10\" class=\"data row4 col10\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row4_col11\" class=\"data row4 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row5\" class=\"row_heading level0 row5\" >raw_data_Monocyte</th>\n",
       "      <td id=\"T_d3fd8_row5_col0\" class=\"data row5 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col1\" class=\"data row5 col1\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col2\" class=\"data row5 col2\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col3\" class=\"data row5 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col4\" class=\"data row5 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col5\" class=\"data row5 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col6\" class=\"data row5 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col7\" class=\"data row5 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col8\" class=\"data row5 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col9\" class=\"data row5 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col10\" class=\"data row5 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row5_col11\" class=\"data row5 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row6\" class=\"row_heading level0 row6\" >raw_data_NK_cell</th>\n",
       "      <td id=\"T_d3fd8_row6_col0\" class=\"data row6 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col1\" class=\"data row6 col1\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col2\" class=\"data row6 col2\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col3\" class=\"data row6 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col4\" class=\"data row6 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col5\" class=\"data row6 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col6\" class=\"data row6 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col7\" class=\"data row6 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col8\" class=\"data row6 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col9\" class=\"data row6 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col10\" class=\"data row6 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row6_col11\" class=\"data row6 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row7\" class=\"row_heading level0 row7\" >raw_data_B_cell</th>\n",
       "      <td id=\"T_d3fd8_row7_col0\" class=\"data row7 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col1\" class=\"data row7 col1\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col2\" class=\"data row7 col2\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col3\" class=\"data row7 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col4\" class=\"data row7 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col5\" class=\"data row7 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col6\" class=\"data row7 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col7\" class=\"data row7 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col8\" class=\"data row7 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col9\" class=\"data row7 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col10\" class=\"data row7 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row7_col11\" class=\"data row7 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row8\" class=\"row_heading level0 row8\" >raw_data_Epithelial_cells</th>\n",
       "      <td id=\"T_d3fd8_row8_col0\" class=\"data row8 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col1\" class=\"data row8 col1\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col2\" class=\"data row8 col2\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col3\" class=\"data row8 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col4\" class=\"data row8 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col5\" class=\"data row8 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col6\" class=\"data row8 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col7\" class=\"data row8 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col8\" class=\"data row8 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col9\" class=\"data row8 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col10\" class=\"data row8 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row8_col11\" class=\"data row8 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row9\" class=\"row_heading level0 row9\" >raw_data_BM</th>\n",
       "      <td id=\"T_d3fd8_row9_col0\" class=\"data row9 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row9_col1\" class=\"data row9 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row9_col2\" class=\"data row9 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row9_col3\" class=\"data row9 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row9_col4\" class=\"data row9 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row9_col5\" class=\"data row9 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row9_col6\" class=\"data row9 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row9_col7\" class=\"data row9 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row9_col8\" class=\"data row9 col8\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row9_col9\" class=\"data row9 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row9_col10\" class=\"data row9 col10\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row9_col11\" class=\"data row9 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row10\" class=\"row_heading level0 row10\" >raw_data_Pro-B_cell_CD34+</th>\n",
       "      <td id=\"T_d3fd8_row10_col0\" class=\"data row10 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row10_col1\" class=\"data row10 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row10_col2\" class=\"data row10 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row10_col3\" class=\"data row10 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row10_col4\" class=\"data row10 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row10_col5\" class=\"data row10 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row10_col6\" class=\"data row10 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row10_col7\" class=\"data row10 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row10_col8\" class=\"data row10 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row10_col9\" class=\"data row10 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row10_col10\" class=\"data row10 col10\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row10_col11\" class=\"data row10 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row11\" class=\"row_heading level0 row11\" >raw_data_HSC_-G-CSF</th>\n",
       "      <td id=\"T_d3fd8_row11_col0\" class=\"data row11 col0\" >+</td>\n",
       "      <td id=\"T_d3fd8_row11_col1\" class=\"data row11 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row11_col2\" class=\"data row11 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row11_col3\" class=\"data row11 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row11_col4\" class=\"data row11 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row11_col5\" class=\"data row11 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row11_col6\" class=\"data row11 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row11_col7\" class=\"data row11 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row11_col8\" class=\"data row11 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row11_col9\" class=\"data row11 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row11_col10\" class=\"data row11 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row11_col11\" class=\"data row11 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row12\" class=\"row_heading level0 row12\" >raw_data_Neutrophils</th>\n",
       "      <td id=\"T_d3fd8_row12_col0\" class=\"data row12 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row12_col1\" class=\"data row12 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row12_col2\" class=\"data row12 col2\" >-</td>\n",
       "      <td id=\"T_d3fd8_row12_col3\" class=\"data row12 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row12_col4\" class=\"data row12 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row12_col5\" class=\"data row12 col5\" >+</td>\n",
       "      <td id=\"T_d3fd8_row12_col6\" class=\"data row12 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row12_col7\" class=\"data row12 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row12_col8\" class=\"data row12 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row12_col9\" class=\"data row12 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row12_col10\" class=\"data row12 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row12_col11\" class=\"data row12 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row13\" class=\"row_heading level0 row13\" >raw_data_CMP</th>\n",
       "      <td id=\"T_d3fd8_row13_col0\" class=\"data row13 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row13_col1\" class=\"data row13 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row13_col2\" class=\"data row13 col2\" >-</td>\n",
       "      <td id=\"T_d3fd8_row13_col3\" class=\"data row13 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row13_col4\" class=\"data row13 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row13_col5\" class=\"data row13 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row13_col6\" class=\"data row13 col6\" >+</td>\n",
       "      <td id=\"T_d3fd8_row13_col7\" class=\"data row13 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row13_col8\" class=\"data row13 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row13_col9\" class=\"data row13 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row13_col10\" class=\"data row13 col10\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row13_col11\" class=\"data row13 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row14\" class=\"row_heading level0 row14\" >raw_data_GMP</th>\n",
       "      <td id=\"T_d3fd8_row14_col0\" class=\"data row14 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row14_col1\" class=\"data row14 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row14_col2\" class=\"data row14 col2\" >-</td>\n",
       "      <td id=\"T_d3fd8_row14_col3\" class=\"data row14 col3\" >+</td>\n",
       "      <td id=\"T_d3fd8_row14_col4\" class=\"data row14 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row14_col5\" class=\"data row14 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row14_col6\" class=\"data row14 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row14_col7\" class=\"data row14 col7\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row14_col8\" class=\"data row14 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row14_col9\" class=\"data row14 col9\" >+</td>\n",
       "      <td id=\"T_d3fd8_row14_col10\" class=\"data row14 col10\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row14_col11\" class=\"data row14 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row15\" class=\"row_heading level0 row15\" >raw_data_Gametocytes</th>\n",
       "      <td id=\"T_d3fd8_row15_col0\" class=\"data row15 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row15_col1\" class=\"data row15 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row15_col2\" class=\"data row15 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row15_col3\" class=\"data row15 col3\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row15_col4\" class=\"data row15 col4\" >+</td>\n",
       "      <td id=\"T_d3fd8_row15_col5\" class=\"data row15 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row15_col6\" class=\"data row15 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row15_col7\" class=\"data row15 col7\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row15_col8\" class=\"data row15 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row15_col9\" class=\"data row15 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row15_col10\" class=\"data row15 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row15_col11\" class=\"data row15 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row16\" class=\"row_heading level0 row16\" >raw_data_Erythroblast</th>\n",
       "      <td id=\"T_d3fd8_row16_col0\" class=\"data row16 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col1\" class=\"data row16 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col2\" class=\"data row16 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col3\" class=\"data row16 col3\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col4\" class=\"data row16 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col5\" class=\"data row16 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col6\" class=\"data row16 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col7\" class=\"data row16 col7\" >+</td>\n",
       "      <td id=\"T_d3fd8_row16_col8\" class=\"data row16 col8\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col9\" class=\"data row16 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col10\" class=\"data row16 col10\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row16_col11\" class=\"data row16 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row17\" class=\"row_heading level0 row17\" >raw_data_Neurons</th>\n",
       "      <td id=\"T_d3fd8_row17_col0\" class=\"data row17 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col1\" class=\"data row17 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col2\" class=\"data row17 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col3\" class=\"data row17 col3\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col4\" class=\"data row17 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col5\" class=\"data row17 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col6\" class=\"data row17 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col7\" class=\"data row17 col7\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col8\" class=\"data row17 col8\" >+</td>\n",
       "      <td id=\"T_d3fd8_row17_col9\" class=\"data row17 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col10\" class=\"data row17 col10\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row17_col11\" class=\"data row17 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row18\" class=\"row_heading level0 row18\" >raw_data_Fibroblasts</th>\n",
       "      <td id=\"T_d3fd8_row18_col0\" class=\"data row18 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col1\" class=\"data row18 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col2\" class=\"data row18 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col3\" class=\"data row18 col3\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col4\" class=\"data row18 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col5\" class=\"data row18 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col6\" class=\"data row18 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col7\" class=\"data row18 col7\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col8\" class=\"data row18 col8\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col9\" class=\"data row18 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row18_col10\" class=\"data row18 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row18_col11\" class=\"data row18 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row19\" class=\"row_heading level0 row19\" >raw_data_Smooth_muscle_cells</th>\n",
       "      <td id=\"T_d3fd8_row19_col0\" class=\"data row19 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col1\" class=\"data row19 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col2\" class=\"data row19 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col3\" class=\"data row19 col3\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col4\" class=\"data row19 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col5\" class=\"data row19 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col6\" class=\"data row19 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col7\" class=\"data row19 col7\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col8\" class=\"data row19 col8\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col9\" class=\"data row19 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row19_col10\" class=\"data row19 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row19_col11\" class=\"data row19 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row20\" class=\"row_heading level0 row20\" >raw_data_Hepatocytes</th>\n",
       "      <td id=\"T_d3fd8_row20_col0\" class=\"data row20 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col1\" class=\"data row20 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col2\" class=\"data row20 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col3\" class=\"data row20 col3\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col4\" class=\"data row20 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col5\" class=\"data row20 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col6\" class=\"data row20 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col7\" class=\"data row20 col7\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col8\" class=\"data row20 col8\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col9\" class=\"data row20 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row20_col10\" class=\"data row20 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row20_col11\" class=\"data row20 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row21\" class=\"row_heading level0 row21\" >raw_data_Keratinocytes</th>\n",
       "      <td id=\"T_d3fd8_row21_col0\" class=\"data row21 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col1\" class=\"data row21 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col2\" class=\"data row21 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col3\" class=\"data row21 col3\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col4\" class=\"data row21 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col5\" class=\"data row21 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col6\" class=\"data row21 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col7\" class=\"data row21 col7\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col8\" class=\"data row21 col8\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col9\" class=\"data row21 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row21_col10\" class=\"data row21 col10\" >+</td>\n",
       "      <td id=\"T_d3fd8_row21_col11\" class=\"data row21 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_d3fd8_level0_row22\" class=\"row_heading level0 row22\" >raw_data_Pro-Myelocyte</th>\n",
       "      <td id=\"T_d3fd8_row22_col0\" class=\"data row22 col0\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col1\" class=\"data row22 col1\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col2\" class=\"data row22 col2\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col3\" class=\"data row22 col3\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col4\" class=\"data row22 col4\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col5\" class=\"data row22 col5\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col6\" class=\"data row22 col6\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col7\" class=\"data row22 col7\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col8\" class=\"data row22 col8\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col9\" class=\"data row22 col9\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col10\" class=\"data row22 col10\" >nan</td>\n",
       "      <td id=\"T_d3fd8_row22_col11\" class=\"data row22 col11\" >+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7fe3cb6fcad0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36mCTX\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_6f768_row0_col0, #T_6f768_row0_col1, #T_6f768_row0_col2, #T_6f768_row0_col3, #T_6f768_row0_col4, #T_6f768_row0_col5, #T_6f768_row0_col6, #T_6f768_row0_col7, #T_6f768_row0_col8, #T_6f768_row0_col9, #T_6f768_row0_col10, #T_6f768_row0_col11, #T_6f768_row1_col0, #T_6f768_row1_col1, #T_6f768_row1_col2, #T_6f768_row1_col3, #T_6f768_row1_col4, #T_6f768_row1_col5, #T_6f768_row1_col6, #T_6f768_row1_col7, #T_6f768_row1_col8, #T_6f768_row1_col9, #T_6f768_row1_col10, #T_6f768_row1_col11, #T_6f768_row2_col0, #T_6f768_row2_col1, #T_6f768_row2_col2, #T_6f768_row2_col3, #T_6f768_row2_col4, #T_6f768_row2_col5, #T_6f768_row2_col6, #T_6f768_row2_col7, #T_6f768_row2_col8, #T_6f768_row2_col9, #T_6f768_row2_col10, #T_6f768_row2_col11, #T_6f768_row3_col0, #T_6f768_row3_col1, #T_6f768_row3_col2, #T_6f768_row3_col3, #T_6f768_row3_col4, #T_6f768_row3_col5, #T_6f768_row3_col6, #T_6f768_row3_col7, #T_6f768_row3_col8, #T_6f768_row3_col9, #T_6f768_row3_col10, #T_6f768_row3_col11, #T_6f768_row4_col0, #T_6f768_row4_col1, #T_6f768_row4_col3, #T_6f768_row4_col6, #T_6f768_row4_col7, #T_6f768_row4_col8, #T_6f768_row4_col9, #T_6f768_row4_col11, #T_6f768_row5_col0, #T_6f768_row5_col1, #T_6f768_row5_col2, #T_6f768_row5_col3, #T_6f768_row5_col4, #T_6f768_row5_col5, #T_6f768_row5_col6, #T_6f768_row5_col7, #T_6f768_row5_col8, #T_6f768_row5_col9, #T_6f768_row5_col10, #T_6f768_row5_col11, #T_6f768_row6_col0, #T_6f768_row6_col1, #T_6f768_row6_col2, #T_6f768_row6_col3, #T_6f768_row6_col4, #T_6f768_row6_col5, #T_6f768_row6_col6, #T_6f768_row6_col7, #T_6f768_row6_col8, #T_6f768_row6_col9, #T_6f768_row6_col10, #T_6f768_row6_col11, #T_6f768_row7_col0, #T_6f768_row7_col2, #T_6f768_row7_col3, #T_6f768_row7_col4, #T_6f768_row7_col5, #T_6f768_row7_col6, #T_6f768_row7_col7, #T_6f768_row7_col8, #T_6f768_row7_col9, #T_6f768_row7_col10, #T_6f768_row7_col11, #T_6f768_row8_col0, #T_6f768_row8_col1, #T_6f768_row8_col2, #T_6f768_row8_col3, #T_6f768_row8_col4, #T_6f768_row8_col5, #T_6f768_row8_col6, #T_6f768_row8_col7, #T_6f768_row8_col8, #T_6f768_row8_col9, #T_6f768_row8_col10, #T_6f768_row8_col11, #T_6f768_row9_col0, #T_6f768_row9_col11, #T_6f768_row10_col0, #T_6f768_row10_col11, #T_6f768_row11_col5, #T_6f768_row11_col8, #T_6f768_row11_col9, #T_6f768_row12_col3, #T_6f768_row12_col4, #T_6f768_row12_col5, #T_6f768_row12_col7, #T_6f768_row12_col8, #T_6f768_row12_col9, #T_6f768_row12_col10, #T_6f768_row12_col11, #T_6f768_row13_col3, #T_6f768_row13_col4, #T_6f768_row13_col6, #T_6f768_row14_col11 {\n",
       "  background-color: green;\n",
       "}\n",
       "#T_6f768_row4_col2, #T_6f768_row4_col5, #T_6f768_row7_col1, #T_6f768_row9_col3, #T_6f768_row9_col7, #T_6f768_row10_col3, #T_6f768_row10_col4, #T_6f768_row10_col6, #T_6f768_row10_col7, #T_6f768_row10_col8, #T_6f768_row10_col9, #T_6f768_row11_col0, #T_6f768_row11_col3, #T_6f768_row11_col7, #T_6f768_row11_col10, #T_6f768_row12_col2, #T_6f768_row13_col2, #T_6f768_row13_col7, #T_6f768_row13_col8, #T_6f768_row14_col2, #T_6f768_row14_col3, #T_6f768_row14_col4, #T_6f768_row14_col8, #T_6f768_row14_col9, #T_6f768_row15_col4, #T_6f768_row15_col8, #T_6f768_row15_col10, #T_6f768_row16_col7, #T_6f768_row17_col8, #T_6f768_row18_col10, #T_6f768_row19_col10, #T_6f768_row20_col10, #T_6f768_row21_col10, #T_6f768_row22_col11 {\n",
       "  background-color: red;\n",
       "}\n",
       "#T_6f768_row4_col4, #T_6f768_row4_col10, #T_6f768_row9_col1, #T_6f768_row9_col2, #T_6f768_row9_col4, #T_6f768_row9_col5, #T_6f768_row9_col6, #T_6f768_row9_col8, #T_6f768_row9_col9, #T_6f768_row9_col10, #T_6f768_row10_col1, #T_6f768_row10_col2, #T_6f768_row10_col5, #T_6f768_row10_col10, #T_6f768_row11_col1, #T_6f768_row11_col2, #T_6f768_row11_col4, #T_6f768_row11_col6, #T_6f768_row11_col11, #T_6f768_row12_col0, #T_6f768_row12_col1, #T_6f768_row12_col6, #T_6f768_row13_col0, #T_6f768_row13_col1, #T_6f768_row13_col5, #T_6f768_row13_col9, #T_6f768_row13_col10, #T_6f768_row13_col11, #T_6f768_row14_col0, #T_6f768_row14_col1, #T_6f768_row14_col5, #T_6f768_row14_col6, #T_6f768_row14_col7, #T_6f768_row14_col10, #T_6f768_row15_col0, #T_6f768_row15_col1, #T_6f768_row15_col2, #T_6f768_row15_col3, #T_6f768_row15_col5, #T_6f768_row15_col6, #T_6f768_row15_col7, #T_6f768_row15_col9, #T_6f768_row15_col11, #T_6f768_row16_col0, #T_6f768_row16_col1, #T_6f768_row16_col2, #T_6f768_row16_col3, #T_6f768_row16_col4, #T_6f768_row16_col5, #T_6f768_row16_col6, #T_6f768_row16_col8, #T_6f768_row16_col9, #T_6f768_row16_col10, #T_6f768_row16_col11, #T_6f768_row17_col0, #T_6f768_row17_col1, #T_6f768_row17_col2, #T_6f768_row17_col3, #T_6f768_row17_col4, #T_6f768_row17_col5, #T_6f768_row17_col6, #T_6f768_row17_col7, #T_6f768_row17_col9, #T_6f768_row17_col10, #T_6f768_row17_col11, #T_6f768_row18_col0, #T_6f768_row18_col1, #T_6f768_row18_col2, #T_6f768_row18_col3, #T_6f768_row18_col4, #T_6f768_row18_col5, #T_6f768_row18_col6, #T_6f768_row18_col7, #T_6f768_row18_col8, #T_6f768_row18_col9, #T_6f768_row18_col11, #T_6f768_row19_col0, #T_6f768_row19_col1, #T_6f768_row19_col2, #T_6f768_row19_col3, #T_6f768_row19_col4, #T_6f768_row19_col5, #T_6f768_row19_col6, #T_6f768_row19_col7, #T_6f768_row19_col8, #T_6f768_row19_col9, #T_6f768_row19_col11, #T_6f768_row20_col0, #T_6f768_row20_col1, #T_6f768_row20_col2, #T_6f768_row20_col3, #T_6f768_row20_col4, #T_6f768_row20_col5, #T_6f768_row20_col6, #T_6f768_row20_col7, #T_6f768_row20_col8, #T_6f768_row20_col9, #T_6f768_row20_col11, #T_6f768_row21_col0, #T_6f768_row21_col1, #T_6f768_row21_col2, #T_6f768_row21_col3, #T_6f768_row21_col4, #T_6f768_row21_col5, #T_6f768_row21_col6, #T_6f768_row21_col7, #T_6f768_row21_col8, #T_6f768_row21_col9, #T_6f768_row21_col11, #T_6f768_row22_col0, #T_6f768_row22_col1, #T_6f768_row22_col2, #T_6f768_row22_col3, #T_6f768_row22_col4, #T_6f768_row22_col5, #T_6f768_row22_col6, #T_6f768_row22_col7, #T_6f768_row22_col8, #T_6f768_row22_col9, #T_6f768_row22_col10 {\n",
       "  background: white;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_6f768_\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th class=\"col_heading level0 col0\" >C51</th>\n",
       "      <th class=\"col_heading level0 col1\" >C52</th>\n",
       "      <th class=\"col_heading level0 col2\" >C141</th>\n",
       "      <th class=\"col_heading level0 col3\" >C142</th>\n",
       "      <th class=\"col_heading level0 col4\" >C148</th>\n",
       "      <th class=\"col_heading level0 col5\" >C149</th>\n",
       "      <th class=\"col_heading level0 col6\" >C100</th>\n",
       "      <th class=\"col_heading level0 col7\" >C144</th>\n",
       "      <th class=\"col_heading level0 col8\" >C143</th>\n",
       "      <th class=\"col_heading level0 col9\" >C145</th>\n",
       "      <th class=\"col_heading level0 col10\" >C146</th>\n",
       "      <th class=\"col_heading level0 col11\" >C152</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row0\" class=\"row_heading level0 row0\" >raw_data</th>\n",
       "      <td id=\"T_6f768_row0_col0\" class=\"data row0 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col1\" class=\"data row0 col1\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col2\" class=\"data row0 col2\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col3\" class=\"data row0 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col4\" class=\"data row0 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col5\" class=\"data row0 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col6\" class=\"data row0 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col7\" class=\"data row0 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col8\" class=\"data row0 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col9\" class=\"data row0 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col10\" class=\"data row0 col10\" >+</td>\n",
       "      <td id=\"T_6f768_row0_col11\" class=\"data row0 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row1\" class=\"row_heading level0 row1\" >raw_data_Macrophage</th>\n",
       "      <td id=\"T_6f768_row1_col0\" class=\"data row1 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col1\" class=\"data row1 col1\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col2\" class=\"data row1 col2\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col3\" class=\"data row1 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col4\" class=\"data row1 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col5\" class=\"data row1 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col6\" class=\"data row1 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col7\" class=\"data row1 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col8\" class=\"data row1 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col9\" class=\"data row1 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col10\" class=\"data row1 col10\" >+</td>\n",
       "      <td id=\"T_6f768_row1_col11\" class=\"data row1 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row2\" class=\"row_heading level0 row2\" >raw_data_T_cells</th>\n",
       "      <td id=\"T_6f768_row2_col0\" class=\"data row2 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col1\" class=\"data row2 col1\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col2\" class=\"data row2 col2\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col3\" class=\"data row2 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col4\" class=\"data row2 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col5\" class=\"data row2 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col6\" class=\"data row2 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col7\" class=\"data row2 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col8\" class=\"data row2 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col9\" class=\"data row2 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col10\" class=\"data row2 col10\" >+</td>\n",
       "      <td id=\"T_6f768_row2_col11\" class=\"data row2 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row3\" class=\"row_heading level0 row3\" >raw_data_DC</th>\n",
       "      <td id=\"T_6f768_row3_col0\" class=\"data row3 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col1\" class=\"data row3 col1\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col2\" class=\"data row3 col2\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col3\" class=\"data row3 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col4\" class=\"data row3 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col5\" class=\"data row3 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col6\" class=\"data row3 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col7\" class=\"data row3 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col8\" class=\"data row3 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col9\" class=\"data row3 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col10\" class=\"data row3 col10\" >+</td>\n",
       "      <td id=\"T_6f768_row3_col11\" class=\"data row3 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row4\" class=\"row_heading level0 row4\" >raw_data_Pre-B_cell_CD34-</th>\n",
       "      <td id=\"T_6f768_row4_col0\" class=\"data row4 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row4_col1\" class=\"data row4 col1\" >+</td>\n",
       "      <td id=\"T_6f768_row4_col2\" class=\"data row4 col2\" >-</td>\n",
       "      <td id=\"T_6f768_row4_col3\" class=\"data row4 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row4_col4\" class=\"data row4 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row4_col5\" class=\"data row4 col5\" >-</td>\n",
       "      <td id=\"T_6f768_row4_col6\" class=\"data row4 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row4_col7\" class=\"data row4 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row4_col8\" class=\"data row4 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row4_col9\" class=\"data row4 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row4_col10\" class=\"data row4 col10\" >nan</td>\n",
       "      <td id=\"T_6f768_row4_col11\" class=\"data row4 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row5\" class=\"row_heading level0 row5\" >raw_data_Monocyte</th>\n",
       "      <td id=\"T_6f768_row5_col0\" class=\"data row5 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col1\" class=\"data row5 col1\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col2\" class=\"data row5 col2\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col3\" class=\"data row5 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col4\" class=\"data row5 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col5\" class=\"data row5 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col6\" class=\"data row5 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col7\" class=\"data row5 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col8\" class=\"data row5 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col9\" class=\"data row5 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col10\" class=\"data row5 col10\" >+</td>\n",
       "      <td id=\"T_6f768_row5_col11\" class=\"data row5 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row6\" class=\"row_heading level0 row6\" >raw_data_NK_cell</th>\n",
       "      <td id=\"T_6f768_row6_col0\" class=\"data row6 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col1\" class=\"data row6 col1\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col2\" class=\"data row6 col2\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col3\" class=\"data row6 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col4\" class=\"data row6 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col5\" class=\"data row6 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col6\" class=\"data row6 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col7\" class=\"data row6 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col8\" class=\"data row6 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col9\" class=\"data row6 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col10\" class=\"data row6 col10\" >+</td>\n",
       "      <td id=\"T_6f768_row6_col11\" class=\"data row6 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row7\" class=\"row_heading level0 row7\" >raw_data_B_cell</th>\n",
       "      <td id=\"T_6f768_row7_col0\" class=\"data row7 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col1\" class=\"data row7 col1\" >-</td>\n",
       "      <td id=\"T_6f768_row7_col2\" class=\"data row7 col2\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col3\" class=\"data row7 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col4\" class=\"data row7 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col5\" class=\"data row7 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col6\" class=\"data row7 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col7\" class=\"data row7 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col8\" class=\"data row7 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col9\" class=\"data row7 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col10\" class=\"data row7 col10\" >+</td>\n",
       "      <td id=\"T_6f768_row7_col11\" class=\"data row7 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row8\" class=\"row_heading level0 row8\" >raw_data_Epithelial_cells</th>\n",
       "      <td id=\"T_6f768_row8_col0\" class=\"data row8 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col1\" class=\"data row8 col1\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col2\" class=\"data row8 col2\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col3\" class=\"data row8 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col4\" class=\"data row8 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col5\" class=\"data row8 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col6\" class=\"data row8 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col7\" class=\"data row8 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col8\" class=\"data row8 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col9\" class=\"data row8 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col10\" class=\"data row8 col10\" >+</td>\n",
       "      <td id=\"T_6f768_row8_col11\" class=\"data row8 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row9\" class=\"row_heading level0 row9\" >raw_data_BM</th>\n",
       "      <td id=\"T_6f768_row9_col0\" class=\"data row9 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row9_col1\" class=\"data row9 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row9_col2\" class=\"data row9 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row9_col3\" class=\"data row9 col3\" >-</td>\n",
       "      <td id=\"T_6f768_row9_col4\" class=\"data row9 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row9_col5\" class=\"data row9 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row9_col6\" class=\"data row9 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row9_col7\" class=\"data row9 col7\" >-</td>\n",
       "      <td id=\"T_6f768_row9_col8\" class=\"data row9 col8\" >nan</td>\n",
       "      <td id=\"T_6f768_row9_col9\" class=\"data row9 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row9_col10\" class=\"data row9 col10\" >nan</td>\n",
       "      <td id=\"T_6f768_row9_col11\" class=\"data row9 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row10\" class=\"row_heading level0 row10\" >raw_data_Pro-B_cell_CD34+</th>\n",
       "      <td id=\"T_6f768_row10_col0\" class=\"data row10 col0\" >+</td>\n",
       "      <td id=\"T_6f768_row10_col1\" class=\"data row10 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row10_col2\" class=\"data row10 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row10_col3\" class=\"data row10 col3\" >-</td>\n",
       "      <td id=\"T_6f768_row10_col4\" class=\"data row10 col4\" >-</td>\n",
       "      <td id=\"T_6f768_row10_col5\" class=\"data row10 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row10_col6\" class=\"data row10 col6\" >-</td>\n",
       "      <td id=\"T_6f768_row10_col7\" class=\"data row10 col7\" >-</td>\n",
       "      <td id=\"T_6f768_row10_col8\" class=\"data row10 col8\" >-</td>\n",
       "      <td id=\"T_6f768_row10_col9\" class=\"data row10 col9\" >-</td>\n",
       "      <td id=\"T_6f768_row10_col10\" class=\"data row10 col10\" >nan</td>\n",
       "      <td id=\"T_6f768_row10_col11\" class=\"data row10 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row11\" class=\"row_heading level0 row11\" >raw_data_HSC_-G-CSF</th>\n",
       "      <td id=\"T_6f768_row11_col0\" class=\"data row11 col0\" >-</td>\n",
       "      <td id=\"T_6f768_row11_col1\" class=\"data row11 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row11_col2\" class=\"data row11 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row11_col3\" class=\"data row11 col3\" >-</td>\n",
       "      <td id=\"T_6f768_row11_col4\" class=\"data row11 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row11_col5\" class=\"data row11 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row11_col6\" class=\"data row11 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row11_col7\" class=\"data row11 col7\" >-</td>\n",
       "      <td id=\"T_6f768_row11_col8\" class=\"data row11 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row11_col9\" class=\"data row11 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row11_col10\" class=\"data row11 col10\" >-</td>\n",
       "      <td id=\"T_6f768_row11_col11\" class=\"data row11 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row12\" class=\"row_heading level0 row12\" >raw_data_Neutrophils</th>\n",
       "      <td id=\"T_6f768_row12_col0\" class=\"data row12 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row12_col1\" class=\"data row12 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row12_col2\" class=\"data row12 col2\" >-</td>\n",
       "      <td id=\"T_6f768_row12_col3\" class=\"data row12 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row12_col4\" class=\"data row12 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row12_col5\" class=\"data row12 col5\" >+</td>\n",
       "      <td id=\"T_6f768_row12_col6\" class=\"data row12 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row12_col7\" class=\"data row12 col7\" >+</td>\n",
       "      <td id=\"T_6f768_row12_col8\" class=\"data row12 col8\" >+</td>\n",
       "      <td id=\"T_6f768_row12_col9\" class=\"data row12 col9\" >+</td>\n",
       "      <td id=\"T_6f768_row12_col10\" class=\"data row12 col10\" >+</td>\n",
       "      <td id=\"T_6f768_row12_col11\" class=\"data row12 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row13\" class=\"row_heading level0 row13\" >raw_data_CMP</th>\n",
       "      <td id=\"T_6f768_row13_col0\" class=\"data row13 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row13_col1\" class=\"data row13 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row13_col2\" class=\"data row13 col2\" >-</td>\n",
       "      <td id=\"T_6f768_row13_col3\" class=\"data row13 col3\" >+</td>\n",
       "      <td id=\"T_6f768_row13_col4\" class=\"data row13 col4\" >+</td>\n",
       "      <td id=\"T_6f768_row13_col5\" class=\"data row13 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row13_col6\" class=\"data row13 col6\" >+</td>\n",
       "      <td id=\"T_6f768_row13_col7\" class=\"data row13 col7\" >-</td>\n",
       "      <td id=\"T_6f768_row13_col8\" class=\"data row13 col8\" >-</td>\n",
       "      <td id=\"T_6f768_row13_col9\" class=\"data row13 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row13_col10\" class=\"data row13 col10\" >nan</td>\n",
       "      <td id=\"T_6f768_row13_col11\" class=\"data row13 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row14\" class=\"row_heading level0 row14\" >raw_data_GMP</th>\n",
       "      <td id=\"T_6f768_row14_col0\" class=\"data row14 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row14_col1\" class=\"data row14 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row14_col2\" class=\"data row14 col2\" >-</td>\n",
       "      <td id=\"T_6f768_row14_col3\" class=\"data row14 col3\" >-</td>\n",
       "      <td id=\"T_6f768_row14_col4\" class=\"data row14 col4\" >-</td>\n",
       "      <td id=\"T_6f768_row14_col5\" class=\"data row14 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row14_col6\" class=\"data row14 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row14_col7\" class=\"data row14 col7\" >nan</td>\n",
       "      <td id=\"T_6f768_row14_col8\" class=\"data row14 col8\" >-</td>\n",
       "      <td id=\"T_6f768_row14_col9\" class=\"data row14 col9\" >-</td>\n",
       "      <td id=\"T_6f768_row14_col10\" class=\"data row14 col10\" >nan</td>\n",
       "      <td id=\"T_6f768_row14_col11\" class=\"data row14 col11\" >+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row15\" class=\"row_heading level0 row15\" >raw_data_Gametocytes</th>\n",
       "      <td id=\"T_6f768_row15_col0\" class=\"data row15 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row15_col1\" class=\"data row15 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row15_col2\" class=\"data row15 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row15_col3\" class=\"data row15 col3\" >nan</td>\n",
       "      <td id=\"T_6f768_row15_col4\" class=\"data row15 col4\" >-</td>\n",
       "      <td id=\"T_6f768_row15_col5\" class=\"data row15 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row15_col6\" class=\"data row15 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row15_col7\" class=\"data row15 col7\" >nan</td>\n",
       "      <td id=\"T_6f768_row15_col8\" class=\"data row15 col8\" >-</td>\n",
       "      <td id=\"T_6f768_row15_col9\" class=\"data row15 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row15_col10\" class=\"data row15 col10\" >-</td>\n",
       "      <td id=\"T_6f768_row15_col11\" class=\"data row15 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row16\" class=\"row_heading level0 row16\" >raw_data_Erythroblast</th>\n",
       "      <td id=\"T_6f768_row16_col0\" class=\"data row16 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col1\" class=\"data row16 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col2\" class=\"data row16 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col3\" class=\"data row16 col3\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col4\" class=\"data row16 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col5\" class=\"data row16 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col6\" class=\"data row16 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col7\" class=\"data row16 col7\" >-</td>\n",
       "      <td id=\"T_6f768_row16_col8\" class=\"data row16 col8\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col9\" class=\"data row16 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col10\" class=\"data row16 col10\" >nan</td>\n",
       "      <td id=\"T_6f768_row16_col11\" class=\"data row16 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row17\" class=\"row_heading level0 row17\" >raw_data_Neurons</th>\n",
       "      <td id=\"T_6f768_row17_col0\" class=\"data row17 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col1\" class=\"data row17 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col2\" class=\"data row17 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col3\" class=\"data row17 col3\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col4\" class=\"data row17 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col5\" class=\"data row17 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col6\" class=\"data row17 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col7\" class=\"data row17 col7\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col8\" class=\"data row17 col8\" >-</td>\n",
       "      <td id=\"T_6f768_row17_col9\" class=\"data row17 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col10\" class=\"data row17 col10\" >nan</td>\n",
       "      <td id=\"T_6f768_row17_col11\" class=\"data row17 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row18\" class=\"row_heading level0 row18\" >raw_data_Fibroblasts</th>\n",
       "      <td id=\"T_6f768_row18_col0\" class=\"data row18 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col1\" class=\"data row18 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col2\" class=\"data row18 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col3\" class=\"data row18 col3\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col4\" class=\"data row18 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col5\" class=\"data row18 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col6\" class=\"data row18 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col7\" class=\"data row18 col7\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col8\" class=\"data row18 col8\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col9\" class=\"data row18 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row18_col10\" class=\"data row18 col10\" >-</td>\n",
       "      <td id=\"T_6f768_row18_col11\" class=\"data row18 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row19\" class=\"row_heading level0 row19\" >raw_data_Smooth_muscle_cells</th>\n",
       "      <td id=\"T_6f768_row19_col0\" class=\"data row19 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col1\" class=\"data row19 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col2\" class=\"data row19 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col3\" class=\"data row19 col3\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col4\" class=\"data row19 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col5\" class=\"data row19 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col6\" class=\"data row19 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col7\" class=\"data row19 col7\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col8\" class=\"data row19 col8\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col9\" class=\"data row19 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row19_col10\" class=\"data row19 col10\" >-</td>\n",
       "      <td id=\"T_6f768_row19_col11\" class=\"data row19 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row20\" class=\"row_heading level0 row20\" >raw_data_Hepatocytes</th>\n",
       "      <td id=\"T_6f768_row20_col0\" class=\"data row20 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col1\" class=\"data row20 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col2\" class=\"data row20 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col3\" class=\"data row20 col3\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col4\" class=\"data row20 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col5\" class=\"data row20 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col6\" class=\"data row20 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col7\" class=\"data row20 col7\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col8\" class=\"data row20 col8\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col9\" class=\"data row20 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row20_col10\" class=\"data row20 col10\" >-</td>\n",
       "      <td id=\"T_6f768_row20_col11\" class=\"data row20 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row21\" class=\"row_heading level0 row21\" >raw_data_Keratinocytes</th>\n",
       "      <td id=\"T_6f768_row21_col0\" class=\"data row21 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col1\" class=\"data row21 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col2\" class=\"data row21 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col3\" class=\"data row21 col3\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col4\" class=\"data row21 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col5\" class=\"data row21 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col6\" class=\"data row21 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col7\" class=\"data row21 col7\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col8\" class=\"data row21 col8\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col9\" class=\"data row21 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row21_col10\" class=\"data row21 col10\" >-</td>\n",
       "      <td id=\"T_6f768_row21_col11\" class=\"data row21 col11\" >nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_6f768_level0_row22\" class=\"row_heading level0 row22\" >raw_data_Pro-Myelocyte</th>\n",
       "      <td id=\"T_6f768_row22_col0\" class=\"data row22 col0\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col1\" class=\"data row22 col1\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col2\" class=\"data row22 col2\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col3\" class=\"data row22 col3\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col4\" class=\"data row22 col4\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col5\" class=\"data row22 col5\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col6\" class=\"data row22 col6\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col7\" class=\"data row22 col7\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col8\" class=\"data row22 col8\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col9\" class=\"data row22 col9\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col10\" class=\"data row22 col10\" >nan</td>\n",
       "      <td id=\"T_6f768_row22_col11\" class=\"data row22 col11\" >-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x7fe3d0b11450>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Setting pathways to files\n",
    "_PROJ_PATH = '/gpfs/projects/bsc08/bsc08890'\n",
    "_FMETA = os.path.join(_PROJ_PATH, 'data/GSE145926_RAW/metadata.tsv')\n",
    "_DATA_HOME = os.path.join(_PROJ_PATH, 'res/covid_19')\n",
    "\n",
    "# Data types\n",
    "_DATA_TYPES = ['all', 'TF', 'ctx']\n",
    "_DTYPE_TITLES = ['gene-gene links', 'TF-gene links', 'enriched TF-gene links']\n",
    "\n",
    "# Limiting the metadata to certain patients to fit into RAM\n",
    "_FIXED_PATIENTS = ['C51', 'C52', 'C141', 'C142', 'C148', 'C149']\n",
    "_FIXED_CELL_TYPES = ['T_cells', 'Macrophage', 'NK_cell', 'Monocyte']\n",
    "\n",
    "# Loading sample meta data, reordering patients\n",
    "full_meta = pd.read_csv(_FMETA, sep='\\t', index_col=0)\n",
    "full_meta = full_meta.reindex(index=_FIXED_PATIENTS + ['C100', 'C144', 'C143', 'C145', 'C146', 'C152'])\n",
    "\n",
    "# Adding cell-type frequency to the metadat\n",
    "for patient_id in full_meta.index:\n",
    "    cells_metadata = pd.read_csv(os.path.join(_DATA_HOME, patient_id, 'data', 'Seurat', 'cells_metadata.tsv'), sep='\\t')\n",
    "    full_meta.loc[patient_id, 'num_cells'] = len(cells_metadata)\n",
    "    for cell_type, cell_count in cells_metadata['cell_type'].value_counts().iteritems():\n",
    "        full_meta.loc[patient_id, cell_type] = cell_count   \n",
    "full_cell_type_freq = full_meta.iloc[:, 3:]\n",
    "\n",
    "# Finding common cell types\n",
    "_COMMON_CELL_TYPES = full_cell_type_freq.T.dropna().index.to_list()\n",
    "_COMMON_FILE_PREFIXES = ['raw_data', 'raw_data_HVG'] + list(map(lambda x: 'raw_data_' + x, _COMMON_CELL_TYPES))\n",
    "\n",
    "# Limiting the metadata to certain patients to fit into RAM\n",
    "meta = full_meta.loc[_FIXED_PATIENTS, ['group', 'file', 'num_cells'] + _FIXED_CELL_TYPES]\n",
    "cell_type_freq = full_cell_type_freq[_FIXED_CELL_TYPES]\n",
    "\n",
    "# Getting information about all patients\n",
    "_ALL_PATIENTS = full_meta.index.to_list()\n",
    "_ALL_PATIENT_CELL_TYPES = {\n",
    "    pat: full_cell_type_freq.loc[pat].dropna().index.to_list() for pat in _ALL_PATIENTS\n",
    "}\n",
    "_ALL_FILE_PREFIXES = {\n",
    "    pat: ['raw_data'] + list(map(lambda x: 'raw_data_' + x, cell_types)) for pat, cell_types in _ALL_PATIENT_CELL_TYPES.items()\n",
    "}\n",
    "_ALL_TITLES = {\n",
    "    pat: list(map(lambda x: 'all data' if x == 'raw_data' else 'all data, only HVGs' if 'HVG' in x else x.replace('raw_data_', '').replace('_', ' '), files)) \n",
    "    for pat, files in _ALL_FILE_PREFIXES.items()\n",
    "}\n",
    "_ALL_FIG_DIRS = {\n",
    "    pat: os.path.join(_DATA_HOME, pat, 'figs/grnboost2') for pat in _ALL_PATIENTS\n",
    "}\n",
    "        \n",
    "# Setting patient-specific params\n",
    "_PATIENTS = meta.index.to_list()\n",
    "_PATIENT_CELL_TYPES = {\n",
    "    pat: cell_type_freq.loc[pat].dropna().index.to_list() for pat in _PATIENTS\n",
    "}\n",
    "_FILE_PREFIXES = {\n",
    "    pat: ['raw_data'] + list(map(lambda x: 'raw_data_' + x, cell_types)) for pat, cell_types in _PATIENT_CELL_TYPES.items()\n",
    "}\n",
    "_TITLES = {\n",
    "    pat: list(map(lambda x: 'all data' if x == 'raw_data' else 'all data, only HVGs' if 'HVG' in x else x.replace('raw_data_', '').replace('_', ' '), files)) \n",
    "    for pat, files in _FILE_PREFIXES.items()\n",
    "}\n",
    "_FIG_DIRS = {\n",
    "    pat: os.path.join(_DATA_HOME, pat, 'figs/grnboost2') for pat in _PATIENTS\n",
    "}\n",
    "\n",
    "for fig_dir in _FIG_DIRS.values():\n",
    "    os.makedirs(fig_dir, exist_ok=True)\n",
    "    \n",
    "# Loading TFs\n",
    "TF_names = pd.read_csv(os.path.join(_PROJ_PATH, 'data/TF_lists/lambert2018.txt'), header=None)[0].to_list()\n",
    "\n",
    "# Loading gene GO description\n",
    "gene_func = pd.read_csv(os.path.join(_PROJ_PATH, 'data/gene_go_func.txt'), sep='\\t', index_col=0) \\\n",
    "    .rename(columns={'GO term name': 'GO_func'}).drop_duplicates().dropna()['GO_func']\n",
    "\n",
    "# Loading exemplary data\n",
    "_EXAMPLE_PATIENT = 'C141'\n",
    "_INPUT_NORM_FN = os.path.join(_DATA_HOME, f'{_EXAMPLE_PATIENT}/data/Seurat/norm_data.tsv')\n",
    "_INPUT_NORM_HVG_FN = os.path.join(_DATA_HOME, f'{_EXAMPLE_PATIENT}/data/Seurat/norm_data_HVG.tsv')\n",
    "ex_df = pd.read_csv(_INPUT_NORM_FN, sep='\\t').T\n",
    "ex_HVG_df = pd.read_csv(_INPUT_NORM_HVG_FN, sep='\\t').T\n",
    "ex_HVGs = ex_HVG_df.columns\n",
    "\n",
    "# Quick reporting\n",
    "print(f'As an example, the expression matrix of {_EXAMPLE_PATIENT} patient contains {ex_df.shape[0]} cells and {ex_df.shape[1]} genes:')\n",
    "print(f'And as an example, the dataset from {_EXAMPLE_PATIENT} patient contains {len(ex_HVGs)} highly-variable genes.')\n",
    "display(ex_df.head())\n",
    "print(colored('Full metadata:', 'green'))\n",
    "display(full_meta)\n",
    "print(colored('Current metadata:', 'green'))\n",
    "display(meta)\n",
    "print(colored('Common cell types among different patients (full metadata):', 'green'))\n",
    "display(full_cell_type_freq.T.dropna())\n",
    "\n",
    "# Creating data availability dictionary\n",
    "_AVAILABLE_DATA = {}\n",
    "for pat in _ALL_PATIENTS:\n",
    "    _AVAILABLE_DATA[pat] = {}\n",
    "    for data_type, suf in zip(['all', 'TF', 'ctx'], ['cor', 'TF_cor', 'TF_ctx']):\n",
    "        _AVAILABLE_DATA[pat][data_type] = [\n",
    "            fn for fn in _ALL_FILE_PREFIXES[pat] \\\n",
    "                if os.path.exists(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_{suf}.pickle')) and \\\n",
    "                (os.stat(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_TF_cor.pickle')).st_size != 0)\n",
    "        ]\n",
    "\n",
    "# Displaying available data\n",
    "data_availability = {}\n",
    "for data_type, suf in zip(['all', 'TF', 'ctx'], ['cor', 'TF_cor', 'TF_ctx']):\n",
    "    data_availability[data_type] = {}\n",
    "    for pat in _ALL_PATIENTS:\n",
    "        data_availability[data_type][pat] = {}\n",
    "        for fn in _ALL_FILE_PREFIXES[pat]:\n",
    "            data_availability[data_type][pat][fn] = None\n",
    "            if os.path.exists(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_{suf}.pickle')):\n",
    "                if os.stat(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_TF_cor.pickle')).st_size != 0:\n",
    "                    data_availability[data_type][pat][fn] = '+'\n",
    "                else:\n",
    "                    data_availability[data_type][pat][fn] = '-'\n",
    "            else:\n",
    "                data_availability[data_type][pat][fn] = '-'\n",
    "    print(colored(data_type.upper(), 'cyan'))\n",
    "    data_availability[data_type] = pd.DataFrame(data_availability[data_type])\n",
    "    display(style_data_availability(data_availability[data_type]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## NDEx database\n",
    "\n",
    "Firstly, we will download the interaction pathways from [NDEx](https://www.ndexbio.org/#/) database as a reference. We will save manually each networkx into `.cx` file and then load them here using NDEx2 API.\n",
    "\n",
    "| Type        | Name                                                                                                                                                      | Source                                                                                                         |   |   |\n",
    "|-------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------|---|---|\n",
    "| NK cells    | [NK Signaling](https://www.ndexbio.org/viewer/networks/b953a31b-9e15-11eb-9e72-0ac135e8bacf)                                                              | [causalbionet](http://causalbionet.com/Home/NetworkVisualization/#/!/networkGraph?ID=nk_signaling_2.0_hs)      |   |   |\n",
    "| B cells     | [WP23 - B cell receptor signaling pathway - Homo sapiens](https://www.ndexbio.org/viewer/networks/00a68ce0-5c65-11ec-b3be-0ac135e8bacf)                   | [wikipathways](https://www.wikipathways.org/index.php/Pathway:WP23)                                            |   |   |\n",
    "| B cells     | [B-cell Signaling](https://www.ndexbio.org/viewer/networks/51d0f5ed-9e16-11eb-9e72-0ac135e8bacf)                                                          | [causalbionet](http://causalbionet.com/Home/NetworkVisualization/#/!/networkGraph?ID=5df92bd801dd760eacf8ebf0) |   |   |\n",
    "| DC cells    | [Dendritic Cell Signaling](https://www.ndexbio.org/viewer/networks/79b8d4a0-9e15-11eb-9e72-0ac135e8bacf)                                                  | [causalbionet](http://causalbionet.com/Home/NetworkVisualization/#/!/networkGraph?ID=765110535)                |   |   |\n",
    "| Macrophages | [Macrophage polarization](https://www.ndexbio.org/viewer/networks/1b3873d6-8f9a-11e8-a4bf-0ac135e8bacf)                                                   | [signor](https://signor.uniroma2.it/pathway_browser.php?organism=&pathway_list=SIGNOR-M1M2&x=24&y=17)          |   |   |\n",
    "| Macrophages | [Macrophage Signaling](https://www.ndexbio.org/viewer/networks/3285ca84-9e16-11eb-9e72-0ac135e8bacf)                                                      | [causalbionet](http://causalbionet.com/Home/NetworkVisualization/#/!/networkGraph?ID=1397028844)               |   |   |\n",
    "| T cells     | [Cytotoxic T-cell Signaling](https://www.ndexbio.org/viewer/networks/bc49319e-9e14-11eb-9e72-0ac135e8bacf)                                                | [causalbionet](http://causalbionet.com/Home/NetworkVisualization/#/!/networkGraph?ID=5df9349f01dd760eacf98b55) |   |   |\n",
    "| T cells     | [TCR signaling in nave CD8 T cells](https://www.ndexbio.org/viewer/networks/15a017bb-6196-11e5-8ac5-06603eb7f303)                                         | nci-pid                                                                                                        |   |   |\n",
    "| T cells     | [WP2583 - T cell receptor and co-stimulatory signaling - Homo sapiens](https://www.ndexbio.org/viewer/networks/13b30225-5c6a-11ec-b3be-0ac135e8bacf)      | [wikipathways](https://www.wikipathways.org/index.php/Pathway:WP2583)                                          |   |   |\n",
    "| T cells     | [T cell activation](https://www.ndexbio.org/viewer/networks/568d095e-1e5f-11e8-b939-0ac135e8bacf)                                                         | [signor](https://signor.uniroma2.it/pathway_browser.php?organism=&pathway_list=SIGNOR-TCA&x=26&y=15)           |   |   |\n",
    "| T cells     | [TCR signaling in nave CD4 T cells](https://www.ndexbio.org/viewer/networks/0c2862fa-6196-11e5-8ac5-06603eb7f303)                                         | nci-pid                                                                                                        |   |   |\n",
    "| T cells     | [WP5072 - Modulators of TCR signaling and T cell activation - Homo sapiens](https://www.ndexbio.org/viewer/networks/6ea1c5d6-5c74-11ec-b3be-0ac135e8bacf) | [wikipathways](https://www.wikipathways.org/index.php/Pathway:WP5072)                                          |   |   |\n",
    "| T cells     | [Downstream signaling in nave CD8 T cells](https://www.ndexbio.org/viewer/networks/fc5e2bc0-b7be-11e9-8bb4-0ac135e8bacf)                                  | nci-pid                                                                                                        |   |   |\n",
    "| T cells     | [WP69 - T-cell receptor (TCR) signaling pathway - Homo sapiens](https://www.ndexbio.org/viewer/networks/937153e5-5c66-11ec-b3be-0ac135e8bacf)             | [wikipathways](https://www.wikipathways.org/index.php/Pathway:WP69)                                            |   |   |\n",
    "| Pathways    | [NF-KB Canonical](https://www.ndexbio.org/viewer/networks/3b9113b1-1e5f-11e8-b939-0ac135e8bacf)                                                           | [signor](https://signor.uniroma2.it/pathway_browser.php?organism=&pathway_list=SIGNOR-NFKBC&x=15&y=22)         |   |   |\n",
    "| Pathways    | [PI3K_AKT Signaling](https://www.ndexbio.org/viewer/networks/0b6cbc27-97f1-11eb-9e72-0ac135e8bacf)                                                        | [signor](https://signor.uniroma2.it/pathway_browser.php?organism=&pathway_list=SIGNOR-PI3K-AKT&x=28&y=10)      |   |   |\n",
    "| Pathways    | [Toll like receptors](https://www.ndexbio.org/viewer/networks/5a2ec134-1e5f-11e8-b939-0ac135e8bacf)                                                       | [signor](https://signor.uniroma2.it/pathway_browser.php?organism=&pathway_list=SIGNOR-TLR&x=14&y=16)           |   |   |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "_NET_FNAMES = glob(os.path.join(_PROJ_PATH, 'data/ndex/*/*.cx'))\n",
    "_NET_NAMES = [net_fn[net_fn.find('ndex/')+5:-3] for net_fn in _NET_FNAMES]\n",
    "_REMOVE_SINGLE_NODES = True\n",
    "\n",
    "published_nets = {\n",
    "    net_n: process_ndex_net(ndex2.create_nice_cx_from_file(net_fn).to_networkx(mode='default'), net_n, _REMOVE_SINGLE_NODES) for net_fn, net_n in zip(_NET_FNAMES, _NET_NAMES)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c806d3ac434e450bb3b0c53d44277368",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='data', options=('NK_cell/NK Signaling', 'B_cell/WP23 - B cell receâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def draw_published_net(data):\n",
    "    seed = 42\n",
    "    G = published_nets[data]\n",
    "    pos = nx.nx_agraph.pygraphviz_layout(G, prog=\"dot\")  # nx.nx.spring_layout(G, seed=seed, k=0.2, iterations=20)\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(40, 20))      \n",
    "    nodes = nx.draw(G, pos, with_labels=True, arrowstyle=\"->\", arrowsize=30, edge_color='k', \n",
    "                    node_color='pink', ax=ax, node_size=_NODE_SIZE)\n",
    "    nx.draw_networkx_nodes(G.subgraph(TF_names), ax=ax, pos=pos, node_color='limegreen', \n",
    "                           node_size=_NODE_SIZE)\n",
    "            \n",
    "    ax.set_title(f'\"{data}\" network from NDEx', fontsize=20)\n",
    "    plt.axis('off')    \n",
    "    \n",
    "widget = interact(draw_published_net, data=_NET_NAMES)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## <span style=\"color:salmon\">GRNBoost2 + pySCENIC</span>\n",
    "\n",
    "Here we just depict how to run this algorithms on a handful of genes in the notebook. Normally one would use the following way:\n",
    "\n",
    "To run in CLI:\n",
    "\n",
    "```\n",
    "cd\n",
    "python src/run_grnboost2.py -i <path_to_input> -o <path_to_output>\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6 Âµs, sys: 2 Âµs, total: 8 Âµs\n",
      "Wall time: 13.1 Âµs\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "if False:\n",
    "    print(f'Running GRNBoost2 with {expression_data.shape[1]} genes..')\n",
    "    \n",
    "    expression_data = HVG_ex_df\n",
    "\n",
    "    # compute the GRN using grnboost2\n",
    "    grnboost2_net = grnboost2(expression_data=expression_data, gene_names=None, tf_names=TF_names)\n",
    "\n",
    "    grnboost2_net.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Instead of running everything here, we will run `grnboost2` in CLI and load results here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "359b53cc4bfa4b0d827a4d91fd944784",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=6.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The first 2 rows of GRNBoost2 output as an example (based on all genes and all cells)..\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>target</th>\n",
       "      <th>importance</th>\n",
       "      <th>regulation</th>\n",
       "      <th>rho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LYZ</td>\n",
       "      <td>AC020656.1</td>\n",
       "      <td>315.984263</td>\n",
       "      <td>1</td>\n",
       "      <td>0.696023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC020656.1</td>\n",
       "      <td>LYZ</td>\n",
       "      <td>293.557246</td>\n",
       "      <td>1</td>\n",
       "      <td>0.696023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           TF      target  importance  regulation       rho\n",
       "0         LYZ  AC020656.1  315.984263           1  0.696023\n",
       "1  AC020656.1         LYZ  293.557246           1  0.696023"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After processing of ctx output..\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>target</th>\n",
       "      <th>importance</th>\n",
       "      <th>regulation</th>\n",
       "      <th>rho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARNT</td>\n",
       "      <td>MFSD13A</td>\n",
       "      <td>1.471749</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.045436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARNT</td>\n",
       "      <td>TRA2A</td>\n",
       "      <td>1.189392</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.046711</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     TF   target  importance  regulation       rho\n",
       "0  ARNT  MFSD13A    1.471749         1.0  0.045436\n",
       "1  ARNT    TRA2A    1.189392         1.0  0.046711"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 11.5 s, sys: 2.91 s, total: 14.4 s\n",
      "Wall time: 26.1 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "grnboost2_meta = {}\n",
    "p_bar = tqdm(_PATIENTS)\n",
    "\n",
    "for pat in p_bar:\n",
    "    p_bar.set_description(f'{pat} patient')\n",
    "    grnboost2_meta[pat] = {}\n",
    "    for fn in _FILE_PREFIXES[pat]:\n",
    "        try:\n",
    "            input_TF = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_TF_cor.pickle'))\n",
    "            input_ctx = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_TF_ctx.pickle'))\n",
    "        except (FileNotFoundError, pd.errors.EmptyDataError):\n",
    "            input_TF = None\n",
    "            ctx = None\n",
    "            input_ctx = None\n",
    "        try:\n",
    "            input_all = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_cor.pickle'))\n",
    "        except (FileNotFoundError, pd.errors.EmptyDataError):\n",
    "            input_all = None\n",
    "        grnboost2_meta[pat][fn] = {\n",
    "        'input_TF': input_TF,\n",
    "        'input_all': input_all, \n",
    "        'filtered_TF': None,\n",
    "        'filtered_all': None,\n",
    "        'graph_TF': None,\n",
    "        'graph_all': None,\n",
    "        'full_graph_TF': None,\n",
    "        'full_graph_all': None,\n",
    "        'input_ctx': input_ctx,\n",
    "        'filtered_ctx': None,\n",
    "        'graph': None,\n",
    "        'full_graph_ctx': None\n",
    "        }    \n",
    "\n",
    "print('The first 2 rows of GRNBoost2 output as an example (based on all genes and all cells)..')\n",
    "display(grnboost2_meta[_PATIENTS[0]]['raw_data']['input_all'].head(2))\n",
    "print('After processing of ctx output..')\n",
    "display(grnboost2_meta[_PATIENTS[0]]['raw_data']['input_ctx'].head(2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### GENERAL STATISTICS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f299ddfcaba24525b1701d9f8f34876b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='pat', options=('C51', 'C52', 'C141', 'C142', 'C148', 'C149', 'C100â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def cell_freq(pat, save_fig):\n",
    "    \n",
    "    color_bars = True\n",
    "    add_non_cell_type = True\n",
    "    start_index = 0 if add_non_cell_type else 2\n",
    "    \n",
    "    sns.set_style(\"whitegrid\")\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(20, 10))\n",
    "    cell_types = full_cell_type_freq.loc[pat].dropna().astype('int32')\n",
    "    if add_non_cell_type:\n",
    "        cell_types = pd.concat([\n",
    "            pd.Series([sum(cell_types)], index=['all']),\n",
    "            cell_types\n",
    "        ])\n",
    "    \n",
    "    if not color_bars:\n",
    "        cell_types = full_cell_type_freq.loc[pat].dropna().astype('int32')\n",
    "        plots = cell_types.plot(ax=ax, kind='bar') \n",
    "        \n",
    "        # Iterrating over the bars one-by-one\n",
    "        for bar in plots.patches:\n",
    "            plots.annotate(bar.get_height(), (bar.get_x() + bar.get_width() / 2,\n",
    "                           bar.get_height()), ha='center', va='center',\n",
    "                           size=15, xytext=(0, 8), textcoords='offset points')\n",
    "    else:\n",
    "        tmp_tf = pd.Series(\n",
    "            ['g' if fn in _AVAILABLE_DATA[pat]['ctx'] else 'r' for fn in _ALL_FILE_PREFIXES[pat][start_index:]], \n",
    "            name='TF_present', \n",
    "            index=_ALL_FILE_PREFIXES[pat][start_index:]\n",
    "        )\n",
    "        tmp_tf.index = tmp_tf.index.map(lambda x: 'all' if x == 'raw_data' else x.replace('raw_data_', ''))\n",
    "        tmp_all = pd.Series(\n",
    "            ['g' if fn in _AVAILABLE_DATA[pat]['all'] else 'r' for fn in _ALL_FILE_PREFIXES[pat][start_index:]], \n",
    "            name='all_present', \n",
    "            index=_ALL_FILE_PREFIXES[pat][start_index:]\n",
    "        )\n",
    "        tmp_all.index = tmp_tf.index.map(lambda x: 'all' if x == 'raw_data' else x.replace('raw_data_', ''))\n",
    "        tmp_df = pd.DataFrame(\n",
    "            [\n",
    "                (cell_types / 2).rename('TF'),\n",
    "                (cell_types / 2).rename('all'), \n",
    "                tmp_tf,\n",
    "                tmp_all\n",
    "            ]\n",
    "        ).T\n",
    "    \n",
    "        plots = tmp_df[['TF', 'all']].plot(ax=ax, kind='bar', stacked=True, legend=None)\n",
    "        \n",
    "        # Iterrating over the bars one-by-one\n",
    "        for bar in plots.patches[:len(plots.patches) // 2]:\n",
    "            plots.annotate(int(2 * bar.get_height()), (bar.get_x() + bar.get_width() / 2,\n",
    "                           2 * bar.get_height()), ha='center', va='center',\n",
    "                           size=15, xytext=(0, 8), textcoords='offset points')\n",
    "            \n",
    "        for bar, c in zip(plots.patches, tmp_tf.to_list() + tmp_all.to_list()):\n",
    "            bar.set_color(c)\n",
    "    \n",
    "    backslash = '\\n'\n",
    "    ax.set_title(f\"{'Cell' if not add_non_cell_type else 'Data/cell'} type frequency in {pat} patient ({int(full_meta['num_cells'][pat])} cells) \"\n",
    "                 f\"{'' if not color_bars else f'{backslash}(TF_ctx - bottom, all - top)'}\")\n",
    "    ax.set_xlabel('Cell type')\n",
    "    ax.set_ylabel('Frequency')\n",
    "    \n",
    "    plt.xticks(rotation=45)\n",
    "    \n",
    "    if save_fig:\n",
    "        plt.savefig(os.path.join(_ALL_FIG_DIRS[pat], f'{pat}_full_cell_type_freq.png'), dpi=200, bbox_inches='tight')\n",
    "    \n",
    "widget = interact(cell_freq, pat=_ALL_PATIENTS, save_fig=[False, True])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6fe8453b33a54c629ede5ecf1d558624",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='Patient:', options=(('C51 (C)', 'C51'), ('C52 (C)', 'C52'), ('C141â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def genstats(pat, data_type, data, save_fig):\n",
    "    \n",
    "    sns.set_style(\"whitegrid\")\n",
    "    n_bins = 100\n",
    "    dtype_title = _DTYPE_TITLES[_DATA_TYPES.index(data_type)]\n",
    "    data_title = _ALL_TITLES[pat][_ALL_FILE_PREFIXES[pat].index(data)]\n",
    "    \n",
    "    try:\n",
    "        df = grnboost2_meta[pat][data][f'input_{data_type}']\n",
    "    except KeyError:\n",
    "        df = get_adj_list(pat, data, data_type)\n",
    "    \n",
    "    num_links = len(df)\n",
    "    pos_links, neg_links = (df[\"rho\"] > 0).sum(), (df[\"rho\"] < 0).sum()\n",
    "    top_tfs = df[['TF', 'target']].groupby('TF').size().nlargest(3)\n",
    "    top_targets = df[['TF', 'target']].groupby('target').size().nlargest(3)\n",
    "    import_50, import_75, import_max = df['importance'].median(), df['importance'].quantile(0.75), df['importance'].max()\n",
    "    top_high_tfs = df.loc[lambda x: x.importance > import_50][['TF', 'target']].groupby('TF').size().nlargest(3)\n",
    "    top_high_targets = df.loc[lambda x: x.importance > import_50][['TF', 'target']].groupby('target').size().nlargest(3)    \n",
    "    \n",
    "    print(f'The chosen dataset is \"{data}_{data_type}\" which contains {num_links} connections')\n",
    "    print(f'Positive connections - {pos_links} ({pos_links/num_links*100:.1f} %), '\n",
    "          f'negative connections - {neg_links} ({neg_links/num_links*100:.1f} %)')\n",
    "    print(f'The dataset contains {df[\"TF\"].nunique()} TFs out of {len(TF_names) if \"TF\" in data_type else ex_df.shape[1]} and {df[\"target\"].nunique()} targets')\n",
    "    print(f'TFs with maximum connections: {top_tfs.index[0]} ({top_tfs[0]}), {top_tfs.index[1]} ({top_tfs[1]}), {top_tfs.index[2]} ({top_tfs[2]}); '\n",
    "          f'targets with maximum connections: {top_targets.index[0]} ({top_targets[0]}), {top_targets.index[1]} ({top_targets[1]}), {top_targets.index[2]} ({top_targets[2]})')\n",
    "    print(f'TFs with maximum connections (importance > median): {top_high_tfs.index[0]} ({top_high_tfs[0]}), {top_high_tfs.index[1]} ({top_high_tfs[1]}), {top_high_tfs.index[2]} ({top_high_tfs[2]}); '\n",
    "          f'targets with maximum connections (importance > median): {top_high_targets.index[0]} ({top_high_targets[0]}), {top_high_targets.index[1]} ({top_high_targets[1]}), {top_high_targets.index[2]} ({top_high_targets[2]})..')\n",
    "    print()\n",
    "    display(df.describe().drop(index=['count'], columns=['regulation']).T)\n",
    "    print()\n",
    "    \n",
    "    cmap=plt.cm.plasma\n",
    "    color_bars = False\n",
    "    \n",
    "    # Plotting gene-gene link importance distribution\n",
    "    f, ax = plt.subplots(1, 2, figsize=(20, 8))\n",
    "\n",
    "    if color_bars:\n",
    "        y, x = np.histogram(df['importance'], n_bins, density=False)\n",
    "        colors = [cmap((x_i - x.min()) / (x.max() - x.min())) for x_i in x]\n",
    "        ax[0].bar(x[:-1], y, color=colors, width=x[1]-x[0])\n",
    "        ax[0].set_yscale('log')\n",
    "    else:\n",
    "        df['importance'].plot(ax=ax[0], bins=n_bins, kind='hist', logy=True)\n",
    "    ax[0].set_title(f'GRNBoost2-inferred gene-gene connection importance')\n",
    "    ax[0].set_xlabel('Gene-gene connection importance')\n",
    "    \n",
    "    if color_bars:\n",
    "        df_rho = df['rho'] if not np.all(np.isfinite(df['rho'])) else df['rho'][np.isfinite(df['rho'])]\n",
    "        y_rho, x_rho = np.histogram(df_rho, n_bins, density=False)\n",
    "        colors_rho = [None]*len(x)\n",
    "        for i, rho in enumerate(x_rho):\n",
    "            importance_closest = df.iloc[(df['rho']-rho).abs().argsort()[:3]]['importance'].mean()\n",
    "            id_closest = (np.abs(x - importance_closest)).argmin()\n",
    "            colors_rho[i] = colors[id_closest]\n",
    "        ax[1].bar(x_rho[:-1], y_rho, color=colors_rho, width=x_rho[1] - x_rho[0])\n",
    "        ax[1].set_yscale('log')\n",
    "    else:\n",
    "        df['rho'].plot(ax=ax[1], bins=n_bins, density=True, kind='hist', logy=True)\n",
    "    ax[1].set_title(f'Correlation between linked genes')\n",
    "    ax[1].set_xlabel('Spearman correlation')\n",
    "    \n",
    "    f.suptitle(f'{pat}, {dtype_title}, {data_title}', fontsize=20)\n",
    "    \n",
    "    if save_fig:\n",
    "        plt.savefig(os.path.join(_ALL_FIG_DIRS[pat], f'{pat}_{data}_{data_type}_genstats.png'), dpi=200, bbox_inches='tight') \n",
    "\n",
    "annotate_option = lambda pat, data: \\\n",
    "    f'{data} ({int(full_meta.loc[pat, data.replace(\"raw_data_\", \"\")]) if data != \"raw_data\" else int(full_meta.loc[pat, \"num_cells\"])} patients)'\n",
    "        \n",
    "def pick_pat_dtypes(_):\n",
    "    data_dropdown.options = [(annotate_option(pat_dropdown.value, data), data) for data in _AVAILABLE_DATA[pat_dropdown.value][dtype_dropdown.value]]\n",
    "        \n",
    "pat_dropdown = Dropdown(description='Patient:', options=[(f\"{p} ({full_meta.loc[p, 'group']})\", p) for p in _ALL_PATIENTS])\n",
    "dtype_dropdown = Dropdown(description='Data type:', options=_DATA_TYPES)\n",
    "data_dropdown = Dropdown(description='Data:', options=[(annotate_option(_ALL_PATIENTS[0], data), data) for data in _ALL_FILE_PREFIXES[_ALL_PATIENTS[0]]])\n",
    "pat_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "dtype_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "\n",
    "widget = interact(genstats, pat=pat_dropdown, data_type=dtype_dropdown, data=data_dropdown, save_fig=[False, True])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FILTERING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36mPATIENT: C51\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 3007789 links -> 150390 links; \u001b[31mALL\u001b[0m: 4356932 links -> 217847 links; \u001b[31mCTX\u001b[0m: 46102 links -> 2306 links\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 726639 links -> 36332 links; \u001b[31mALL\u001b[0m: 919907 links -> 45996 links; \u001b[31mCTX\u001b[0m: 28443 links -> 1423 links\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 2637459 links -> 131873 links; \u001b[31mALL\u001b[0m: 3787950 links -> 189398 links; \u001b[31mCTX\u001b[0m: 39495 links -> 1975 links\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 313398 links -> 15670 links; \u001b[31mALL\u001b[0m: 434537 links -> 21727 links; \u001b[31mCTX\u001b[0m: 9649 links -> 483 links\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 481252 links -> 24063 links; \u001b[31mALL\u001b[0m: 860872 links -> 43044 links; \u001b[31mCTX\u001b[0m: 20736 links -> 1037 links\n",
      "\n",
      "\u001b[36mPATIENT: C52\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 2137957 links -> 106898 links; \u001b[31mALL\u001b[0m: 3154829 links -> 157742 links; \u001b[31mCTX\u001b[0m: 30810 links -> 1541 links\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 92998 links -> 4650 links; \u001b[31mALL\u001b[0m: 122732 links -> 6137 links; \u001b[31mCTX\u001b[0m: 1638 links -> 82 links\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 2045399 links -> 102270 links; \u001b[31mALL\u001b[0m: 2919329 links -> 145967 links; \u001b[31mCTX\u001b[0m: 28267 links -> 1414 links\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 42441 links -> 2122 links; \u001b[31mALL\u001b[0m: 60439 links -> 3022 links; \u001b[31mCTX\u001b[0m: 2909 links -> 146 links\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 139836 links -> 6992 links; \u001b[31mALL\u001b[0m: 264819 links -> 13241 links; \u001b[31mCTX\u001b[0m: 6428 links -> 322 links\n",
      "\n",
      "\u001b[36mPATIENT: C141\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 2295632 links -> 114782 links; \u001b[31mALL\u001b[0m: 3869515 links -> 193476 links; \u001b[31mCTX\u001b[0m: 68903 links -> 3446 links\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1477707 links -> 73886 links; \u001b[31mALL\u001b[0m: 2228782 links -> 111440 links; \u001b[31mCTX\u001b[0m: 46778 links -> 2339 links\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1289637 links -> 64482 links; \u001b[31mALL\u001b[0m: 2566550 links -> 128328 links; \u001b[31mCTX\u001b[0m: 42421 links -> 2121 links\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1033947 links -> 51698 links; \u001b[31mALL\u001b[0m: 1774524 links -> 88727 links; \u001b[31mCTX\u001b[0m: 34956 links -> 1748 links\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1148476 links -> 57424 links; \u001b[31mALL\u001b[0m: 2361062 links -> 118054 links; \u001b[31mCTX\u001b[0m: 43486 links -> 2175 links\n",
      "\n",
      "\u001b[36mPATIENT: C142\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 2519309 links -> 125966 links; \u001b[31mALL\u001b[0m: 4464518 links -> 223226 links; \u001b[31mCTX\u001b[0m: 68197 links -> 3410 links\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1494034 links -> 74702 links; \u001b[31mALL\u001b[0m: 2267582 links -> 113380 links; \u001b[31mCTX\u001b[0m: 46232 links -> 2312 links\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1592498 links -> 79625 links; \u001b[31mALL\u001b[0m: 3046259 links -> 152313 links; \u001b[31mCTX\u001b[0m: 41594 links -> 2080 links\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1009902 links -> 50496 links; \u001b[31mALL\u001b[0m: 1664900 links -> 83245 links; \u001b[31mCTX\u001b[0m: 31729 links -> 1587 links\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1109231 links -> 55462 links; \u001b[31mALL\u001b[0m: 2252570 links -> 112629 links; \u001b[31mCTX\u001b[0m: 41403 links -> 2071 links\n",
      "\n",
      "\u001b[36mPATIENT: C148\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 2235804 links -> 111791 links; \u001b[31mALL\u001b[0m: 4133344 links -> 206668 links; \u001b[31mCTX\u001b[0m: 76143 links -> 3808 links\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 991162 links -> 49559 links; \u001b[31mALL\u001b[0m: 1841923 links -> 92097 links; \u001b[31mCTX\u001b[0m: 39765 links -> 1989 links\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1142564 links -> 57129 links; \u001b[31mALL\u001b[0m: 2251919 links -> 112596 links; \u001b[31mCTX\u001b[0m: 39495 links -> 1975 links\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 653862 links -> 32694 links; \u001b[31mALL\u001b[0m: 1155320 links -> 57766 links; \u001b[31mCTX\u001b[0m: 26766 links -> 1339 links\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1519719 links -> 75986 links; \u001b[31mALL\u001b[0m: 2689833 links -> 134492 links; \u001b[31mCTX\u001b[0m: 59267 links -> 2964 links\n",
      "\n",
      "\u001b[36mPATIENT: C149\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 2422985 links -> 121150 links; \u001b[31mALL\u001b[0m: 4248674 links -> 212434 links; \u001b[31mCTX\u001b[0m: 70797 links -> 3540 links\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1225351 links -> 61268 links; \u001b[31mALL\u001b[0m: 1837885 links -> 91895 links; \u001b[31mCTX\u001b[0m: 41414 links -> 2071 links\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1241191 links -> 62060 links; \u001b[31mALL\u001b[0m: 2315477 links -> 115774 links; \u001b[31mCTX\u001b[0m: 43885 links -> 2195 links\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 728975 links -> 36449 links; \u001b[31mALL\u001b[0m: 1196133 links -> 59807 links; \u001b[31mCTX\u001b[0m: 24181 links -> 1209 links\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "Filtering - \u001b[31mTF\u001b[0m: 1524507 links -> 76226 links; \u001b[31mALL\u001b[0m: 2589563 links -> 129479 links; \u001b[31mCTX\u001b[0m: 54264 links -> 2714 links\n",
      "\n",
      "CPU times: user 717 ms, sys: 179 ms, total: 896 ms\n",
      "Wall time: 3.78 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "_RELOAD = False\n",
    "filter_quantile = 0.95\n",
    "filter_suffix = str(filter_quantile).replace(\".\", \"_\")\n",
    "\n",
    "if _RELOAD:\n",
    "    print(colored(f'RELOADING -> CHOSEN QUANTILE THRESHOLD: {filter_quantile}', 'green'))\n",
    "    for pat in _PATIENTS:\n",
    "        for data in _FILE_PREFIXES[pat]:\n",
    "            input_TF = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{data}_TF_cor.tsv')\n",
    "            input_ctx = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{data}_TF_ctx.tsv')\n",
    "            input_all = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{data}_cor.tsv')\n",
    "            !python /gpfs/home/bsc08/bsc08890/src/adj_list_to_filter.py -f {input_TF} -q {filter_quantile} &> /dev/null \n",
    "            !python /gpfs/home/bsc08/bsc08890/src/adj_list_to_filter.py -f {input_ctx} -q {filter_quantile} &> /dev/null \n",
    "            !python /gpfs/home/bsc08/bsc08890/src/adj_list_to_filter.py -f {input_all} -q {filter_quantile} &> /dev/null \n",
    "else:\n",
    "    for pat in _PATIENTS:\n",
    "        print(colored(f'PATIENT: {pat}', 'cyan'))\n",
    "        for data in _FILE_PREFIXES[pat]:\n",
    "            try:\n",
    "                grnboost2_meta[pat][data]['filtered_TF'] = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{data}_TF_cor_filtered_{filter_suffix}.pickle'))\n",
    "                grnboost2_meta[pat][data]['filtered_ctx'] = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{data}_TF_ctx_filtered_{filter_suffix}.pickle'))\n",
    "                grnboost2_meta[pat][data]['filtered_all'] = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{data}_cor_filtered_{filter_suffix}.pickle'))\n",
    "                print(colored(data.upper(), 'magenta'))\n",
    "                print(f\"Filtering - {colored('TF', 'red')}: {len(grnboost2_meta[pat][data]['input_TF'])} links -> {len(grnboost2_meta[pat][data]['filtered_TF'])} links; \"\n",
    "                      f\"{colored('ALL', 'red')}: {len(grnboost2_meta[pat][data]['input_all'])} links -> {len(grnboost2_meta[pat][data]['filtered_all'])} links; \"\n",
    "                      f\"{colored('CTX', 'red')}: {len(grnboost2_meta[pat][data]['input_ctx'])} links -> {len(grnboost2_meta[pat][data]['filtered_ctx'])} links\")\n",
    "            except FileNotFoundError:\n",
    "                pass\n",
    "        print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GETTING NETWORKX GRAPHS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36mPATIENT: C51\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 14038 nodes and 150390 edges; \u001b[31mALL\u001b[0m: 16555 nodes and 217847 edges; \u001b[31mCTX\u001b[0m: 1628 nodes and 2306 edges\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 11166 nodes and 36332 edges; \u001b[31mALL\u001b[0m: 11504 nodes and 45996 edges; \u001b[31mCTX\u001b[0m: 1408 nodes and 1423 edges\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 14058 nodes and 131873 edges; \u001b[31mALL\u001b[0m: 16419 nodes and 189398 edges; \u001b[31mCTX\u001b[0m: 1355 nodes and 1975 edges\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 5568 nodes and 15670 edges; \u001b[31mALL\u001b[0m: 6204 nodes and 21727 edges; \u001b[31mCTX\u001b[0m: 521 nodes and 483 edges\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 7532 nodes and 24063 edges; \u001b[31mALL\u001b[0m: 8202 nodes and 43044 edges; \u001b[31mCTX\u001b[0m: 1002 nodes and 1037 edges\n",
      "\n",
      "\u001b[36mPATIENT: C52\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 13653 nodes and 106898 edges; \u001b[31mALL\u001b[0m: 15626 nodes and 157742 edges; \u001b[31mCTX\u001b[0m: 1069 nodes and 1541 edges\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 1945 nodes and 4650 edges; \u001b[31mALL\u001b[0m: 2234 nodes and 6137 edges; \u001b[31mCTX\u001b[0m: 89 nodes and 82 edges\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 13641 nodes and 102270 edges; \u001b[31mALL\u001b[0m: 15493 nodes and 145967 edges; \u001b[31mCTX\u001b[0m: 962 nodes and 1414 edges\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 1061 nodes and 2122 edges; \u001b[31mALL\u001b[0m: 1366 nodes and 3022 edges; \u001b[31mCTX\u001b[0m: 150 nodes and 146 edges\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 2603 nodes and 6992 edges; \u001b[31mALL\u001b[0m: 3347 nodes and 13241 edges; \u001b[31mCTX\u001b[0m: 335 nodes and 322 edges\n",
      "\n",
      "\u001b[36mPATIENT: C141\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 15449 nodes and 114782 edges; \u001b[31mALL\u001b[0m: 17166 nodes and 193476 edges; \u001b[31mCTX\u001b[0m: 1839 nodes and 3446 edges\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 14558 nodes and 73886 edges; \u001b[31mALL\u001b[0m: 15164 nodes and 111440 edges; \u001b[31mCTX\u001b[0m: 1236 nodes and 2339 edges\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 11554 nodes and 64482 edges; \u001b[31mALL\u001b[0m: 12392 nodes and 128328 edges; \u001b[31mCTX\u001b[0m: 1359 nodes and 2121 edges\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 11350 nodes and 51698 edges; \u001b[31mALL\u001b[0m: 11662 nodes and 88727 edges; \u001b[31mCTX\u001b[0m: 1363 nodes and 1748 edges\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 9755 nodes and 57424 edges; \u001b[31mALL\u001b[0m: 10666 nodes and 118054 edges; \u001b[31mCTX\u001b[0m: 1072 nodes and 2175 edges\n",
      "\n",
      "\u001b[36mPATIENT: C142\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 15108 nodes and 125966 edges; \u001b[31mALL\u001b[0m: 17344 nodes and 223226 edges; \u001b[31mCTX\u001b[0m: 1933 nodes and 3410 edges\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 14358 nodes and 74702 edges; \u001b[31mALL\u001b[0m: 15065 nodes and 113380 edges; \u001b[31mCTX\u001b[0m: 1267 nodes and 2312 edges\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 12026 nodes and 79625 edges; \u001b[31mALL\u001b[0m: 13049 nodes and 152313 edges; \u001b[31mCTX\u001b[0m: 1344 nodes and 2080 edges\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 11322 nodes and 50496 edges; \u001b[31mALL\u001b[0m: 11606 nodes and 83245 edges; \u001b[31mCTX\u001b[0m: 1180 nodes and 1587 edges\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 9760 nodes and 55462 edges; \u001b[31mALL\u001b[0m: 10606 nodes and 112629 edges; \u001b[31mCTX\u001b[0m: 1088 nodes and 2071 edges\n",
      "\n",
      "\u001b[36mPATIENT: C148\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 14541 nodes and 111791 edges; \u001b[31mALL\u001b[0m: 16645 nodes and 206668 edges; \u001b[31mCTX\u001b[0m: 2091 nodes and 3808 edges\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 10920 nodes and 49559 edges; \u001b[31mALL\u001b[0m: 11398 nodes and 92097 edges; \u001b[31mCTX\u001b[0m: 1304 nodes and 1989 edges\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 10379 nodes and 57129 edges; \u001b[31mALL\u001b[0m: 11134 nodes and 112596 edges; \u001b[31mCTX\u001b[0m: 1119 nodes and 1975 edges\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 8012 nodes and 32694 edges; \u001b[31mALL\u001b[0m: 8625 nodes and 57766 edges; \u001b[31mCTX\u001b[0m: 1038 nodes and 1339 edges\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 12631 nodes and 75986 edges; \u001b[31mALL\u001b[0m: 13806 nodes and 134492 edges; \u001b[31mCTX\u001b[0m: 1484 nodes and 2964 edges\n",
      "\n",
      "\u001b[36mPATIENT: C149\u001b[0m\n",
      "\u001b[35mRAW_DATA\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 14588 nodes and 121150 edges; \u001b[31mALL\u001b[0m: 17134 nodes and 212434 edges; \u001b[31mCTX\u001b[0m: 2007 nodes and 3540 edges\n",
      "\u001b[35mRAW_DATA_T_CELLS\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 13739 nodes and 61268 edges; \u001b[31mALL\u001b[0m: 14205 nodes and 91895 edges; \u001b[31mCTX\u001b[0m: 1352 nodes and 2071 edges\n",
      "\u001b[35mRAW_DATA_MACROPHAGE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 11574 nodes and 62060 edges; \u001b[31mALL\u001b[0m: 12199 nodes and 115774 edges; \u001b[31mCTX\u001b[0m: 1282 nodes and 2195 edges\n",
      "\u001b[35mRAW_DATA_NK_CELL\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 9525 nodes and 36449 edges; \u001b[31mALL\u001b[0m: 10001 nodes and 59807 edges; \u001b[31mCTX\u001b[0m: 1150 nodes and 1209 edges\n",
      "\u001b[35mRAW_DATA_MONOCYTE\u001b[0m\n",
      "The graphs contain - \u001b[31mTF\u001b[0m: 13266 nodes and 76226 edges; \u001b[31mALL\u001b[0m: 14332 nodes and 129479 edges; \u001b[31mCTX\u001b[0m: 1430 nodes and 2714 edges\n",
      "\n",
      "CPU times: user 2min 39s, sys: 32.9 s, total: 3min 12s\n",
      "Wall time: 3min 20s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "if _RELOAD:\n",
    "    print(colored(f'RE-CREATING GRAPH -> CHOSEN QUANTILE THRESHOLD: {filter_quantile}', 'green'))\n",
    "    for pat in _PATIENTS:\n",
    "        for data in _FILE_PREFIXES[pat]:\n",
    "            input_TF = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{data}_TF_cor.pickle')\n",
    "            input_ctx = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{data}_TF_ctx.pickle')\n",
    "            input_all = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{data}_cor.pickle')\n",
    "            !python /gpfs/home/bsc08/bsc08890/src/adj_list_to_graph.py -f {input_TF} &> /dev/null \n",
    "            !python /gpfs/home/bsc08/bsc08890/src/adj_list_to_graph.py -f {input_ctx} &> /dev/null \n",
    "            !python /gpfs/home/bsc08/bsc08890/src/adj_list_to_graph.py -f {input_all} &> /dev/null \n",
    "else:\n",
    "    for pat in _PATIENTS:\n",
    "        print(colored(f'PATIENT: {pat}', 'cyan'))\n",
    "        for data in _FILE_PREFIXES[pat]:\n",
    "            try:\n",
    "                # all\n",
    "                grnboost2_meta[pat][data]['full_graph_all'] = nx.read_gpickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'nx_graph', f'{data}_cor.gpickle'))\n",
    "                grnboost2_meta[pat][data]['graph_all'] = nx.read_gpickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'nx_graph', f'{data}_cor_filtered_{filter_suffix}.gpickle'))\n",
    "                # TF\n",
    "                grnboost2_meta[pat][data]['full_graph_TF'] = nx.read_gpickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'nx_graph', f'{data}_TF_cor.gpickle'))\n",
    "                grnboost2_meta[pat][data]['graph_TF'] = nx.read_gpickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'nx_graph', f'{data}_TF_cor_filtered_{filter_suffix}.gpickle'))\n",
    "                # ctx\n",
    "                grnboost2_meta[pat][data]['full_graph_ctx'] = nx.read_gpickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'nx_graph', f'{data}_TF_ctx.gpickle'))\n",
    "                grnboost2_meta[pat][data]['graph_ctx'] = nx.read_gpickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'nx_graph', f'{data}_TF_ctx_filtered_{filter_suffix}.gpickle'))\n",
    "                print(colored(data.upper(), 'magenta'))\n",
    "                print(f\"The graphs contain - {colored('TF', 'red')}: {grnboost2_meta[pat][data]['graph_TF'].number_of_nodes()} nodes and {grnboost2_meta[pat][data]['graph_TF'].number_of_edges()} edges; \"\n",
    "                      f\"{colored('ALL', 'red')}: {grnboost2_meta[pat][data]['graph_all'].number_of_nodes()} nodes and {grnboost2_meta[pat][data]['graph_all'].number_of_edges()} edges; \"\n",
    "                      f\"{colored('CTX', 'red')}: {grnboost2_meta[pat][data]['graph_ctx'].number_of_nodes()} nodes and {grnboost2_meta[pat][data]['graph_ctx'].number_of_edges()} edges\")\n",
    "            except FileNotFoundError:\n",
    "                pass\n",
    "        print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### LOOKING AT HIGHEST IMPORTANCES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "323c99806cc64dbfb3b8e9009587f3ca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='Patient:', options=(('C51 (C)', 'C51'), ('C52 (C)', 'C52'), ('C141â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def draw_top_importances_graph(pat, data_type, data, save_fig, top_n):\n",
    "    \n",
    "    seed = 42\n",
    "    dtype_title = _DTYPE_TITLES[_DATA_TYPES.index(data_type)]\n",
    "    data_title = _ALL_TITLES[pat][_ALL_FILE_PREFIXES[pat].index(data)]\n",
    "    \n",
    "    try:\n",
    "        full_G = grnboost2_meta[pat][data][f'full_graph_{data_type}']\n",
    "    except KeyError:\n",
    "        full_G = get_nx_graph(pat, data, data_type)\n",
    "    \n",
    "    out_link_dict = {(st, end): info['importance'] for st, end, info in full_G.edges(data=True)}\n",
    "    out_dict_sorted = dict(sorted(out_link_dict.items(), key=lambda x: x[1], reverse=True)[:top_n])\n",
    "    edges = list(out_dict_sorted.keys())\n",
    "    G = full_G.edge_subgraph(edges)\n",
    "    \n",
    "    label_edges = False\n",
    "    if_alpha_edges = False\n",
    "    alpha = 0.5\n",
    "    pos = nx.nx_agraph.pygraphviz_layout(G, prog=\"dot\")  # nx.nx.spring_layout(G, seed=seed, k=0.2, iterations=20)\n",
    "    cmap=plt.cm.plasma\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(28, 16))   \n",
    "    \n",
    "    draw_graph(G, pos=pos, ax=ax, TF_names=TF_names, label_edges=label_edges, node_size=_NODE_SIZE, if_alpha_edges=if_alpha_edges, cmap=cmap)\n",
    "    \n",
    "    f.suptitle(f'Top {top_n} links that have highest importances\\n{pat}, {dtype_title}, {data_title}', fontsize=20)\n",
    "    \n",
    "    if save_fig:\n",
    "        plt.savefig(os.path.join(_ALL_FIG_DIRS[pat], f'{pat}_{data}_{data_type}_top_{top_n}_link_importances.png'), bbox_inches='tight', dpi=200)    \n",
    "\n",
    "annotate_option = lambda pat, data: \\\n",
    "    f'{data} ({int(full_meta.loc[pat, data.replace(\"raw_data_\", \"\")]) if data != \"raw_data\" else int(full_meta.loc[pat, \"num_cells\"])} patients)'\n",
    "        \n",
    "def pick_pat_dtypes(_):\n",
    "    data_dropdown.options = [(annotate_option(pat_dropdown.value, data), data) for data in _AVAILABLE_DATA[pat_dropdown.value][dtype_dropdown.value]]\n",
    "        \n",
    "pat_dropdown = Dropdown(description='Patient:', options=[(f\"{p} ({full_meta.loc[p, 'group']})\", p) for p in _ALL_PATIENTS])\n",
    "dtype_dropdown = Dropdown(description='Data type:', options=_DATA_TYPES)\n",
    "data_dropdown = Dropdown(description='Data:', options=[(annotate_option(_ALL_PATIENTS[0], data), data) for data in _ALL_FILE_PREFIXES[_ALL_PATIENTS[0]]])\n",
    "pat_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "dtype_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "        \n",
    "widget = interact(\n",
    "    draw_top_importances_graph, \n",
    "    pat=pat_dropdown, \n",
    "    data_type=dtype_dropdown, \n",
    "    data=data_dropdown,\n",
    "    save_fig=[False, True],\n",
    "    top_n=IntSlider(min=1, max=151, step=1, value=50),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### LOOKING AT FIXED TFs AND THEIR TOP TARGETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7ae2c1831515457ba076b58c916ac9ae",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='Patient:', options=(('C51 (C)', 'C51'), ('C52 (C)', 'C52'), ('C141â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def draw_top_link_importance_fixed_TFs_graph(pat, data_type, data, top_n, TFs, save_fig):\n",
    "    \n",
    "    seed = 42\n",
    "    dtype_title = _DTYPE_TITLES[_DATA_TYPES.index(data_type)]\n",
    "    data_title = _ALL_TITLES[pat][_ALL_FILE_PREFIXES[pat].index(data)]\n",
    "    top_n = top_n\n",
    "    edges = []\n",
    "    \n",
    "    try:\n",
    "        full_G = grnboost2_meta[pat][data][f'full_graph_{data_type}']\n",
    "    except KeyError:\n",
    "        full_G = get_nx_graph(pat, data, data_type)\n",
    "    \n",
    "    # Check if all TFs are present in the graph\n",
    "    if not all([tf in full_G.nodes() for tf in TFs]):\n",
    "        warnings.warn(f\"\"\"Next TFs are not present in the graph:\"\"\"\n",
    "                      f\"\"\"{\", \".join([f\"'{tf}'\" for tf in TFs if tf not in full_G.nodes()])}\"\"\")\n",
    "    TFs_to_plot = [tf for tf in TFs if tf in full_G.nodes()]\n",
    "    \n",
    "    for TF in TFs_to_plot:\n",
    "        out_link_dict = {end: info['importance'] for st, end, info in full_G.edges(TF, data=True)}\n",
    "        out_dict_sorted = dict(sorted(out_link_dict.items(), key=lambda x: x[1], reverse=True)[:top_n])\n",
    "        edges += [(TF, target) for target in out_dict_sorted.keys()]\n",
    "    G = full_G.edge_subgraph(edges)\n",
    "    \n",
    "    label_edges = True\n",
    "    if_alpha_edges = False\n",
    "    alpha = 0.5\n",
    "    pos = nx.nx_agraph.pygraphviz_layout(G, prog=\"dot\")  # nx.nx.spring_layout(G, seed=seed, k=0.2, iterations=20)\n",
    "    cmap=plt.cm.plasma\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(30, 15))   \n",
    "    \n",
    "    draw_graph(G, pos=pos, ax=ax, TF_names=TFs_to_plot, label_edges=label_edges, node_size=_NODE_SIZE, if_alpha_edges=if_alpha_edges, cmap=cmap)\n",
    "    \n",
    "    ax.set_title(f'Graph with fixed genes (in green) and their {top_n} targets with highest importances ({pat}, {dtype_title}, {data_title})\\n'\n",
    "                 f'TFs: {\", \".join(TFs_to_plot)}', fontsize=20)\n",
    "    \n",
    "    if save_fig:\n",
    "        plt.savefig(os.path.join(_ALL_FIG_DIRS[pat], f'{pat}_{data}_{data_type}_top_{top_n}_tf_target_link_importances.png'), bbox_inches='tight', dpi=200) \n",
    "\n",
    "annotate_option = lambda pat, data: \\\n",
    "    f'{data} ({int(full_meta.loc[pat, data.replace(\"raw_data_\", \"\")]) if data != \"raw_data\" else int(full_meta.loc[pat, \"num_cells\"])} patients)'\n",
    "        \n",
    "def pick_pat_dtypes(_):\n",
    "    data_dropdown.options = [(annotate_option(pat_dropdown.value, data), data) for data in _AVAILABLE_DATA[pat_dropdown.value][dtype_dropdown.value]]\n",
    "        \n",
    "pat_dropdown = Dropdown(description='Patient:', options=[(f\"{p} ({full_meta.loc[p, 'group']})\", p) for p in _ALL_PATIENTS])\n",
    "dtype_dropdown = Dropdown(description='Data type:', options=_DATA_TYPES)\n",
    "data_dropdown = Dropdown(description='Data:', options=[(annotate_option(_ALL_PATIENTS[0], data), data) for data in _ALL_FILE_PREFIXES[_ALL_PATIENTS[0]]])\n",
    "pat_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "dtype_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "        \n",
    "TFs = ['NFKB1', 'STAT3', 'MYC', 'ETS1'] \n",
    "widget = interact(\n",
    "    draw_top_link_importance_fixed_TFs_graph, \n",
    "    pat=pat_dropdown, \n",
    "    data_type=dtype_dropdown, \n",
    "    data=data_dropdown,\n",
    "    top_n=IntSlider(min=1, max=151, step=1, value=5),\n",
    "    TFs=fixed(TFs),\n",
    "    save_fig=[False, True]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DRAWING GENES OF INTEREST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c9db8217883a4fd3b3670456c6c490b5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='Patient:', options=(('C51 (C)', 'C51'), ('C52 (C)', 'C52'), ('C141â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def draw_genes_from_list_graph(pat, data_type, data, genes, save_fig):\n",
    "    \n",
    "    seed = 42\n",
    "    dtype_title = _DTYPE_TITLES[_DATA_TYPES.index(data_type)]\n",
    "    data_title = _ALL_TITLES[pat][_ALL_FILE_PREFIXES[pat].index(data)]\n",
    "    \n",
    "    try:\n",
    "        full_G = grnboost2_meta[pat][data][f'full_graph_{data_type}']\n",
    "    except KeyError:\n",
    "        full_G = get_nx_graph(pat, data, data_type)\n",
    "    \n",
    "    # Check if all genes are present in the graph\n",
    "    if not all([g in full_G.nodes() for g in genes]):\n",
    "        warnings.warn(f\"\"\"Next genes are not present in the graph: \"\"\"\n",
    "                      f\"\"\"{\", \".join([f\"'{g}'\" for g in genes if g not in full_G.nodes()])}\"\"\")\n",
    "        \n",
    "    genes_to_plot = [g for g in genes if g in full_G.nodes()]\n",
    "\n",
    "    G = full_G.subgraph(genes_to_plot)\n",
    "    \n",
    "    label_edges = True\n",
    "    if_alpha_edges = False\n",
    "    alpha = 0.5\n",
    "    pos = nx.nx_agraph.pygraphviz_layout(G, prog=\"dot\")  # nx.nx.spring_layout(G, seed=seed, k=0.2, iterations=20)\n",
    "    cmap=plt.cm.plasma\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(30, 15))   \n",
    "    \n",
    "    draw_graph(G, pos=pos, ax=ax, TF_names=TF_names, label_edges=label_edges, node_size=_NODE_SIZE, if_alpha_edges=if_alpha_edges, cmap=cmap)\n",
    "        \n",
    "    ax.set_title(f'Graph with fixed genes of interest ({pat}, {dtype_title}, {data_title})', fontsize=20)\n",
    "    \n",
    "    if save_fig:\n",
    "        plt.savefig(os.path.join(_ALL_FIG_DIRS['pat'], f'{pat}_{data}_{data_type}_genes_from_list.png'), bbox_inches='tight', dpi=200) \n",
    "\n",
    "annotate_option = lambda pat, data: \\\n",
    "    f'{data} ({int(full_meta.loc[pat, data.replace(\"raw_data_\", \"\")]) if data != \"raw_data\" else int(full_meta.loc[pat, \"num_cells\"])} patients)'\n",
    "        \n",
    "def pick_pat_dtypes(_):\n",
    "    data_dropdown.options = [(annotate_option(pat_dropdown.value, data), data) for data in _AVAILABLE_DATA[pat_dropdown.value][dtype_dropdown.value]]\n",
    "        \n",
    "pat_dropdown = Dropdown(description='Patient:', options=[(f\"{p} ({full_meta.loc[p, 'group']})\", p) for p in _ALL_PATIENTS])\n",
    "dtype_dropdown = Dropdown(description='Data type:', options=_DATA_TYPES)\n",
    "data_dropdown = Dropdown(description='Data:', options=[(annotate_option(_ALL_PATIENTS[0], data), data) for data in _ALL_FILE_PREFIXES[_ALL_PATIENTS[0]]])\n",
    "pat_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "dtype_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "        \n",
    "genes = ['NFKB1', 'STAT3', 'MYC', 'ETS1'] \n",
    "widget = interact(\n",
    "    draw_genes_from_list_graph, \n",
    "    pat=pat_dropdown, \n",
    "    data_type=dtype_dropdown, \n",
    "    data=data_dropdown,\n",
    "    genes=fixed(genes),\n",
    "    save_fig=[False, True]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COMPARING WITH NDEX/SIGNOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "90734665be844b249fa4b683ec0b5d70",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='Patient:', options=(('C51 (C)', 'C51'), ('C52 (C)', 'C52'), ('C141â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def compare_with_ndex(pat, data_ndex, data_type, data_infer, save_fig):\n",
    "    \n",
    "    seed = 42\n",
    "    dtype_title = _DTYPE_TITLES[_DATA_TYPES.index(data_type)]\n",
    "    data_title = _ALL_TITLES[pat][_ALL_FILE_PREFIXES[pat].index(data_infer)]\n",
    "    G_ndex = published_nets[data_ndex]\n",
    "    # G_ndex = nx.bfs_tree(G_ndex, source='CREB1')  # Creating a tree-like structure of the networkx starting from `source`\n",
    "    \n",
    "    try:\n",
    "        full_G = grnboost2_meta[pat][data_infer][f'full_graph_{data_type}']\n",
    "    except KeyError:\n",
    "        full_G = get_nx_graph(pat, data_infer, data_type)\n",
    "    \n",
    "    f, ax = plt.subplots(1, 2, figsize=(40, 20), sharex=True, sharey=True)  \n",
    "    \n",
    "    pos_ndex = nx.nx_agraph.pygraphviz_layout(G_ndex, prog=\"dot\")  # nx.nx.spring_layout(G, seed=seed, k=0.2, iterations=20)\n",
    "    nodes = nx.draw(G_ndex, pos_ndex, ax=ax[0], with_labels=True, arrowstyle=\"->\", arrowsize=30, \n",
    "                    edge_color='k', node_color='pink', node_size=_NODE_SIZE)\n",
    "    nx.draw_networkx_nodes(G_ndex.subgraph(TF_names), ax=ax[0], pos=pos_ndex, node_color='limegreen', \n",
    "                           node_size=_NODE_SIZE)\n",
    "    ax[0].set_title(f'\"{data_ndex[data_ndex.find(\"/\") + 1:]}\" network from NDEx', fontsize=20)\n",
    "    \n",
    "    ndex_connections = {}\n",
    "    for G_node in G_ndex.nodes():\n",
    "        node = G_node.replace('_', '')\n",
    "        neighbors = set(map(lambda x: x.replace('_', ''), G_ndex.neighbors(G_node))).difference({node})\n",
    "        if node not in ndex_connections:\n",
    "            ndex_connections[node] = neighbors\n",
    "        else:\n",
    "            ndex_connections[node] = ndex_connections[node].union(neighbors)\n",
    "    \n",
    "    # Check if all genes are present in the graph\n",
    "    if not all([g in full_G.nodes() for g in G_ndex.nodes()]):\n",
    "        warnings.warn(f\"\"\"Next genes are not present in the graph: \"\"\"\n",
    "                      f\"\"\"{\", \".join([f\"'{g}'\" for g in G_ndex.nodes() if g not in full_G.nodes()])}\"\"\")\n",
    "        \n",
    "    genes_to_plot = [g for g in G_ndex.nodes() if g in full_G.nodes()]\n",
    "    G = full_G.subgraph(genes_to_plot)\n",
    "    \n",
    "    label_edges = True\n",
    "    if_alpha_edges = False\n",
    "    alpha = 0.5\n",
    "    pos = {node: pos_ndex[node] for node in G.nodes() if node in pos_ndex.keys()}  # nx.nx_agraph.pygraphviz_layout(G, prog=\"dot\")  # nx.nx.spring_layout(G, seed=seed, k=0.2, iterations=20)\n",
    "    cmap=plt.cm.plasma\n",
    "    \n",
    "    draw_graph(G, pos=pos, ax=ax[1], TF_names=TF_names, label_edges=label_edges, node_size=_NODE_SIZE, if_alpha_edges=if_alpha_edges, plot_cmap=True)\n",
    "    ax[1].set_title(f'Inferred GRN - same list of genes ({pat}, {dtype_title}, {data_title})', fontsize=20)\n",
    "    \n",
    "    direct_connections = {}\n",
    "    indirect_connections = {}\n",
    "    \n",
    "    for G_node in G.nodes():\n",
    "        node = G_node.replace('_', '')\n",
    "        # Neighbors search\n",
    "        neighbors = set(map(lambda x: x.replace('_', ''), G.neighbors(G_node))).difference({node})\n",
    "        if node not in direct_connections:\n",
    "            direct_connections[node] = neighbors\n",
    "        else:\n",
    "            direct_connections[node] = direct_connections[node].union(neighbors)\n",
    "        # Descendants search\n",
    "        descendants = set(map(lambda x: x.replace('_', ''), nx.algorithms.dag.descendants(G, G_node))).difference({node})\n",
    "        if node not in indirect_connections:\n",
    "            indirect_connections[node] = descendants\n",
    "        else:\n",
    "            indirect_connections[node] = indirect_connections[node].union(descendants)\n",
    "        \n",
    "    direct_edge_matches = 0\n",
    "    indirect_edge_matches = 0\n",
    "    \n",
    "    for ndex_node, ndex_neighbors in ndex_connections.items():\n",
    "        if ndex_node in direct_connections.keys():\n",
    "            direct_edge_matches += len(ndex_neighbors.intersection(direct_connections[ndex_node]))\n",
    "        if ndex_node in indirect_connections.keys():\n",
    "            indirect_edge_matches += len(ndex_neighbors.intersection(indirect_connections[ndex_node]))\n",
    "            \n",
    "    print(f'Overall, we have {direct_edge_matches} matched connections out of total {G_ndex.number_of_edges()} connections ({direct_edge_matches/G_ndex.number_of_edges()*100:.1f} %) that are in NDEx graph..')\n",
    "    print(f'If we take indirect connections too (descendants), then we have {indirect_edge_matches} matched connections out of total {G_ndex.number_of_edges()} ({indirect_edge_matches/G_ndex.number_of_edges()*100:.1f} %) that are in NDEx graph..')\n",
    "    \n",
    "    if save_fig:\n",
    "        plt.savefig(os.path.join(_ALL_FIG_DIRS[pat], f\"{pat}_{data_infer}_{data_type}_with_ndex_{data_ndex.replace('/', '_').replace(' ', '_')}.png\"), bbox_inches='tight', dpi=200) \n",
    "    \n",
    "annotate_option = lambda pat, data: \\\n",
    "    f'{data} ({int(full_meta.loc[pat, data.replace(\"raw_data_\", \"\")]) if data != \"raw_data\" else int(full_meta.loc[pat, \"num_cells\"])} patients)'\n",
    "        \n",
    "def pick_pat_dtypes(_):\n",
    "    data_dropdown.options = [(annotate_option(pat_dropdown.value, data), data) for data in _AVAILABLE_DATA[pat_dropdown.value][dtype_dropdown.value]]\n",
    "        \n",
    "pat_dropdown = Dropdown(description='Patient:', options=[(f\"{p} ({full_meta.loc[p, 'group']})\", p) for p in _ALL_PATIENTS])\n",
    "dtype_dropdown = Dropdown(description='Data type:', options=_DATA_TYPES)\n",
    "data_dropdown = Dropdown(description='Data:', options=[(annotate_option(_ALL_PATIENTS[0], data), data) for data in _ALL_FILE_PREFIXES[_ALL_PATIENTS[0]]])\n",
    "pat_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "dtype_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "    \n",
    "widgets = interactive(\n",
    "    compare_with_ndex, \n",
    "    pat=pat_dropdown,\n",
    "    data_ndex=_NET_NAMES,\n",
    "    data_type=dtype_dropdown, \n",
    "    data_infer=data_dropdown,\n",
    "    save_fig=[False, True]\n",
    ")   \n",
    "\n",
    "#controls = HBox(widgets.children, layout = Layout(flex_flow='row wrap'))\n",
    "#controls\n",
    "widgets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### FINDING COMMUNITIES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "39b6b928ae3b4891b8e7f473f68b9d1e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(Dropdown(description='Patient:', options=(('C51 (C)', 'C51'), ('C52 (C)', 'C52'), ('C141â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 61.9 ms, sys: 3.85 ms, total: 65.8 ms\n",
      "Wall time: 63.1 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from src.func import *\n",
    "%autoreload 2\n",
    "def plot_communities(pat, data_type, data, algo, wordcloud_display, save_fig):\n",
    "    \n",
    "    seed = 42\n",
    "    dtype_title = _DTYPE_TITLES[_DATA_TYPES.index(data_type)]\n",
    "    data_title = _ALL_TITLES[pat][_ALL_FILE_PREFIXES[pat].index(data)]\n",
    "    \n",
    "    try:\n",
    "        G = grnboost2_meta[pat][data][f'graph_{data_type}']\n",
    "    except KeyError:\n",
    "        G = get_nx_graph(pat, data, data_type, is_filter=True)\n",
    "        \n",
    "    print(f'The chosen graph for pat={colored(pat, \"red\")}, data={colored(data, \"red\")}, data_type={colored(data_type, \"red\")}:')\n",
    "    print(f'Number of nodes: {colored(G.number_of_nodes(), \"cyan\")}; Number of edges: {colored(G.number_of_edges(), \"cyan\")}')\n",
    "        \n",
    "    if algo == 'louvain':\n",
    "        partition = community_louvain.best_partition(G.to_undirected(), weight='importance', random_state=seed)\n",
    "    else:\n",
    "        G_igraph = ig.Graph.from_networkx(G.to_undirected())\n",
    "        la_partition = la.find_partition(G_igraph, la.ModularityVertexPartition, weights='importance', seed=seed)\n",
    "        partition = {G_igraph.vs[node]['_nx_name']: i for i, cluster_nodes in enumerate(la_partition) for node in cluster_nodes}\n",
    "\n",
    "    num_partitions = len(set(partition.values()))\n",
    "    print(f'Number of partitions using {algo} algorithm: {colored(num_partitions, \"cyan\")}')\n",
    "        \n",
    "    f, ax = plt.subplots(figsize=(25, 45))\n",
    "    cmap = ListedColormap(sns.color_palette(cc.glasbey_bw, n_colors=num_partitions).as_hex())\n",
    "    \n",
    "    # Getting positions of squeezed graph\n",
    "    squeezed_G, squeezed_partition = squeeze_graph(G, partition)\n",
    "    print('Computed a squeezed graph representation..')\n",
    "    \n",
    "    squeezed_pos = netgraph_community_layout(squeezed_G, squeezed_partition, seed=seed)  # nx.nx_agraph.pygraphviz_layout(G.to_undirected(), prog=\"sfdp\")  # nx.nx.spring_layout(G, seed=seed, k=0.2, iterations=20)\n",
    "    print('Computed node positions of the squeezed graph representation..')\n",
    "    plot_cloud(G, partition, squeezed_pos, ax=ax, gene_func=gene_func, \n",
    "               filter_genes=True, limit_anno_until=50, \n",
    "               display_func=True if wordcloud_display == 'funcs' else False)\n",
    "    print('Finished plotting a word cloud..')\n",
    "    nx.draw(squeezed_G, squeezed_pos, ax=ax, arrowstyle=\"->\", arrowsize=20, connectionstyle=f'arc3, rad = 0.25', edge_color='k', width=0.4, node_color='k', node_size=50, alpha=0.01)\n",
    "    nx.draw_networkx_nodes(squeezed_G, squeezed_pos, ax=ax, node_size=100, nodelist=list(squeezed_partition.keys()), node_color=list(squeezed_partition.values()), cmap=cmap, alpha=0.01)\n",
    "    print('Finished plotting gene nodes..')\n",
    "    ax.set_title(f'Found communities ({pat}, {dtype_title}, {data_title})', fontsize=30)\n",
    "    \n",
    "    plt.axis('off')\n",
    "    \n",
    "    if save_fig:\n",
    "        plt.savefig(os.path.join(_ALL_FIG_DIRS[pat], f\"{pat}_{data}_{data_type}_communities_{wordcloud_display}.png\"), bbox_inches='tight', dpi=400)\n",
    "    \n",
    "annotate_option = lambda pat, data: \\\n",
    "    f'{data} ({int(full_meta.loc[pat, data.replace(\"raw_data_\", \"\")]) if data != \"raw_data\" else int(full_meta.loc[pat, \"num_cells\"])} patients)'\n",
    "        \n",
    "def pick_pat_dtypes(_):\n",
    "    data_dropdown.options = [(annotate_option(pat_dropdown.value, data), data) for data in _AVAILABLE_DATA[pat_dropdown.value][dtype_dropdown.value]]\n",
    "        \n",
    "pat_dropdown = Dropdown(description='Patient:', options=[(f\"{p} ({full_meta.loc[p, 'group']})\", p) for p in _ALL_PATIENTS])\n",
    "dtype_dropdown = Dropdown(description='Data type:', options=_DATA_TYPES)\n",
    "data_dropdown = Dropdown(description='Data:', options=[(annotate_option(_ALL_PATIENTS[0], data), data) for data in _ALL_FILE_PREFIXES[_ALL_PATIENTS[0]]])\n",
    "pat_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "dtype_dropdown.observe(pick_pat_dtypes, names='value')\n",
    "        \n",
    "widget = interactive(\n",
    "    plot_communities, \n",
    "    {'manual': True},\n",
    "    pat=pat_dropdown, \n",
    "    data_type=dtype_dropdown, \n",
    "    data=data_dropdown,\n",
    "    algo=['leiden', 'louvain'],\n",
    "    wordcloud_display=['funcs', 'genes'],\n",
    "    save_fig=[False, True]\n",
    ")\n",
    "display(widget)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### COMPARING BETWEEN PATIENTS - looking at highest importances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_top_importances_graph(pat_1, pat_2, data, data_type, top_n):\n",
    "    \n",
    "    seed = 42\n",
    "    \n",
    "    # First patient\n",
    "    full_G_1 = grnboost2_meta[pat_1][data][f'graph_{data_type}']\n",
    "    out_link_dict = {(st, end): info['importance'] for st, end, info in full_G_1.edges(data=True)}\n",
    "    out_dict_sorted = dict(sorted(out_link_dict.items(), key=lambda x: x[1], reverse=True)[:top_n])\n",
    "    edges = list(out_dict_sorted.keys())\n",
    "    G_1 = full_G_1.edge_subgraph(edges)\n",
    "    \n",
    "    # Second patient\n",
    "    full_G_2 = grnboost2_meta[pat_2][data][f'graph_{data_type}']\n",
    "    out_link_dict = {(st, end): info['importance'] for st, end, info in full_G_2.edges(data=True)}\n",
    "    out_dict_sorted = dict(sorted(out_link_dict.items(), key=lambda x: x[1], reverse=True)[:top_n])\n",
    "    edges = list(out_dict_sorted.keys())\n",
    "    G_2 = full_G_2.edge_subgraph(edges)\n",
    "    \n",
    "    label_edges = True\n",
    "    if_alpha_edges = False\n",
    "    alpha = 0.5\n",
    "    pos_1 = nx.nx_agraph.pygraphviz_layout(G_1, prog=\"dot\")  # nx.nx.spring_layout(G, seed=seed, k=0.2, iterations=20)\n",
    "    pos_2 = nx.nx_agraph.pygraphviz_layout(G_2, prog=\"dot\")  # nx.nx.spring_layout(G, seed=seed, k=0.2, iterations=20)\n",
    "    cmap=plt.cm.plasma\n",
    "    \n",
    "    f, ax = plt.subplots(1, 2, figsize=(30, 15))   \n",
    "    \n",
    "    draw_graph(G_1, pos=pos_1, ax=ax[0], TF_names=TF_names, label_edges=label_edges, node_size=_NODE_SIZE, if_alpha_edges=if_alpha_edges, cmap=cmap)\n",
    "    draw_graph(G_2, pos=pos_2, ax=ax[1], TF_names=TF_names, label_edges=label_edges, node_size=_NODE_SIZE, if_alpha_edges=if_alpha_edges, cmap=cmap)\n",
    "\n",
    "    ax[0].set_title(pat_1, fontsize=20)\n",
    "    ax[1].set_title(pat_2, fontsize=20)\n",
    "    \n",
    "    f.suptitle(f'Top {top_n} links that have highest importances ({data})', fontsize=22)\n",
    "    \n",
    "pat_1, pat_2 = 'C141', 'C146'\n",
    "        \n",
    "widget = interact(\n",
    "    draw_top_importances_graph, \n",
    "    pat_1=fixed(pat_1),\n",
    "    pat_2=fixed(pat_2),\n",
    "    data=_FILES, \n",
    "    data_type=['ctx', 'all', 'TF'], \n",
    "    top_n=IntSlider(min=1, max=151, step=1, value=10)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COMPARING BETWEEN PATIENTS - looking at fixed TFs and their top targets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COMPARING BETWEEN PATIENTS - comparing with ndex/signor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <span style=\"color:salmon\">GENIE3 + pySCENIC</span>\n",
    "\n",
    "To run in the CLI:\n",
    "```\n",
    "cd\n",
    "python src/run_genie3.py -i res/PilotWorkflow/data/norm_data.tsv -o res/PilotWorkflow/data/genie3_res.tsv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "if False:\n",
    "    print(f'Running GENIE3 with {expression_data.shape[1]} genes..')\n",
    "\n",
    "    # compute the GRN using grnboost2\n",
    "    genie3_net = genie3(expression_data=expression_data, gene_names=None, tf_names=TF_names)\n",
    "\n",
    "    genie3_net"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### <span style=\"color:Aquamarine\">TF-based</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_INPUT_GENIE3 = 'res/PilotWorkflow/data/genie3'\n",
    "_DATA_TYPE = 'genie3_TF_meta'\n",
    "\n",
    "genie3_TF_meta = {\n",
    "    fn: {\n",
    "        'input': pd.read_csv(os.path.join(_INPUT_GENIE3, f'{fn}_TF_cor.tsv'), sep='\\t'),\n",
    "        'filtered': None,\n",
    "        'graph': None\n",
    "    } for fn in _FILES\n",
    "}\n",
    "\n",
    "print('The first 5 rows of GENIE3 output as an example (based on all genes and all cells..')\n",
    "genie3_TF_meta['raw_data']['input'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GENERAL STATISTICS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "widget = interact(genstats, data=_FILES)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will now plot only edges with `importance > 99% quantile`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for data in _FILES:\n",
    "    filter_quantile = 0.99\n",
    "    print(data.upper())\n",
    "    thresh = genie3_TF_meta[data]['input']['importance'].quantile(filter_quantile)\n",
    "    genie3_TF_meta[data]['filtered'] = genie3_TF_meta[data]['input'].loc[lambda x: x.importance > thresh]\n",
    "    genie3_TF_meta[data]['graph'] = nx.from_pandas_edgelist(genie3_TF_meta[data]['filtered'], 'TF', 'target', ['importance', 'rho'], create_using=nx.DiGraph)\n",
    "    print(f\"Filtering: {len(genie3_TF_meta[data]['input'])} links -> {len(genie3_TF_meta[data]['filtered'])} links\")\n",
    "    print(f\"The graph contains {genie3_TF_meta[data]['graph'].number_of_nodes()} nodes and {genie3_TF_meta[data]['graph'].number_of_edges()} edges..\")\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### LOOKING AT HIGHEST IMPORTANCES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_top_importances_graph(data, top_n):\n",
    "    seed = 42\n",
    "    full_G = eval(f\"{_DATA_TYPE}[data]['graph']\")\n",
    "    edges = []\n",
    "    out_link_dict = {(st, end): info['importance'] for st, end, info in full_G.edges(data=True)}\n",
    "    out_dict_sorted = dict(sorted(out_link_dict.items(), key=lambda x: x[1], reverse=True)[:top_n])\n",
    "    edges += list(out_dict_sorted.keys())\n",
    "    G = full_G.edge_subgraph(edges)\n",
    "    \n",
    "    if_alpha_edges = False\n",
    "    alpha = 0.5\n",
    "    pos = nx.nx.spring_layout(G, seed=seed)\n",
    "    cmap=plt.cm.plasma\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(30, 15))      \n",
    "    \n",
    "    edges, importances = zip(*nx.get_edge_attributes(G, 'importance').items())\n",
    "    edges, rhos = zip(*nx.get_edge_attributes(G, 'rho').items())\n",
    "    widths = [w / 20 for w in importances]\n",
    "    \n",
    "    nodes = nx.draw_networkx_nodes(G, pos, with_labels=True, node_color='pink', ax=ax, node_size=_NODE_SIZE)\n",
    "    nx.draw_networkx_nodes(G.subgraph(TF_names), pos=pos, node_color='limegreen', ax=ax, node_size=_NODE_SIZE)\n",
    "    nx.draw_networkx_labels(G, pos, ax=ax)\n",
    "    edges = nx.draw_networkx_edges(\n",
    "        G, pos, connectionstyle='arc3, rad = 0.1', arrowstyle=\"->\", arrowsize=30,\n",
    "        ax=ax, edge_color=rhos, edge_cmap=cmap, width=widths\n",
    "    )\n",
    "\n",
    "    if if_alpha_edges:\n",
    "        edge_importances = [w_dict['importance'] for st, end, w_dict in G.edges(data=True)]\n",
    "        alpha_importances = [(w - min(edge_importances))/(max(edge_importances) - min(edge_importances))*alpha + alpha for w in edge_importances]\n",
    "        # set alpha value for each edge\n",
    "        for i in range(len(alpha_importances)):\n",
    "            edges[i].set_alpha(alpha_importances[i])\n",
    "            \n",
    "    ax.set_title(f'Top {top_n} links that have highest importances ({data})', fontsize=20)\n",
    "            \n",
    "    pc = mpl.collections.PatchCollection(edges, cmap=cmap)\n",
    "    pc.set_array(rhos)\n",
    "    cbar = plt.colorbar(pc)\n",
    "    cbar.ax.set_ylabel('Spearman correlation', rotation=270, fontsize=15, labelpad=15)\n",
    "    \n",
    "    plt.axis('off')\n",
    "        \n",
    "widget = interact(\n",
    "    draw_top_importances_graph, \n",
    "    data=_FILES, \n",
    "    top_n=IntSlider(min=1, max=151, step=1, value=20)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### LOOKING AT FIXED TFs AND THEIR TOP TARGETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def draw_top_link_importance_fixed_TFs_graph(data, top_n):\n",
    "    seed = 42\n",
    "    full_G = eval(f\"{_DATA_TYPE}[data]['graph']\")\n",
    "    TFs_to_plot = ['NFKB1', 'STAT3', 'MYC', 'ETS1']\n",
    "    top_n = top_n\n",
    "    edges = []\n",
    "    for TF in TFs_to_plot:\n",
    "        out_link_dict = {end: info['importance'] for st, end, info in full_G.edges(TF, data=True)}\n",
    "        out_dict_sorted = dict(sorted(out_link_dict.items(), key=lambda x: x[1], reverse=True)[:top_n])\n",
    "        edges += [(TF, target) for target in out_dict_sorted.keys()]\n",
    "    G = full_G.edge_subgraph(edges)\n",
    "    \n",
    "    if_alpha_edges = False\n",
    "    alpha = 0.5\n",
    "    pos = nx.nx.spring_layout(G, seed=seed)\n",
    "    cmap=plt.cm.plasma\n",
    "    \n",
    "    f, ax = plt.subplots(figsize=(30, 15))        \n",
    "    \n",
    "    edges, importances = zip(*nx.get_edge_attributes(G, 'importance').items())\n",
    "    edges, rhos = zip(*nx.get_edge_attributes(G, 'rho').items())\n",
    "    widths = [w / 10 for w in importances]\n",
    "    \n",
    "    nodes = nx.draw_networkx_nodes(G, pos, node_color='pink', ax=ax, node_size=_NODE_SIZE)\n",
    "    nx.draw_networkx_nodes(G.subgraph(TFs_to_plot), pos=pos, node_color='limegreen', ax=ax, node_size=_NODE_SIZE)\n",
    "    nx.draw_networkx_labels(G, pos, ax=ax)\n",
    "    edges = nx.draw_networkx_edges(\n",
    "        G, pos, connectionstyle='arc3, rad = 0.1', arrowstyle=\"->\", arrowsize=30,\n",
    "        ax=ax, edge_color=rhos, edge_cmap=cmap, width=widths\n",
    "    )\n",
    "\n",
    "    if if_alpha_edges:\n",
    "        edge_importances = [w_dict['importance'] for st, end, w_dict in G.edges(data=True)]\n",
    "        alpha_importances = [(w - min(edge_importances))/(max(edge_importances) - min(edge_importances))*alpha + alpha for w in edge_importances]\n",
    "        # set alpha value for each edge\n",
    "        for i in range(len(alpha_importances)):\n",
    "            edges[i].set_alpha(alpha_importances[i])\n",
    "    \n",
    "    ax.set_title(f'Graph with fixed TFs (in green) and their {top_n} targets with highest importances ({data})\\n'\n",
    "                 f'TFs: {\", \".join(TFs_to_plot)}', fontsize=20)\n",
    "            \n",
    "    pc = mpl.collections.PatchCollection(edges, cmap=cmap)\n",
    "    pc.set_array(rhos)\n",
    "    cbar = plt.colorbar(pc)\n",
    "    cbar.ax.set_ylabel('Spearman correlation', rotation=270, fontsize=15, labelpad=15)\n",
    "    \n",
    "    plt.axis('off')\n",
    "        \n",
    "widget = interact(\n",
    "    draw_top_link_importance_fixed_TFs_graph, \n",
    "    data=_FILES,\n",
    "    top_n=IntSlider(min=1, max=151, step=1, value=10)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### <span style=\"color:Aquamarine\">All possible interactions</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_INPUT_GENIE3 = 'res/PilotWorkflow/data/genie3'\n",
    "_FILES = _FILES[1:]  # FIXME BUG: PLEASE CHANGE TO FULL LIST ONCE SBATCH FINISHED\n",
    "_DATA_TYPE = 'genie3_meta'\n",
    "\n",
    "genie3_meta = {\n",
    "    fn: {\n",
    "        'input': pd.read_csv(os.path.join(_INPUT_GENIE3, f'{fn}_cor.tsv'), sep='\\t'),\n",
    "        'filtered': None,\n",
    "        'graph': None\n",
    "    } for fn in _FILES\n",
    "}\n",
    "\n",
    "print('The first 5 rows of GENIE3 output as an example (based on all genes and all cells..')\n",
    "genie3_meta['raw_data_NK_cell']['input'].head()  # FIXME BUG: CHANGE THIS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GENERAL STATISTICS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "widget = interact(genstats, data=_FILES)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will now plot only edges with `importance > 50%/75%/90% quantile`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for data in _FILES:\n",
    "    filter_quantile = 0.99\n",
    "    print(data.upper())\n",
    "    thresh = genie3_meta[data]['input']['importance'].quantile(filter_quantile)\n",
    "    genie3_meta[data]['filtered'] = genie3_meta[data]['input'].loc[lambda x: x.importance > thresh]\n",
    "    genie3_meta[data]['graph'] = nx.from_pandas_edgelist(genie3_meta[data]['filtered'], 'TF', 'target', ['importance', 'rho'], create_using=nx.DiGraph)\n",
    "    print(f\"Filtering: {len(genie3_meta[data]['input'])} links -> {len(genie3_meta[data]['filtered'])} links\")\n",
    "    print(f\"The graph contains {genie3_meta[data]['graph'].number_of_nodes()} nodes and {genie3_meta[data]['graph'].number_of_edges()} edges..\")\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### LOOKING AT HIGHEST IMPORTANCES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "widget = interact(\n",
    "    draw_top_importances_graph, \n",
    "    data=_FILES, \n",
    "    top_n=IntSlider(min=1, max=151, step=1, value=20)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### LOOKING AT FIXED TFs AND THEIR TOP TARGETS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "widget = interact(\n",
    "    draw_top_link_importance_fixed_TFs_graph, \n",
    "    data=_FILES,\n",
    "    top_n=IntSlider(min=1, max=151, step=1, value=10)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RELOAD DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'os' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-3b94069ddec2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m# Setting pathways to files\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0m_PROJ_PATH\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'/gpfs/projects/bsc08/bsc08890'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0m_FMETA\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_PROJ_PATH\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'data/GSE145926_RAW/metadata.tsv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0m_DATA_HOME\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_PROJ_PATH\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'res/covid_19'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'os' is not defined"
     ]
    }
   ],
   "source": [
    "###### RELOAD DATA ######\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Setting pathways to files\n",
    "_PROJ_PATH = '/gpfs/projects/bsc08/bsc08890'\n",
    "_FMETA = os.path.join(_PROJ_PATH, 'data/GSE145926_RAW/metadata.tsv')\n",
    "_DATA_HOME = os.path.join(_PROJ_PATH, 'res/covid_19')\n",
    "\n",
    "# Data types\n",
    "_DATA_TYPES = ['all', 'TF', 'ctx']\n",
    "_DTYPE_TITLES = ['gene-gene links', 'TF-gene links', 'enriched TF-gene links']\n",
    "\n",
    "# Limiting the metadata to certain patients to fit into RAM\n",
    "_FIXED_PATIENTS = ['C51', 'C52', 'C141', 'C142', 'C148', 'C149']\n",
    "_FIXED_CELL_TYPES = ['T_cells', 'Macrophage', 'NK_cell', 'Monocyte']\n",
    "\n",
    "# Loading sample meta data, reordering patients\n",
    "full_meta = pd.read_csv(_FMETA, sep='\\t', index_col=0)\n",
    "full_meta = full_meta.reindex(index=_FIXED_PATIENTS + ['C100', 'C144', 'C143', 'C145', 'C146', 'C152'])\n",
    "\n",
    "# Adding cell-type frequency to the metadat\n",
    "for patient_id in full_meta.index:\n",
    "    cells_metadata = pd.read_csv(os.path.join(_DATA_HOME, patient_id, 'data', 'Seurat', 'cells_metadata.tsv'), sep='\\t')\n",
    "    full_meta.loc[patient_id, 'num_cells'] = len(cells_metadata)\n",
    "    for cell_type, cell_count in cells_metadata['cell_type'].value_counts().iteritems():\n",
    "        full_meta.loc[patient_id, cell_type] = cell_count   \n",
    "full_cell_type_freq = full_meta.iloc[:, 3:]\n",
    "\n",
    "# Finding common cell types\n",
    "_COMMON_CELL_TYPES = full_cell_type_freq.T.dropna().index.to_list()\n",
    "_COMMON_FILE_PREFIXES = ['raw_data', 'raw_data_HVG'] + list(map(lambda x: 'raw_data_' + x, _COMMON_CELL_TYPES))\n",
    "\n",
    "# Limiting the metadata to certain patients to fit into RAM\n",
    "meta = full_meta.loc[_FIXED_PATIENTS, ['group', 'file', 'num_cells'] + _FIXED_CELL_TYPES]\n",
    "cell_type_freq = full_cell_type_freq[_FIXED_CELL_TYPES]\n",
    "\n",
    "# Getting information about all patients\n",
    "_ALL_PATIENTS = full_meta.index.to_list()\n",
    "_ALL_PATIENT_CELL_TYPES = {\n",
    "    pat: full_cell_type_freq.loc[pat].dropna().index.to_list() for pat in _ALL_PATIENTS\n",
    "}\n",
    "_ALL_FILE_PREFIXES = {\n",
    "    pat: ['raw_data'] + list(map(lambda x: 'raw_data_' + x, cell_types)) for pat, cell_types in _ALL_PATIENT_CELL_TYPES.items()\n",
    "}\n",
    "_ALL_TITLES = {\n",
    "    pat: list(map(lambda x: 'all data' if x == 'raw_data' else 'all data, only HVGs' if 'HVG' in x else x.replace('raw_data_', '').replace('_', ' '), files)) \n",
    "    for pat, files in _ALL_FILE_PREFIXES.items()\n",
    "}\n",
    "_ALL_FIG_DIRS = {\n",
    "    pat: os.path.join(_DATA_HOME, pat, 'figs/grnboost2') for pat in _ALL_PATIENTS\n",
    "}\n",
    "        \n",
    "# Setting patient-specific params\n",
    "_PATIENTS = meta.index.to_list()\n",
    "_PATIENT_CELL_TYPES = {\n",
    "    pat: cell_type_freq.loc[pat].dropna().index.to_list() for pat in _PATIENTS\n",
    "}\n",
    "_FILE_PREFIXES = {\n",
    "    pat: ['raw_data'] + list(map(lambda x: 'raw_data_' + x, cell_types)) for pat, cell_types in _PATIENT_CELL_TYPES.items()\n",
    "}\n",
    "_TITLES = {\n",
    "    pat: list(map(lambda x: 'all data' if x == 'raw_data' else 'all data, only HVGs' if 'HVG' in x else x.replace('raw_data_', '').replace('_', ' '), files)) \n",
    "    for pat, files in _FILE_PREFIXES.items()\n",
    "}\n",
    "_FIG_DIRS = {\n",
    "    pat: os.path.join(_DATA_HOME, pat, 'figs/grnboost2') for pat in _PATIENTS\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "for pat in _ALL_PATIENTS:\n",
    "    for fn in _ALL_FILE_PREFIXES[pat]:\n",
    "        TF_cor = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_TF_cor.tsv')\n",
    "        TF_ctx = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_TF_ctx.tsv')\n",
    "        cor = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_cor.tsv')\n",
    "        !python src/adj_list_to_process.py -f {TF_cor} -q 0.95\n",
    "        !python src/adj_list_to_process.py -f {TF_ctx} -q 0.95\n",
    "        !python src/adj_list_to_process.py -f {cor} -q 0.95"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Additional materials"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for pat in _PATIENTS:\n",
    "    print(colored(f'PATIENT: {pat} ({int(meta[\"num_cells\"][pat])} cells)', 'cyan'))\n",
    "    TF_strings, all_strings = [], []\n",
    "    for fn, cell_dtype in zip(_FILE_PREFIXES[pat], ['all_data', 'HVG_data'] + _PATIENT_CELL_TYPES[pat]):\n",
    "        TF_strings.append(colored(cell_dtype, 'green' if fn in _AVAILABLE_DATAS[pat]['TF'] else 'red'))\n",
    "        all_strings.append(colored(cell_dtype, 'green' if fn in _AVAILABLE_DATAS[pat]['all'] else 'red'))\n",
    "    print('TFs')\n",
    "    print(', '.join(TF_strings))\n",
    "    print('All genes')\n",
    "    print(', '.join(all_strings))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### GREASY TEMPLATES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_cells</th>\n",
       "      <th>Macrophage</th>\n",
       "      <th>T_cells</th>\n",
       "      <th>DC</th>\n",
       "      <th>Pre-B_cell_CD34-</th>\n",
       "      <th>Monocyte</th>\n",
       "      <th>NK_cell</th>\n",
       "      <th>B_cell</th>\n",
       "      <th>Epithelial_cells</th>\n",
       "      <th>BM</th>\n",
       "      <th>Pro-B_cell_CD34+</th>\n",
       "      <th>Neutrophils</th>\n",
       "      <th>GMP</th>\n",
       "      <th>Pro-Myelocyte</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C152</th>\n",
       "      <td>2557.0</td>\n",
       "      <td>466.0</td>\n",
       "      <td>397.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>795.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>322.0</td>\n",
       "      <td>201.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     num_cells Macrophage T_cells    DC Pre-B_cell_CD34- Monocyte NK_cell  \\\n",
       "C152    2557.0      466.0   397.0  41.0            174.0    795.0    74.0   \n",
       "\n",
       "     B_cell Epithelial_cells   BM Pro-B_cell_CD34+ Neutrophils   GMP  \\\n",
       "C152  322.0            201.0  7.0             36.0        30.0  13.0   \n",
       "\n",
       "     Pro-Myelocyte  \n",
       "C152           1.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31msbatch\u001b[0m \u001b[36m--job-name\u001b[0m='C152_GREASY_grnboost2' \u001b[36m--ntasks\u001b[0m=12 \u001b[36m--time\u001b[0m='36:00:00' \u001b[36m--cpus-per-task\u001b[0m=16 \u001b[31mgreasy_run_pyscenic.sh\u001b[0m C152\n",
      "\n",
      "../run_pyscenic_full.sh grnboost2 C152 '' /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 1\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Macrophage /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 2\n",
      "../run_pyscenic_full.sh grnboost2 C152 _T_cells /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 3\n",
      "../run_pyscenic_full.sh grnboost2 C152 _DC /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 4\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Pre-B_cell_CD34- /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 5\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Monocyte /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 6\n",
      "../run_pyscenic_full.sh grnboost2 C152 _NK_cell /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 7\n",
      "../run_pyscenic_full.sh grnboost2 C152 _B_cell /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 8\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Epithelial_cells /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 9\n",
      "../run_pyscenic_full.sh grnboost2 C152 _BM /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 10\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Pro-B_cell_CD34+ /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 11\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Neutrophils /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 12\n",
      "../run_pyscenic_full.sh grnboost2 C152 _GMP /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 13\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Pro-Myelocyte /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 4 GREASY 14\n",
      "../run_pyscenic_full.sh grnboost2 C152 '' '' 4 GREASY 15\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Macrophage '' 4 GREASY 16\n",
      "../run_pyscenic_full.sh grnboost2 C152 _T_cells '' 4 GREASY 17\n",
      "../run_pyscenic_full.sh grnboost2 C152 _DC '' 4 GREASY 18\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Pre-B_cell_CD34- '' 4 GREASY 19\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Monocyte '' 4 GREASY 20\n",
      "../run_pyscenic_full.sh grnboost2 C152 _NK_cell '' 4 GREASY 21\n",
      "../run_pyscenic_full.sh grnboost2 C152 _B_cell '' 4 GREASY 22\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Epithelial_cells '' 4 GREASY 23\n",
      "../run_pyscenic_full.sh grnboost2 C152 _BM '' 4 GREASY 24\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Pro-B_cell_CD34+ '' 4 GREASY 25\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Neutrophils '' 4 GREASY 26\n",
      "../run_pyscenic_full.sh grnboost2 C152 _GMP '' 4 GREASY 27\n",
      "../run_pyscenic_full.sh grnboost2 C152 _Pro-Myelocyte '' 4 GREASY 28\n"
     ]
    }
   ],
   "source": [
    "cyan_color = lambda x: colored(x, 'cyan')\n",
    "red_color = lambda x: colored(x, 'red')\n",
    "\n",
    "# Patients - 'C141', 'C142', 'C143', 'C144', 'C145', 'C146', 'C51', 'C52', 'C100', 'C148', 'C149', 'C152'\n",
    "\n",
    "patient = 'C152'\n",
    "method = 'grnboost2'\n",
    "data_types = list(map(lambda x: x.replace('raw_data', '') if x.replace('raw_data', '') else \"''\", _ALL_FILE_PREFIXES[patient]))\n",
    "cell_type_x = full_meta.loc[patient].drop(['file', 'group']).dropna()\n",
    "\n",
    "GBs_per_worker = 8 if cell_type_x['num_cells'] < 8000 else 16\n",
    "n_nodes = 4\n",
    "hours = 36\n",
    "\n",
    "cpus_per_worker = GBs_per_worker // 2\n",
    "cpus_per_task = cpus_per_worker * 4\n",
    "total_avail_cpus = n_nodes * 48\n",
    "n_tasks = total_avail_cpus // cpus_per_task\n",
    "num_workers_per_task = cpus_per_task // cpus_per_worker\n",
    "\n",
    "# n_tasks = 12\n",
    "# cpus_per_task = 12\n",
    "# hours = 24\n",
    "# num_workers_per_task = cpus_per_task // 3\n",
    "\n",
    "qos_debug = False\n",
    "qos_debug_cmd = f' {cyan_color(\"--qos\")}=debug ' if qos_debug else ' '\n",
    "\n",
    "display(pd.DataFrame(cell_type_x).T)\n",
    "\n",
    "# print(f\"{red_color('sbatch')} {cyan_color('--job-name')}='GREASY_{method}_{patient}' {cyan_color('--output')}=./logs/GREASY_{method}_{patient}_%j.out \"\n",
    "#       f\"{cyan_color('--error')}=./logs/GREASY_{method}_{patient}_%j.err {cyan_color('--ntasks')}={n_tasks} {cyan_color('--time')}='{hours}:00:00' \"\n",
    "#       f\"{cyan_color('--cpus-per-task')}={cpus_per_task}{qos_debug_cmd}{red_color('greasy_run_pyscenic')} {patient}\")\n",
    "\n",
    "print(f\"{red_color('sbatch')} {cyan_color('--job-name')}='{patient}_GREASY_{method}' \"\n",
    "      f\"{cyan_color('--ntasks')}={n_tasks} {cyan_color('--time')}='{hours}:00:00' \"\n",
    "      f\"{cyan_color('--cpus-per-task')}={cpus_per_task}{qos_debug_cmd}{red_color('greasy_run_pyscenic.sh')} {patient}\")\n",
    "\n",
    "tasks = ''\n",
    "print()\n",
    "i = 1\n",
    "for d in data_types:\n",
    "    # TF task\n",
    "    task_1 = f\"../run_pyscenic_full.sh {method} {patient} {d} /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt {num_workers_per_task} GREASY {i}\"\n",
    "    print(task_1)\n",
    "    i += 1\n",
    "for d in data_types:\n",
    "    # All genes task\n",
    "    task_2 = f\"../run_pyscenic_full.sh {method} {patient} {d} '' {num_workers_per_task} GREASY {i}\"\n",
    "    print(task_2)\n",
    "    i += 1\n",
    "    tasks += f'{task_1}\\n{task_2}\\n'\n",
    "\n",
    "tasks = tasks.strip()\n",
    "with open(f'/gpfs/projects/bsc08/bsc08890/sbatch/greasy/greasy_tasks_{patient}_{method}', 'w') as f:\n",
    "    f.write(tasks)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### SBATCH TEMPLATES\n",
    "##### ALL PATIENTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>T_cells</th>\n",
       "      <th>Macrophage</th>\n",
       "      <th>NK_cell</th>\n",
       "      <th>Monocyte</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C145</th>\n",
       "      <td>719.0</td>\n",
       "      <td>6960.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>5616.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      T_cells  Macrophage  NK_cell  Monocyte\n",
       "C145    719.0      6960.0    421.0    5616.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='''_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 '' /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Macrophage_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Macrophage /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_T_cells_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _T_cells /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_DC_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _DC /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Pre-B_cell_CD34-_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Pre-B_cell_CD34- /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Monocyte_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Monocyte /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_NK_cell_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _NK_cell /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_B_cell_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _B_cell /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Epithelial_cells_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Epithelial_cells /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Pro-B_cell_CD34+_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Pro-B_cell_CD34+ /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_HSC_-G-CSF_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _HSC_-G-CSF /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Neutrophils_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Neutrophils /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_GMP_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _GMP /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 8 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='''_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 '' '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Macrophage_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Macrophage '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_T_cells_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _T_cells '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_DC_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _DC '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Pre-B_cell_CD34-_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Pre-B_cell_CD34- '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Monocyte_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Monocyte '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_NK_cell_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _NK_cell '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_B_cell_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _B_cell '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Epithelial_cells_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Epithelial_cells '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Pro-B_cell_CD34+_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Pro-B_cell_CD34+ '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_HSC_-G-CSF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _HSC_-G-CSF '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_Neutrophils_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _Neutrophils '' 8 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='40:00:00' \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='_GMP_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full.sh\u001b[0m grnboost2 C145 _GMP '' 8 SBATCH\n"
     ]
    }
   ],
   "source": [
    "cyan_color = lambda x: colored(x, 'cyan')\n",
    "red_color = lambda x: colored(x, 'red')\n",
    "\n",
    "# Patients - 'C141', 'C142', 'C143', 'C144', 'C145', 'C146', 'C51', 'C52', 'C100', 'C148', 'C149', 'C152'\n",
    "\n",
    "patient = 'C145'\n",
    "method = 'grnboost2'\n",
    "data_types = list(map(lambda x: x.replace('raw_data', '') if x.replace('raw_data', '') else \"''\", _ALL_FILE_PREFIXES[patient]))\n",
    "# cell_type_x = meta.loc[patient].drop(['file', 'group']).dropna()\n",
    "\n",
    "# GBs_per_worker = 8 if cell_type_x['num_cells'] < 8000 else 12\n",
    "# increase_cpus_by = 4\n",
    "# n_nodes = 1\n",
    "# hours = 40\n",
    "\n",
    "# cpus_per_worker = GBs_per_worker // 2\n",
    "# cpus_per_task = cpus_per_worker * increase_cpus_by\n",
    "# total_avail_cpus = n_nodes * 48\n",
    "# num_workers_per_task = cpus_per_task // cpus_per_worker\n",
    "\n",
    "# qos_debug = False\n",
    "# qos_debug_cmd = f' {cyan_color(\"--qos\")}=debug ' if qos_debug else ' '\n",
    "\n",
    "hours = 40\n",
    "cpus_per_task = 48\n",
    "num_workers_per_task = 8\n",
    "high_mem = True if data_size > 20000 else False\n",
    "high_mem_cmd = f' {cyan_color(\"--constraint\")}=highmem ' if high_mem else ' '\n",
    "qos_debug = False\n",
    "qos_debug_cmd = f' {cyan_color(\"--qos\")}=debug ' if qos_debug else ' '\n",
    "\n",
    "display(pd.DataFrame(cell_type_freq.loc[patient].dropna()).T)\n",
    "\n",
    "for d in data_types:\n",
    "    print(f\"{red_color('sbatch')} {cyan_color('--time')}='{hours}:00:00'{qos_debug_cmd}{high_mem_cmd}{cyan_color('--cpus-per-task')}={cpus_per_task} \"\n",
    "          f\"{cyan_color('--job-name')}='{d}_TF_{method}' {cyan_color('--workdir')}=. {cyan_color('--ntasks')}=1 \"\n",
    "          f\"{cyan_color('--output')}=/dev/null {cyan_color('--error')}=/dev/null \"\n",
    "          f\"{red_color('run_pyscenic_full.sh')} {method} {patient} {d} /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt {num_workers_per_task} SBATCH\")\n",
    "print()\n",
    "for d in data_types:\n",
    "    print(f\"{red_color('sbatch')} {cyan_color('--time')}='{hours}:00:00'{qos_debug_cmd}{high_mem_cmd}{cyan_color('--cpus-per-task')}={cpus_per_task} \"\n",
    "          f\"{cyan_color('--job-name')}='{d}_{method}' {cyan_color('--workdir')}=. {cyan_color('--ntasks')}=1 \"\n",
    "          f\"{cyan_color('--output')}=/dev/null {cyan_color('--error')}=/dev/null \"\n",
    "          f\"{red_color('run_pyscenic_full.sh')} {method} {patient} {d} '' {num_workers_per_task} SBATCH\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### CELL TYPES AGGREGATED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Macrophage             28008.0\n",
       "Monocyte               15720.0\n",
       "T_cells                 6515.0\n",
       "Neutrophils             5672.0\n",
       "DC                      1605.0\n",
       "NK_cell                 1502.0\n",
       "Epithelial_cells        1242.0\n",
       "B_cell                   591.0\n",
       "Pre-B_cell_CD34-         386.0\n",
       "HSC_-G-CSF                61.0\n",
       "Pro-B_cell_CD34+          46.0\n",
       "GMP                       20.0\n",
       "CMP                       16.0\n",
       "BM                        14.0\n",
       "Gametocytes                4.0\n",
       "Erythroblast               2.0\n",
       "Fibroblasts                2.0\n",
       "Neurons                    1.0\n",
       "Smooth_muscle_cells        1.0\n",
       "Hepatocytes                1.0\n",
       "Keratinocytes              1.0\n",
       "Pro-Myelocyte              1.0\n",
       "dtype: float64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='48:00:00'  \u001b[36m--constraint\u001b[0m=highmem \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='all_data_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 all_data /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='48:00:00'  \u001b[36m--constraint\u001b[0m=highmem \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='all_data_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 all_data '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='48:00:00'  \u001b[36m--constraint\u001b[0m=highmem \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='Macrophage_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Macrophage /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='48:00:00'  \u001b[36m--constraint\u001b[0m=highmem \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='Macrophage_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Macrophage '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='T_cells_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 T_cells /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='T_cells_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 T_cells '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='DC_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 DC /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='DC_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 DC '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Pre-B_cell_CD34-_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Pre-B_cell_CD34- /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Pre-B_cell_CD34-_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Pre-B_cell_CD34- '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='48:00:00'  \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='Monocyte_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Monocyte /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='48:00:00'  \u001b[36m--cpus-per-task\u001b[0m=48 \u001b[36m--job-name\u001b[0m='Monocyte_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Monocyte '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='NK_cell_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 NK_cell /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='NK_cell_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 NK_cell '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='B_cell_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 B_cell /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='B_cell_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 B_cell '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Epithelial_cells_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Epithelial_cells /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Epithelial_cells_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Epithelial_cells '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='BM_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 BM /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='BM_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 BM '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Pro-B_cell_CD34+_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Pro-B_cell_CD34+ /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Pro-B_cell_CD34+_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Pro-B_cell_CD34+ '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='HSC_-G-CSF_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 HSC_-G-CSF /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='HSC_-G-CSF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 HSC_-G-CSF '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Neutrophils_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Neutrophils /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Neutrophils_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Neutrophils '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='CMP_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 CMP /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='CMP_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 CMP '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='GMP_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 GMP /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='GMP_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 GMP '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Gametocytes_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Gametocytes /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Gametocytes_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Gametocytes '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Erythroblast_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Erythroblast /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Erythroblast_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Erythroblast '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Neurons_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Neurons /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Neurons_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Neurons '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Fibroblasts_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Fibroblasts /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Fibroblasts_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Fibroblasts '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Smooth_muscle_cells_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Smooth_muscle_cells /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Smooth_muscle_cells_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Smooth_muscle_cells '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Hepatocytes_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Hepatocytes /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Hepatocytes_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Hepatocytes '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Keratinocytes_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Keratinocytes /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Keratinocytes_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Keratinocytes '' 6 SBATCH\n",
      "\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Pro-Myelocyte_TF_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Pro-Myelocyte /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt 6 SBATCH\n",
      "\u001b[31msbatch\u001b[0m \u001b[36m--time\u001b[0m='24:00:00'  \u001b[36m--cpus-per-task\u001b[0m=24 \u001b[36m--job-name\u001b[0m='Pro-Myelocyte_grnboost2' \u001b[36m--workdir\u001b[0m=. \u001b[36m--ntasks\u001b[0m=1 \u001b[36m--output\u001b[0m=/dev/null \u001b[36m--error\u001b[0m=/dev/null \u001b[31mrun_pyscenic_full_cell_type.sh\u001b[0m grnboost2 Pro-Myelocyte '' 6 SBATCH\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cyan_color = lambda x: colored(x, 'cyan')\n",
    "red_color = lambda x: colored(x, 'red')\n",
    "\n",
    "method = 'grnboost2'\n",
    "cell_types = ['all_data'] + list(full_cell_type_freq.columns)\n",
    "data_sizes = list(full_meta.sum()[2:])\n",
    "\n",
    "display(full_cell_type_freq.sum().sort_values(ascending=False))\n",
    "\n",
    "for cell_type, data_size in zip(cell_types, data_sizes):\n",
    "    \n",
    "    hours = 48 if data_size > 8000 else 24\n",
    "    cpus_per_task = 48 if data_size > 8000 else 24\n",
    "    num_workers_per_task = 6\n",
    "    high_mem = True if data_size > 20000 else False\n",
    "    high_mem_cmd = f' {cyan_color(\"--constraint\")}=highmem ' if high_mem else ' '\n",
    "    qos_debug = False\n",
    "    qos_debug_cmd = f' {cyan_color(\"--qos\")}=debug ' if qos_debug else ' '\n",
    "    hours = 2 if qos_debug else hours\n",
    "\n",
    "    print(f\"{red_color('sbatch')} {cyan_color('--time')}='{hours}:00:00'{qos_debug_cmd}{high_mem_cmd}{cyan_color('--cpus-per-task')}={cpus_per_task} \"\n",
    "          f\"{cyan_color('--job-name')}='{cell_type}_TF_{method}' {cyan_color('--workdir')}=. {cyan_color('--ntasks')}=1 \"\n",
    "          f\"{cyan_color('--output')}=/dev/null {cyan_color('--error')}=/dev/null \"\n",
    "          f\"{red_color('run_pyscenic_full_cell_type.sh')} {method} {cell_type} /gpfs/projects/bsc08/bsc08890/data/TF_lists/lambert2018.txt {num_workers_per_task} SBATCH\")\n",
    "    print(f\"{red_color('sbatch')} {cyan_color('--time')}='{hours}:00:00'{qos_debug_cmd}{high_mem_cmd}{cyan_color('--cpus-per-task')}={cpus_per_task} \"\n",
    "          f\"{cyan_color('--job-name')}='{cell_type}_{method}' {cyan_color('--workdir')}=. {cyan_color('--ntasks')}=1 \"\n",
    "          f\"{cyan_color('--output')}=/dev/null {cyan_color('--error')}=/dev/null \"\n",
    "          f\"{red_color('run_pyscenic_full_cell_type.sh')} {method} {cell_type} '' {num_workers_per_task} SBATCH\")\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### OTHER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'grnboost2_net' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-d819681b90c1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubplots\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfigsize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m15\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m7\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mgrnboost2_net\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimportance\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbins\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkind\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'hist'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlogy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_title\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'GRNBoost2-inferred gene-gene connection importance'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'grnboost2_net' is not defined"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA34AAAGmCAYAAAAj21Z/AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nO3df6zV9X348RcIVS54uYeCONqgxsu84FbjuBXXOi6Ti3BbWakWQkdE0Q1ZRhOBFJa1wiVhNga6Js5s9AZaF+1WWsUS0DsKyEitjPU2WRiQqytoi1MK6vFyC7eVe+/n+8fiTW/vhXO5v+D7vo9HQtL7znnf8zrJO7d99pzP+QzKsiwLAAAAkjX4Ug8AAABA3xJ+AAAAiRN+AAAAiRN+AAAAiRN+AAAAiRN+AAAAietS+P3sZz+Lhx9+OD7xiU/EFVdcEdOmTevSL29oaIhFixZFLpeLkSNHxoIFC+Ldd9/tybwAAABcpCFdedDhw4fjxRdfjNtvvz3OnTvX5V8+b968eO2112LTpk0xePDgWLVqVcyZMyd+9KMfdXtgAAAALs6grtzAvbW1NQYP/r83B7/whS/EO++8E//+7/9+wT379++PT33qU7Fv376YOnVqRET853/+Z0yZMiV27doVlZWVPZ8eAACAgrr0Uc8Po+9i1NbWxtixY9uiLyLitttuixtuuCFqa2sv+vcBAADQPX325S719fVRVlbWYX3ixIlRX1/fV08LAADA7+jSNX7dkc/no6SkpMN6LpeLY8eOdbqnpqYmampqIuL84QgAADAQvPHGG/HOO+/0yu/qs/DrjsWLF8fixYsjIqK8vDzq6uou8UQAAACXRnl5ea/9rj77qGcul4uGhoYO6/l8PnK5XF89LQAAAL+jz8KvrKys02v5fIQTAACgf/VZ+FVVVcWJEyfi5Zdfblurq6uLY8eORVVVVV89LQAAAL+jS9f4nT17Nl588cWIiPjf//3fOH36dDz77LMREfGZz3wmioqKorS0NCoqKmLz5s0REfHHf/zHcdddd8XChQtjw4YNbTdwv+OOO9zDDwAAoB91KfxOnjwZc+fObbf24c+vv/56XH/99dHc3BwtLS3tHrNly5ZYtmxZPPjgg9Ha2hp33313PPHEE700OgAAAF0xKMuy7FIP0Rnf6gkAAAxkvdlEfXaNHwAAAJcH4QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJA44QcAAJC4LoXfkSNHYvr06VFUVBTjxo2L1atXR0tLS8F9dXV1cdddd8WoUaNi1KhRUVlZGQcOHOjx0AAAAHRdwfDL5/NRWVkZgwYNim3btsXq1avj61//eqxZs+aC+44fPx6VlZXR3NwcTz/9dDz99NPR3NwcM2bMiJ///Oe99gIAAAC4sCGFHrBx48ZoamqKrVu3RnFxccyYMSNOnz4d1dXVsXLlyiguLu503wsvvBCNjY3x/PPPx8iRIyMi4lOf+lSMHj06Xnzxxfirv/qr3n0lAAAAdKrgO361tbUxc+bMdoE3f/78aGpqin379p1337lz52LIkCExfPjwtrURI0bEkCFDIsuyHo4NAABAVxUMv/r6+igrK2u3Nn78+CgqKor6+vrz7rv33nujqKgoVqxYESdPnoyTJ0/GsmXLIpfLxdy5c3s+OQAAAF1S8KOe+Xw+SkpKOqzncrnI5/Pn3Tdu3LjYu3dv3H333fHEE09ERMTv/d7vxc6dO2PMmDGd7qmpqYmampqIiDh16lSXXgAAAAAX1me3c3j77bdj7ty5MXny5KitrY3a2tqYPHlyfPazn41f/OIXne5ZvHhx1NXVRV1d3XnjEAAAgItT8B2/XC4XDQ0NHdbz+Xzkcrnz7lu/fn2cO3cunn322Rg6dGhERNx5550xYcKE2LBhQ9u7gAAAAPStgu/4lZWVdbiW7/jx43H27NkO1/79tvr6+rj55pvboi8i4iMf+UjcfPPNcfTo0R6MDAAAwMUoGH5VVVWxc+fOaGxsbFvbsmVLDBs2LCoqKs6777rrrotDhw7FBx980Lb2m9/8Jg4dOhTXX399z6YGAACgywqG35IlS+LKK6+Me+65J3bv3h01NTVRXV0dy5cvb3eLh9LS0njooYfafv6Lv/iLeOutt+Lzn/98vPDCC7Fjx46YM2dOvP3227F48eK+eTUAAAB0UDD8crlc7NmzJ1paWmL27NmxZs2aWLZsWaxdu7bd45qbm6OlpaXt58mTJ8e//du/RWNjY9x3332xcOHCOHv2bOzatStuueWW3n8lAAAAdGpQdpneTb28vDzq6uou9RgAAACXRG82UZ/dzgEAAIDLg/ADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABInPADAABIXJfC78iRIzF9+vQoKiqKcePGxerVq6OlpaVLT7B169b45Cc/GcOGDYuPfvSjMWvWrDhz5kyPhgYAAKDrCoZfPp+PysrKGDRoUGzbti1Wr14dX//612PNmjUFf/mmTZviz//8z6Oqqipqa2tj06ZNMWHChGhubu6V4QEAAChsSKEHbNy4MZqammLr1q1RXFwcM2bMiNOnT0d1dXWsXLkyiouLO933zjvvxLJly+If/uEf4i//8i/b1j//+c/33vQAAAAUVPAdv9ra2pg5c2a7wJs/f340NTXFvn37zrvve9/7XkRE3H///b0wJgAAAN1VMPzq6+ujrKys3dr48eOjqKgo6uvrz7vvwIEDcdNNN8XmzZvj4x//eAwdOjSmTJkSr7zySs+nBgAAoMu6dI1fSUlJh/VcLhf5fP68+06cOBGvvvpqrFu3Lh5//PHYvn17DB8+PGbNmhW//OUvO91TU1MT5eXlUV5eHqdOnbqIlwEAAMD59NntHLIsi1/96lexefPmWLBgQcyaNSt+8IMfxBVXXBFPPvlkp3sWL14cdXV1UVdXF2PGjOmr0QAAAAaUguGXy+WioaGhw3o+n49cLnfBfYMGDYpp06a1rRUXF8fkyZPjyJEj3ZsWAACAi1Yw/MrKyjpcy3f8+PE4e/Zsh2v/ftvEiRMjy7LIsqzdepZlMXiw+8YDAAD0l4IFVlVVFTt37ozGxsa2tS1btsSwYcOioqLivPvuvvvuiIjYu3dv21pDQ0P89Kc/jVtuuaUnMwMAAHARCobfkiVL4sorr4x77rkndu/eHTU1NVFdXR3Lly9vd4uH0tLSeOihh9p+Li8vj8997nPx0EMPxT//8z/HCy+8EH/2Z38WQ4cOjb/+67/um1cDAABAB126xm/Pnj3R0tISs2fPjjVr1sSyZcti7dq17R7X3NwcLS0t7daeeeaZmDNnTixfvjy+8IUvxNChQ+Oll1664LWBAAAA9K5B2e9ehHeZKC8vj7q6uks9BgAAwCXRm03kW1YAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAASJ/wAAAAS16XwO3LkSEyfPj2Kiopi3LhxsXr16mhpaenyk7S2tkZ5eXkMGjQoduzY0e1hAQAAuHhDCj0gn89HZWVlTJo0KbZt2xZHjx6NFStWRGtra6xbt65LT7Jp06Z48803ezwsAAAAF6/gO34bN26Mpqam2Lp1a8yYMSOWLFkSa9asib//+7+P06dPF3yCfD4fX/nKV+Lv/u7vemVgAAAALk7B8KutrY2ZM2dGcXFx29r8+fOjqakp9u3bV/AJHn300fj0pz8d06dP79mkAAAAdEvB8Kuvr4+ysrJ2a+PHj4+ioqKor6+/4N6DBw/Gt771rdiwYUPPpgQAAKDbCoZfPp+PkpKSDuu5XC7y+fwF937pS1+KpUuXRmlpaZeGqampifLy8igvL49Tp051aQ8AAAAXVvDLXbrru9/9brz66quxffv2Lu9ZvHhxLF68OCIiysvL+2o0AACAAaXgO365XC4aGho6rOfz+cjlcp3uOXfuXHz5y1+OVatWRWtra7z//vttXwRz5syZaGxs7OHYAAAAdFXB8CsrK+twLd/x48fj7NmzHa79+9CZM2fizTffjOXLl0cul4tcLhe33HJLRPzfF8PceuutvTA6AAAAXVHwo55VVVWxfv36aGxsjKuvvjoiIrZs2RLDhg2LioqKTveMGDEi9u7d227txIkT8cUvfjEee+yxuPPOO3thdAAAALqiYPgtWbIknnjiibjnnnti1apVcezYsaiuro7ly5e3u8VDaWlpVFRUxObNm2PIkCExbdq0dr/njTfeiIiIP/zDP4wpU6b06osAAADg/AqGXy6Xiz179sTSpUtj9uzZUVJSEsuWLYvq6up2j2tubo6Wlpa+mhMAAIBu6tK3ek6aNCleeumlCz7mw3f0zuf666+PLMu6PBgAAAC9o+CXuwAAAPD/N+EHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQOOEHAACQuC6F35EjR2L69OlRVFQU48aNi9WrV0dLS8sF9/zkJz+JRYsWRWlpaRQVFcVNN90Ua9eujV//+te9MjgAAABdM6TQA/L5fFRWVsakSZNi27ZtcfTo0VixYkW0trbGunXrzrtvy5YtcfTo0Vi1alVMmDAhDh48GI8++mgcPHgwnnvuuV59EQAAAJxfwfDbuHFjNDU1xdatW6O4uDhmzJgRp0+fjurq6li5cmUUFxd3uu9v/uZvYvTo0W0/T5s2La666qp4+OGH4+c//3lcd911vfcqAAAAOK+CH/Wsra2NmTNntgu8+fPnR1NTU+zbt++8+347+j506623RkTEW2+91Z1ZAQAA6IaC4VdfXx9lZWXt1saPHx9FRUVRX19/UU+2f//+GDx4cNx4440XNyUAAADd1qVr/EpKSjqs53K5yOfzXX6iEydOxLp16+K+++6La665ptPH1NTURE1NTUREnDp1qsu/GwAAgPPrl9s5fPDBBzFv3rwYMWJEfOMb3zjv4xYvXhx1dXVRV1cXY8aM6Y/RAAAAklfwHb9cLhcNDQ0d1vP5fORyuYJPkGVZLFy4MA4fPhw//vGPu7QHAACA3lMw/MrKyjpcy3f8+PE4e/Zsh2v/OvPII4/Etm3bYteuXV16PAAAAL2r4Ec9q6qqYufOndHY2Ni2tmXLlhg2bFhUVFRccO/Xvva1ePLJJ+OZZ56JO+64o+fTAgAAcNEKht+SJUviyiuvjHvuuSd2794dNTU1UV1dHcuXL293i4fS0tJ46KGH2n7+l3/5l/jbv/3bWLhwYXzsYx+L//iP/2j754tbAAAA+k+XrvHbs2dPLF26NGbPnh0lJSWxbNmyqK6ubve45ubmaGlpafv5hz/8YUREPPXUU/HUU0+1e+y3v/3teOCBB3o8PAAAAIUNyrIsu9RDdKa8vDzq6uou9RgAAACXRG82Ub/czgEAAIBLR/gBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkTvgBAAAkrkvhd+TIkZg+fXoUFRXFuHHjYvXq1dHS0lJwX0NDQyxatChyuVyMHDkyFixYEO+++26PhwYAAKDrhhR6QD6fj8rKypg0aVJs27Ytjh49GitWrIjW1tZYt27dBffOmzcvXnvttdi0aVMMHjw4Vq1aFXPmzIkf/ehHvfYCAAAAuLCC4bdx48ZoamqKrVu3RnFxccyYMSNOnz4d1dXVsXLlyiguLu503/79++OHP/xh7Nu3L6ZOnRoRER/72MdiypQpsXv37qisrOzdVwIAAECnCn7Us7a2NmbOnNku8ObPnx9NTU2xb9++C+4bO3ZsW/RFRNx2221xww03RG1tbQ/HBgAAoKsKhl99fX2UlZW1Wxs/fnwUFRVFfX39Re2LiJg4ceIF9wEAANC7unSNX0lJSYf1XC4X+Xy+W/uOHTvW6Z6ampqoqamJiIhDhw5FeXl5ofGg3506dSrGjBlzqceATjmfXK6cTS5nzieXq958w6xg+PWnxYsXx+LFiyMiory8POrq6i7xRNCRs8nlzPnkcuVscjlzPrlc9eYbYQU/6pnL5aKhoaHDej6fj1wu1+v7AAAA6F0Fw6+srKzDW4zHjx+Ps2fPdnoN34X2RZz/2j8AAAD6RsHwq6qqip07d0ZjY2Pb2pYtW2LYsGFRUVFxwX0nTpyIl19+uW2trq4ujh07FlVVVQUH+/Ajn3C5cTa5nDmfXK6cTS5nzieXq948m4OyLMsu9IB8Ph+TJk2KP/iDP4hVq1bFsWPHYvny5fHII4+0u4F7aWlpVFRUxObNm9vWZs6cGf/zP/8TGzZsaLuB+zXXXOMG7gAAAP2oS9f47dmzJ1paWmL27NmxZs2aWLZsWaxdu7bd45qbm6OlpaXd2pYtW6KioiIefPDBWLhwYUyePDmef/753n0FAAAAXFDBd/wAAAD4/1vBd/x625EjR2L69OlRVFQU48aNi9WrV3d4p7AzDQ0NsWjRosjlcjFy5MhYsGBBvPvuu/0wMQNJd87nT37yk1i0aFGUlpZGUVFR3HTTTbF27dr49a9/3U9TMxB092/nh1pbW6O8vDwGDRoUO3bs6MNJGWh6cja3bt0an/zkJ2PYsGHx0Y9+NGbNmhVnzpzp44kZSLp7Puvq6uKuu+6KUaNGxahRo6KysjIOHDjQDxMzUPzsZz+Lhx9+OD7xiU/EFVdcEdOmTevSvp40Ub/exy+fz0dlZWVMmjQptm3bFkePHo0VK1ZEa2tru+sFOzNv3rx47bXXYtOmTW3XC86ZM8f1gvSa7p7PLVu2xNGjR2PVqlUxYcKEOHjwYDz66KNx8ODBeO655/rxFZCqnvzt/NCmTZvizTff7ONJGWh6cjY3bdoUS5cujZUrV8b69esjn8/HSy+9FM3Nzf00Panr7vk8fvx4VFZWxh/90R/F008/HRER69evjxkzZsR///d/x3XXXddfL4GEHT58OF588cW4/fbb49y5c13e16MmyvrRY489lpWUlGQNDQ1ta48//ng2bNiwdmu/65VXXskiItu3b1/b2oEDB7KIyHbt2tWnMzNwdPd8njp1qsPaN7/5zSwisjfeeKNPZmVg6e7Z/NB7772XjR49Otu0aVMWEdn27dv7clwGkJ783RwxYkRWU1PTH2MyQHX3fP7TP/1TNnjw4Oz9999vW3vvvfeywYMHZ//4j//YpzMzcLS0tLT953vvvTerqKgouKenTdSvH/Wsra2NmTNnRnFxcdva/Pnzo6mpKfbt23fBfWPHjo2pU6e2rd12221xww03RG1tbZ/OzMDR3fM5evToDmu33nprRES89dZbvT8oA053z+aHHn300fj0pz8d06dP78sxGYC6eza/973vRUTE/fff3+czMnB193yeO3cuhgwZEsOHD29bGzFiRAwZMiQyX41BLxk8+OIzrKdN1K/h19nN28ePHx9FRUWd3uz9QvsiIiZOnHjBfXAxuns+O7N///4YPHhw3Hjjjb05IgNUT87mwYMH41vf+lZs2LChL0dkgOru2Txw4EDcdNNNsXnz5vj4xz8eQ4cOjSlTpsQrr7zS1yMzgHT3fN57771RVFQUK1asiJMnT8bJkydj2bJlkcvlYu7cuX09NpxXT5uoX8Mvn89HSUlJh/VcLhf5fL7X98HF6K1zduLEiVi3bl3cd999cc011/TmiAxQPTmbX/rSl2Lp0qVRWlraV+MxgHX3bJ44cSJeffXVWLduXTz++OOxffv2GD58eMyaNSt++ctf9uXIDCDdPZ/jxo2LvXv3xnPPPRdjx46NsWPHxtatW2Pnzp0xZsyYvhwZLqin/1u137/VE1L2wQcfxLx582LEiBHxjW9841KPwwD33e9+N1599dX46le/eqlHgXayLItf/epXsXnz5liwYEHMmjUrfvCDH8QVV1wRTz755KUejwHu7bffjrlz58bkyZOjtrY2amtrY/LkyfHZz342fvGLX1zq8aDb+vVbPXO5XDQ0NHRYz+fzkcvlLrjv1KlTF70PLkZ3z+eHsiyLhQsXxuHDh+PHP/6xs0mv6c7ZPHfuXHz5y1+OVatWRWtra7z//vtx+vTpiIg4c+ZMNDY2xtVXX92nc5O+nvz3+qBBg9p9fXlxcXFMnjw5jhw50hejMgB193yuX78+zp07F88++2wMHTo0IiLuvPPOmDBhQmzYsCGeeOKJPpsZLqSnTdSv7/iVlZV1+Pzp8ePH4+zZs51+XvVC+yLO/zlX6I7uns8PPfLII7Ft27bYtm2bc0mv6s7ZPHPmTLz55puxfPnyyOVykcvl4pZbbomI//tygw+/gAh6ort/NydOnBhZlnX4oowsy7r1hQfQme6ez/r6+rj55pvboi8i4iMf+UjcfPPNcfTo0T6bFwrpaRP161/Xqqqq2LlzZzQ2NratbdmyJYYNGxYVFRUX3HfixIl4+eWX29bq6uri2LFjUVVV1aczM3B093xGRHzta1+LJ598Mp555pm44447+npUBpjunM0RI0bE3r172/3713/914iIeOyxx+I73/lOv8xO2rr7d/Puu++OiIi9e/e2rTU0NMRPf/rTtv+DAnqqu+fzuuuui0OHDsUHH3zQtvab3/wmDh06FNdff31fjgwX1OMmushbTvTIe++9l1177bVZZWVltmvXruyb3/xmNnz48OwrX/lKu8fdeOON2YMPPthu7a677spuuOGG7Lnnnsuef/757Pd///ezO+64oz/HJ3HdPZ/f+c53sojIHnjggWz//v3t/p08ebK/XwYJ6snfzt/2+uuvu48fvaonZ/Nzn/tcdu2112ZPPfVUtmPHjmzq1KnZ6NGjs/fee68/XwIJ6+75rKury4YMGZJ95jOfyXbs2JFt3749mzVrVjZkyJDsv/7rv/r7ZZCoM2fOZN///vez73//+9ntt9+eTZo0qe3nM2fOZCfkwPkAAAD3SURBVFnW+03Ur+GXZVl2+PDh7E//9E+zq666Krv22muzr371q1lzc3O7x1x33XXZ/fff324tn89nDzzwQDZy5Mjs6quvzr74xS92euNs6InunM/7778/i4hO/33729/u3xdAsrr7t/O3CT/6QnfPZmNjY7ZkyZJs1KhR2VVXXZVNnz49O3jwYD9OzkDQ3fO5e/fu7E/+5E+yXC6X5XK5bOrUqdnevXv7b3CS9+F/J3f27/XXX8+yrPebaFCWuRMlAABAylxBDQAAkDjhBwAAkDjhBwAAkDjhBwAAkDjhBwAAkDjhBwAAkDjhBwAAkDjhBwAAkLj/B+BgeAjdPqZCAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1080x504 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plotting gene-gene link importance distribution\n",
    "f, ax = plt.subplots(figsize=(15, 7))\n",
    "\n",
    "grnboost2_net.importance.plot(ax=ax, bins=100, kind='hist', logy=True)\n",
    "ax.set_title('GRNBoost2-inferred gene-gene connection importance')\n",
    "\n",
    "ax.set_xlabel('Gene-gene connection importance');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "We will now plot only edges with `importance > 20`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "_THRESH = 40\n",
    "\n",
    "# Getting a weighted adjacency matrix\n",
    "adj_mat = pd.pivot_table(grnboost2_net, values='importance', index=['TF'], columns=['target'])\n",
    "n_genes = adj_mat.shape[0]\n",
    "\n",
    "# Filling NaN as zero (no edge)\n",
    "adj_mat = adj_mat.fillna(0)\n",
    "\n",
    "# Setting diagonal to zero (self loops)\n",
    "adj_mat.values[[np.arange(adj_mat.shape[0])]*2] = 0\n",
    "\n",
    "# Filling edges' weights as zero if below threshold\n",
    "adj_mat[adj_mat < _THRESH] = 0\n",
    "\n",
    "# Removing nodes (or nodes) with zeros everywhere\n",
    "zero_rows, zero_cols = set(adj_mat.loc[(adj_mat==0).all(axis=1)].index), set(adj_mat.loc[:, (adj_mat==0).all(axis=0)].columns)\n",
    "zero_genes = zero_rows.intersection(zero_cols)\n",
    "adj_mat = adj_mat.drop(zero_genes, axis=1).drop(zero_genes, axis=0)\n",
    "\n",
    "# Reording dataframe\n",
    "adj_mat = adj_mat[adj_mat.index]\n",
    "\n",
    "print(f'Initially we had {n_genes} genes, now after filtering we have {adj_mat.shape[0]} genes..')\n",
    "\n",
    "adj_mat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "Now we can have a look at how the gene graph of connections looks like."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Creating a graph\n",
    "G = nx.convert_matrix.from_pandas_adjacency(adj_mat, create_using=nx.DiGraph)\n",
    "pos = nx.spring_layout(G, seed=100, k=0.7)\n",
    "\n",
    "f, ax = plt.subplots(figsize=(15, 15))\n",
    "\n",
    "edge_weights = [w_dict['weight'] for st, end, w_dict in G.edges(data=True)]\n",
    "alpha_weights = [(w - min(edge_weights))/(max(edge_weights) - min(edge_weights))*0.5 + 0.5 for w in edge_weights]\n",
    "\n",
    "edges, weights = zip(*nx.get_edge_attributes(G,'weight').items())\n",
    "\n",
    "nodes = nx.draw_networkx_nodes(G, pos, with_labels=True, node_color='pink', ax=ax)\n",
    "nx.draw_networkx_labels(G, pos, ax=ax)\n",
    "edges = nx.draw_networkx_edges(\n",
    "    G, pos, connectionstyle='arc3, rad = 0.1',\n",
    "    ax=ax, edgelist=edges, edge_color=weights, edge_cmap=plt.cm.winter\n",
    ")\n",
    "\n",
    "# set alpha value for each edge\n",
    "for i in range(len(alpha_weights)):\n",
    "    edges[i].set_alpha(alpha_weights[i])\n",
    "    \n",
    "if False:\n",
    "    plt.savefig('res/PilotWorkflow/figs/grnboost2_40_thresh_net.png', dpi=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "q = pd.read_csv('~/res/PilotWorkflow/data/Seurat/raw_data_HVG.tsv', sep='\\t')\n",
    "q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "EXP_MTX_QC_FNAME = 'res/PilotWorkflow/data/Seurat/raw_data_HVG.tsv'\n",
    "HUMAN_TFS_FNAME = 'data/TF_lists/lambert2018.txt'\n",
    "ADJACENCIES_FNAME = 'res/PilotWorkflow/data/grnboost2_res.tsv'\n",
    "ADJACENCIES_COR_FNAME = 'res/PilotWorkflow/data/grnboost2_res_cor.tsv'\n",
    "!pyscenic grn {EXP_MTX_QC_FNAME} {HUMAN_TFS_FNAME} -o {ADJACENCIES_FNAME} --num_workers 18 --transpose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!pyscenic add_cor {ADJACENCIES_FNAME} {EXP_MTX_QC_FNAME} --output {ADJACENCIES_COR_FNAME} --transpose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "q = pd.read_csv('res/PilotWorkflow/data/grnboost2/raw_data_HVG_ctx.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "_PRINT_LOGS = True\n",
    "\n",
    "def get_tf_targ_ctx(df):\n",
    "    tf_target_dict = {'TF': [], 'target': [], 'importance': []}\n",
    "    tf_target_info = (\n",
    "        df.droplevel(axis=0, level=1).droplevel(axis=1, level=0)['TargetGenes']\n",
    "          .map(set)  # transform each list into set\n",
    "          .groupby('TF').agg(lambda x: reduce(lambda a, b: a.union(b), x))  # combine all targets per TF\n",
    "    )\n",
    "    for tf, target_info in tf_target_info.iteritems():\n",
    "        tf_target_dict['TF'] += [tf for target_name, score in target_info]\n",
    "        tf_target_dict['target'] += [target_name for target_name, score in target_info]\n",
    "        tf_target_dict['importance'] += [score for target_name, score in target_info]\n",
    "    return pd.DataFrame(tf_target_dict)\n",
    "\n",
    "\n",
    "_AVAILABLE_DATAS = {}\n",
    "grnboost2_meta = {}\n",
    "p_bar = tqdm(_PATIENTS)\n",
    "\n",
    "for pat in p_bar:\n",
    "    p_bar.set_description(f'{pat} patient')\n",
    "    grnboost2_meta[pat] = {}\n",
    "    _AVAILABLE_DATAS[pat] = {}\n",
    "    _AVAILABLE_DATAS[pat]['TF'] = []\n",
    "    _AVAILABLE_DATAS[pat]['ctx'] = []\n",
    "    _AVAILABLE_DATAS[pat]['all'] = []\n",
    "    for fn in _FILE_PREFIXES[pat]:\n",
    "        try:\n",
    "            input_TF = pd.read_csv(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_TF_cor.tsv'), sep='\\t')\n",
    "            ctx = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_TF_ctx.pickle'))\n",
    "            input_ctx = get_tf_targ_ctx(ctx).merge(input_TF, how='left')\n",
    "            _AVAILABLE_DATAS[pat]['TF'].append(fn)\n",
    "            _AVAILABLE_DATAS[pat]['ctx'].append(fn)\n",
    "        except (FileNotFoundError, pd.errors.EmptyDataError):\n",
    "            input_TF = None\n",
    "            ctx = None\n",
    "            input_ctx = None\n",
    "        try:\n",
    "            input_all = pd.read_csv(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_cor.tsv'), sep='\\t')\n",
    "            _AVAILABLE_DATAS[pat]['all'].append(fn)\n",
    "        except (FileNotFoundError, pd.errors.EmptyDataError):\n",
    "            input_all = None\n",
    "        grnboost2_meta[pat][fn] = {\n",
    "        'input_TF': input_TF,\n",
    "        'input_all': input_all, \n",
    "        'filtered_TF': None,\n",
    "        'filtered_all': None,\n",
    "        'graph_TF': None,\n",
    "        'graph_all': None,\n",
    "        'full_graph_TF': None,\n",
    "        'full_graph_all': None,\n",
    "        'ctx': ctx, \n",
    "        'input_ctx': input_ctx,\n",
    "        'filtered_ctx': None,\n",
    "        'graph': None,\n",
    "        'full_graph_ctx': None\n",
    "        }\n",
    "\n",
    "if _PRINT_LOGS:\n",
    "    for pat in _PATIENTS:\n",
    "        print(colored(f'PATIENT: {pat}', 'cyan'))\n",
    "        TF_strings, all_strings = [], []\n",
    "        for fn, cell_dtype in zip(_FILE_PREFIXES[pat], ['all_data', 'HVG_data'] + _PATIENT_CELL_TYPES[pat]):\n",
    "            TF_strings.append(colored(cell_dtype, 'green' if fn in _AVAILABLE_DATAS[pat]['TF'] else 'red'))\n",
    "            all_strings.append(colored(cell_dtype, 'green' if fn in _AVAILABLE_DATAS[pat]['all'] else 'red'))\n",
    "        print('TFs')\n",
    "        print(', '.join(TF_strings))\n",
    "        print('All genes')\n",
    "        print(', '.join(all_strings))\n",
    "        print()\n",
    "    \n",
    "\n",
    "print('The first 2 rows of GRNBoost2 output as an example (based on all genes and all cells)..')\n",
    "display(grnboost2_meta[_PATIENTS[0]]['raw_data']['input_all'].head(2))\n",
    "print('After processing of ctx output..')\n",
    "display(grnboost2_meta[_PATIENTS[0]]['raw_data']['input_ctx'].head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_process(i):\n",
    "    _AVAILABLE_DATAS = {}\n",
    "    grnboost2_meta = {}\n",
    "    p_bar = tqdm(_PATIENTS[:i])\n",
    "\n",
    "    for pat in p_bar:\n",
    "        p_bar.set_description(f'{pat} patient')\n",
    "        grnboost2_meta[pat] = {}\n",
    "        _AVAILABLE_DATAS[pat] = {}\n",
    "        _AVAILABLE_DATAS[pat]['TF'] = []\n",
    "        _AVAILABLE_DATAS[pat]['ctx'] = []\n",
    "        _AVAILABLE_DATAS[pat]['all'] = []\n",
    "        for fn in _FILE_PREFIXES[pat]:\n",
    "            try:\n",
    "                input_TF = pd.read_csv(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_TF_cor.tsv'), sep='\\t')\n",
    "                ctx = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_TF_ctx.pickle'))\n",
    "                input_ctx = get_tf_targ_ctx(ctx).merge(input_TF, how='left')\n",
    "                _AVAILABLE_DATAS[pat]['TF'].append(fn)\n",
    "                _AVAILABLE_DATAS[pat]['ctx'].append(fn)\n",
    "            except:\n",
    "                input_TF = None\n",
    "                ctx = None\n",
    "                input_ctx = None\n",
    "            try:\n",
    "                input_all = pd.read_csv(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_cor.tsv'), sep='\\t')\n",
    "                _AVAILABLE_DATAS[pat]['all'].append(fn)\n",
    "            except:\n",
    "                input_all = None\n",
    "            grnboost2_meta[pat][fn] = {'input_TF': input_TF, 'input_all': input_all, 'filtered_TF': None,'filtered_all': None,'graph_TF': None,'graph_all': None,'full_graph_TF': None,'full_graph_all': None,'ctx': ctx, 'input_ctx': input_ctx,'filtered_ctx': None,'graph': None,'full_graph_ctx': None}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%lprun -f data_process data_process(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for pat in _PATIENTS:\n",
    "    for fn in _FILE_PREFIXES[pat]:\n",
    "        full_fn = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_cor.tsv')\n",
    "        !python src/grn_file_to_pickle.py -f {full_fn} >> /dev/null\n",
    "        full_fn = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_TF_cor.tsv')\n",
    "        !python src/grn_file_to_pickle.py -f {full_fn} >> /dev/null\n",
    "        full_fn = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{fn}_TF_ctx.tsv')\n",
    "        !python src/grn_file_to_pickle.py -f {full_fn} >> /dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "filter_quantile = 0.8\n",
    "print(colored(f'CHOSEN QUANTILE THRESHOLD: {filter_quantile}', 'green'))\n",
    "for pat in _PATIENTS:\n",
    "    print(colored(f'PATIENT: {pat}', 'cyan'))\n",
    "    for data in _FILE_PREFIXES[pat]:\n",
    "        try:\n",
    "            print(colored(data.upper(), 'magenta'))\n",
    "            # TF\n",
    "            thresh = grnboost2_meta[pat][data]['input_TF']['importance'].quantile(filter_quantile)\n",
    "            grnboost2_meta[pat][data]['filtered_TF'] = grnboost2_meta[pat][data]['input_TF'].loc[lambda x: x.importance > thresh]\n",
    "            grnboost2_meta[pat][data]['graph_TF'] = nx.from_pandas_edgelist(grnboost2_meta[pat][data]['filtered_TF'], 'TF', 'target', ['importance', 'rho'], create_using=nx.DiGraph)\n",
    "            grnboost2_meta[pat][data]['full_graph_TF'] = nx.from_pandas_edgelist(grnboost2_meta[pat][data]['input_TF'], 'TF', 'target', ['importance', 'rho'], create_using=nx.DiGraph)\n",
    "            # all\n",
    "            thresh = grnboost2_meta[pat][data]['input_all']['importance'].quantile(filter_quantile)\n",
    "            grnboost2_meta[pat][data]['filtered_all'] = grnboost2_meta[pat][data]['input_all'].loc[lambda x: x.importance > thresh]\n",
    "            grnboost2_meta[pat][data]['graph_all'] = nx.from_pandas_edgelist(grnboost2_meta[pat][data]['filtered_all'], 'TF', 'target', ['importance', 'rho'], create_using=nx.DiGraph)\n",
    "            grnboost2_meta[pat][data]['full_graph_all'] = nx.from_pandas_edgelist(grnboost2_meta[pat][data]['input_all'], 'TF', 'target', ['importance', 'rho'], create_using=nx.DiGraph)\n",
    "            # ctx\n",
    "            thresh = grnboost2_meta[pat][data]['input_ctx']['importance'].quantile(filter_quantile)\n",
    "            grnboost2_meta[pat][data]['filtered_ctx'] = grnboost2_meta[pat][data]['input_ctx'].loc[lambda x: x.importance > thresh]\n",
    "            grnboost2_meta[pat][data]['graph_ctx'] = nx.from_pandas_edgelist(grnboost2_meta[pat][data]['filtered_ctx'], 'TF', 'target', ['importance', 'rho'], create_using=nx.DiGraph)\n",
    "            grnboost2_meta[pat][data]['full_graph_ctx'] = nx.from_pandas_edgelist(grnboost2_meta[pat][data]['input_ctx'], 'TF', 'target', ['importance', 'rho'], create_using=nx.DiGraph)\n",
    "            print(f\"Filtering - {colored('TF', 'red')}: {len(grnboost2_meta[pat][data]['input_TF'])} links -> {len(grnboost2_meta[pat][data]['filtered_TF'])} links; \"\n",
    "                  f\"{colored('ALL', 'red')}: {len(grnboost2_meta[pat][data]['input_all'])} links -> {len(grnboost2_meta[pat][data]['filtered_all'])}; \"\n",
    "                  f\"{colored('CTX', 'red')}: {len(grnboost2_meta[pat][data]['input_ctx'])} links -> {len(grnboost2_meta[pat][data]['filtered_ctx'])}\")\n",
    "            print(f\"The graphs contain - {colored('TF', 'red')}: {grnboost2_meta[pat][data]['graph_TF'].number_of_nodes()} nodes and {grnboost2_meta[pat][data]['graph_TF'].number_of_edges()} edges; \"\n",
    "                  f\"{colored('ALL', 'red')}: {grnboost2_meta[pat][data]['graph_all'].number_of_nodes()} nodes and {grnboost2_meta[pat][data]['graph_all'].number_of_edges()} edges; \"\n",
    "                  f\"{colored('CTX', 'red')}: {grnboost2_meta[pat][data]['graph_ctx'].number_of_nodes()} nodes and {grnboost2_meta[pat][data]['graph_ctx'].number_of_edges()} edges\")\n",
    "            print()\n",
    "        except TypeError:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for pat in _PATIENTS:\n",
    "    for data in _FILE_PREFIXES[pat]:\n",
    "        input_TF = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{data}_TF_cor.tsv')\n",
    "        input_ctx = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{data}_TF_ctx.tsv')\n",
    "        input_all = os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', f'{data}_cor.tsv')\n",
    "        !python /gpfs/home/bsc08/bsc08890/src/adj_list_to_process.py -f {input_TF} &> /dev/null \n",
    "        !python /gpfs/home/bsc08/bsc08890/src/adj_list_to_process.py -f {input_ctx} &> /dev/null \n",
    "        !python /gpfs/home/bsc08/bsc08890/src/adj_list_to_process.py -f {input_all} &> /dev/null "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e651779b9eff4a5abf9d401175be6787",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=6.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "The first 2 rows of GRNBoost2 output as an example (based on all genes and all cells)..\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>target</th>\n",
       "      <th>importance</th>\n",
       "      <th>regulation</th>\n",
       "      <th>rho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LYZ</td>\n",
       "      <td>AC020656.1</td>\n",
       "      <td>315.984263</td>\n",
       "      <td>1</td>\n",
       "      <td>0.696023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC020656.1</td>\n",
       "      <td>LYZ</td>\n",
       "      <td>293.557246</td>\n",
       "      <td>1</td>\n",
       "      <td>0.696023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           TF      target  importance  regulation       rho\n",
       "0         LYZ  AC020656.1  315.984263           1  0.696023\n",
       "1  AC020656.1         LYZ  293.557246           1  0.696023"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After processing of ctx output..\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>target</th>\n",
       "      <th>importance</th>\n",
       "      <th>regulation</th>\n",
       "      <th>rho</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARNT</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>1.552042</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.031551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ARNT</td>\n",
       "      <td>LCMT1</td>\n",
       "      <td>2.414943</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.032334</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     TF target  importance  regulation       rho\n",
       "0  ARNT  NOC2L    1.552042         1.0  0.031551\n",
       "1  ARNT  LCMT1    2.414943         1.0  0.032334"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 11.6 s, sys: 2.84 s, total: 14.5 s\n",
      "Wall time: 22.7 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "_PRINT_LOGS = False\n",
    "_AVAILABLE_DATAS = {}\n",
    "grnboost2_meta = {}\n",
    "p_bar = tqdm(_PATIENTS)\n",
    "\n",
    "for pat in p_bar:\n",
    "    p_bar.set_description(f'{pat} patient')\n",
    "    grnboost2_meta[pat] = {}\n",
    "    _AVAILABLE_DATAS[pat] = {}\n",
    "    _AVAILABLE_DATAS[pat]['TF'] = []\n",
    "    _AVAILABLE_DATAS[pat]['ctx'] = []\n",
    "    _AVAILABLE_DATAS[pat]['all'] = []\n",
    "    for fn in _FILE_PREFIXES[pat]:\n",
    "        try:\n",
    "            input_TF = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_TF_cor.pickle'))\n",
    "            input_ctx = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_TF_ctx.pickle'))\n",
    "            _AVAILABLE_DATAS[pat]['TF'].append(fn)\n",
    "            _AVAILABLE_DATAS[pat]['ctx'].append(fn)\n",
    "        except (FileNotFoundError, pd.errors.EmptyDataError):\n",
    "            input_TF = None\n",
    "            ctx = None\n",
    "            input_ctx = None\n",
    "        try:\n",
    "            input_all = pd.read_pickle(os.path.join(_DATA_HOME, pat, 'data', 'grnboost2', 'pickle', f'{fn}_cor.pickle'))\n",
    "            _AVAILABLE_DATAS[pat]['all'].append(fn)\n",
    "        except (FileNotFoundError, pd.errors.EmptyDataError):\n",
    "            input_all = None\n",
    "        grnboost2_meta[pat][fn] = {\n",
    "        'input_TF': input_TF,\n",
    "        'input_all': input_all, \n",
    "        'filtered_TF': None,\n",
    "        'filtered_all': None,\n",
    "        'graph_TF': None,\n",
    "        'graph_all': None,\n",
    "        'full_graph_TF': None,\n",
    "        'full_graph_all': None,\n",
    "        'input_ctx': input_ctx,\n",
    "        'filtered_ctx': None,\n",
    "        'graph': None,\n",
    "        'full_graph_ctx': None\n",
    "        }\n",
    "\n",
    "if _PRINT_LOGS:\n",
    "    for pat in _PATIENTS:\n",
    "        print(colored(f'PATIENT: {pat}', 'cyan'))\n",
    "        TF_strings, all_strings = [], []\n",
    "        for fn, cell_dtype in zip(_FILE_PREFIXES[pat], ['all_data'] + _PATIENT_CELL_TYPES[pat]):\n",
    "            TF_strings.append(colored(cell_dtype, 'green' if fn in _AVAILABLE_DATAS[pat]['TF'] else 'red'))\n",
    "            all_strings.append(colored(cell_dtype, 'green' if fn in _AVAILABLE_DATAS[pat]['all'] else 'red'))\n",
    "        print('TFs')\n",
    "        print(', '.join(TF_strings))\n",
    "        print('All genes')\n",
    "        print(', '.join(all_strings))\n",
    "        print()\n",
    "    \n",
    "\n",
    "print('The first 2 rows of GRNBoost2 output as an example (based on all genes and all cells)..')\n",
    "display(grnboost2_meta[_PATIENTS[0]]['raw_data']['input_all'].head(2))\n",
    "print('After processing of ctx output..')\n",
    "display(grnboost2_meta[_PATIENTS[0]]['raw_data']['input_ctx'].head(2))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
